{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bad563eb-f9ea-4a05-9408-f4cb28c53912",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘proxy’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Matrix’:\n",
      "\n",
      "    as.matrix\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    as.dist, dist\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    as.matrix\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0.\u001b[31m9000\u001b[39m     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.1          \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2          \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.1          \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m masks \u001b[34mMatrix\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mpack()\u001b[39m   masks \u001b[34mMatrix\u001b[39m::pack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32munpack()\u001b[39m masks \u001b[34mMatrix\u001b[39m::unpack()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.13.2\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "here() starts at /home/jpm73279/Mendieta_et_al_comparative_single_cell\n",
      "\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m54659\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chrom, acr_number\n",
      "\u001b[32mdbl\u001b[39m (3): start, stop, accessability\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m16271098\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m3\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): gene_name, barcode\n",
      "\u001b[32mdbl\u001b[39m (1): accessability\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "# load libraries\n",
    "library(dplyr)\n",
    "library(\"edgeR\")\n",
    "library(Matrix)\n",
    "library(gplots)\n",
    "library(RColorBrewer)\n",
    "library(irlba)\n",
    "library(proxy)\n",
    "library(png)\n",
    "library(tidyverse)\n",
    "library(ComplexHeatmap)\n",
    "library(here)\n",
    "library(rsample)\n",
    "library(purrr)\n",
    "\n",
    "\n",
    "#working_dir <- \"/scratch/jpm73279/comparative_single_cell/dev_location/entropy_calc.CTs/R_implemenation_test\"\n",
    "\n",
    "# load arguments\n",
    "args <- commandArgs(T)\n",
    "#input_data <- as.character(args[1])\n",
    "#meta <- as.character(args[2])\n",
    "#peak_file <- as.character(args[3])\n",
    "#meta_slot <- as.character(args[4])\n",
    "#replicate_slot <- as.character(args[5])\n",
    "#prefix <- as.character(args[6])\n",
    "#print(prefix)\n",
    "\n",
    "\n",
    "## Testing Data Original \n",
    "input_data <- \"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/replicate_analysis_one_off/zm/zm.peaks_accessability.txt\"\n",
    "meta <- \"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/replicate_analysis_one_off/zm/Zm.leaf_annot.V5.meta.frozen.txt\"\n",
    "peak_file <- \"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/replicate_analysis_one_off/zm/zm.peaks.500bp_peaks.bed\"\n",
    "meta_slot <- \"final_annotation_n\"\n",
    "replicate_slot <- \"sampleID\"\n",
    "prefix <- \"testing_replicate_merge.normalization_edgeR.FDR\"\n",
    "\n",
    "\n",
    "## Read Inputs \n",
    "input <- input_data\n",
    "bed_file_read <- read_delim(peak_file, col_names = c(\"chrom\", \"start\", \"stop\", \"acr_number\", \"accessability\"))\n",
    "meta_data <- read.delim(meta)\n",
    "\n",
    "## Use the column for meta_data for cell type ACR calling \n",
    "meta_slot_var <- c(meta_slot)\n",
    "rep_slot_var <- c(replicate_slot)\n",
    "##!!sym(meta_slot_var)\n",
    "\n",
    "\n",
    "raw_cpm_counts_all_genes <- read_delim(input, delim=\"\\t\", col_names = c(\"gene_name\", \"barcode\", \"accessability\")) %>%\n",
    "    dplyr::mutate(cellID = barcode)  %>%\n",
    "    dplyr::mutate(geneID = gene_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0a745065-d925-4760-9059-772870727dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "calculating_specificity <- function(x, threads=30){\n",
    "    \n",
    "    # add pseudo-count\n",
    "    x <- x+1\n",
    "    \n",
    "    # convert to probability distribution\n",
    "    p <- t(apply(x, 1, function(z){\n",
    "        z/sum(z)\n",
    "    }))\n",
    "    hp <- apply(p, 1, function(z){\n",
    "        z <- z[z > 0]\n",
    "        -1*sum(z*(log2(z)))\n",
    "        \n",
    "    })\n",
    "    sp <- apply(p, 2, function(z){\n",
    "        hp - log2(z)\n",
    "    })\n",
    "    \n",
    "    return(sp)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f59bc316-9617-4c58-95c2-262f01bbd508",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_meta_cpm_information <- left_join(meta_data, raw_cpm_counts_all_genes, by = c(\"cellID\"))  %>%\n",
    "    #mutate(safe_cluster_name = str_c(\"Louvain_C\", LouvainClusters, sep =\"_\"))  %>%\n",
    "    #dplyr::select(-LouvainClusters)  %>%\n",
    "    group_by(!!sym(meta_slot_var), geneID)  %>%\n",
    "    summarise(counts = sum(accessability, na.rm = TRUE))\n",
    "\n",
    "### Alt CPM Calc\n",
    "merged_meta_cpm_information_copied <- merged_meta_cpm_information\n",
    "catch <- merged_meta_cpm_information_copied  %>%\n",
    "    group_by(!!sym(meta_slot_var)) %>%\n",
    "    group_map(~(cpm(.x$counts, log = FALSE, group = .f)), .keep = TRUE) %>% \n",
    "    unlist()\n",
    "\n",
    "\n",
    "\n",
    "caught_values <- as_tibble(catch)\n",
    "see <- ungroup(merged_meta_cpm_information_copied)\n",
    "merged_meta_cpm_information_copied<- bind_cols(merged_meta_cpm_information_copied,caught_values)  %>% \n",
    "    rename(grouped_CPM = value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbc7a876-ef1a-436d-ac5d-c1606be582f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "options(repr.plot.width=20, repr.plot.height=10)\n",
    "ggplot(merged_meta_cpm_information_copied, aes(x = counts, y = grouped_CPM, color = final_annotation_n)) + \n",
    "geom_point()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e0d3a6-4bb3-4f89-91fd-28c70dad95a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined %>% \n",
    "    ungroup() %>% \n",
    "    group_by(geneID) %>% \n",
    "        summarise(mcount = mean(filterd_low_val, na.rm = TRUE),\n",
    "        varcount = var(filterd_low_val, na.rm = TRUE)) %>% \n",
    "    dplyr::filter(is.na(varcount) != TRUE) %>% \n",
    "    ungroup() %>% \n",
    "    arrange(desc(varcount))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93cbb3f6-0b72-48b1-9e99-8e41ee1c4f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to generate null distribution\n",
    "generate_null_distribution_v3 <- function(data, col_name) {\n",
    "  # Determine the number of rows per class (equal for all classes)\n",
    "  n_rows_per_class <- min(250)\n",
    "\n",
    "  ## Replicate splitting is old and being abandoned in place of bootstrapping\n",
    "  # Split the data into replicate1 and replicate2 groups based on the rep_values column\n",
    "  #data_rep1 <- data %>% filter(data[[rep_values]] == \"rep1\")\n",
    "  #data_rep2 <- data %>% filter(data[[rep_values]] == \"rep2\")\n",
    "\n",
    "  # Group data by the specified column and select n_rows_per_class for each class \n",
    "  sampled_data <- data %>%\n",
    "    group_by_at(col_name) %>%\n",
    "    sample_n(n_rows_per_class,replace = TRUE) %>%\n",
    "    ungroup()\n",
    "\n",
    "  # Shuffle the column and make sure less than 20% of cells retain their original cluster ID\n",
    "  #shuffled_column <- shuffle_and_assign_v3(sampled_data, col_name)\n",
    "  #sampled_data[[col_name]] <- shuffled_column\n",
    "   sampled_data[[col_name]] <- sampled_data[[col_name]][sample(nrow(sampled_data))]\n",
    "\n",
    "  return(sampled_data)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b7a749a-9b33-4b64-b069-5adcae739a31",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_to_sparse_matrix <- function(three_col_tribble, meta_slot_var) {\n",
    "    \n",
    "    CPM_matrix_prep <- three_col_tribble  %>% \n",
    "    dplyr::select(!!sym(meta_slot_var), geneID, grouped_CPM)\n",
    "\n",
    "\n",
    "\n",
    "    three_col_prep <- CPM_matrix_prep  %>% \n",
    "        dplyr::rename(\"V2\" = \"geneID\")  %>% \n",
    "        dplyr::rename(\"V1\" = !!sym(meta_slot_var))\n",
    "\n",
    "\n",
    "    # make sure bins/cells are factors\n",
    "    three_col_prep$V1 <- factor(three_col_prep$V1)\n",
    "    three_col_prep$V2 <- factor(three_col_prep$V2)\n",
    "\n",
    "\n",
    "    # convert to sparseMatrix format\n",
    "    sparse_count_matrix <- Matrix::sparseMatrix(i=as.numeric(three_col_prep$V1),\n",
    "        j=as.numeric(three_col_prep$V2),\n",
    "        x=as.numeric(three_col_prep$grouped_CPM),\n",
    "        dimnames=list(levels(three_col_prep$V1),levels(three_col_prep$V2)))\n",
    "\n",
    "\n",
    "    return(sparse_count_matrix)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22d957f4-c7fd-4bb9-bb0a-491de4d1e9f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to subset the matrix and sample all rows\n",
    "subset_and_sample <- function(matrix, replicate_list) {\n",
    "  # Find the columns matching the replicate list\n",
    "  selected_cols <- grep(paste(replicate_list, collapse = \"|\"), colnames(matrix))\n",
    "  \n",
    "  # Subset the matrix\n",
    "  subset_matrix <- matrix[, selected_cols]\n",
    "  \n",
    "  # Sample all rows\n",
    "  sampled_rows <- subset_matrix[sample(nrow(subset_matrix)), ]\n",
    "  \n",
    "  return(sampled_rows)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60fbc7bd-d498-4486-ad79-6057b912cc60",
   "metadata": {},
   "outputs": [],
   "source": [
    "null_CPM_gen <- function(meta_data, raw_cpm_counts_all_genes, meta_slot_var){\n",
    "#    meta_slot_var <- c(\"final_annotation\")\n",
    "    merged_meta_cpm_information <- left_join(meta_data, raw_cpm_counts_all_genes, by = c(\"cellID\"), relationship = \"many-to-many\")  %>%\n",
    "        group_by(!!sym(meta_slot_var), geneID)  %>%\n",
    "        summarise(counts = sum(accessability, na.rm = TRUE))\n",
    "\n",
    "    ### Alt CPM Calc\n",
    "    merged_meta_cpm_information_copied <- merged_meta_cpm_information\n",
    "    catch <- merged_meta_cpm_information_copied  %>%\n",
    "        group_by(!!sym(meta_slot_var)) %>%\n",
    "        group_map(~(cpm(.x$counts, log = FALSE, group = .f)), .keep = TRUE)  %>%\n",
    "        unlist()\n",
    "\n",
    "    caught_values <- as_tibble(catch)\n",
    "    see <- ungroup(merged_meta_cpm_information_copied)\n",
    "    merged_meta_cpm_information_copied<- bind_cols(merged_meta_cpm_information_copied,caught_values)  %>% \n",
    "        rename(grouped_CPM = value)\n",
    "    \n",
    "    return(merged_meta_cpm_information_copied)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f67080cd-a2d8-4bc7-b199-52b293aa33ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_null_dist_values <- function(meta_data, meta_slot_var, raw_cpm_counts_all_genes){\n",
    "#    meta_slot_var <- c(\"final_annotation\")\n",
    "    merged_meta_cpm_information <- left_join(meta_data, raw_cpm_counts_all_genes, by = c(\"cellID\"), relationship = \"many-to-many\")  %>%\n",
    "        group_by(!!sym(meta_slot_var), geneID)  %>%\n",
    "        summarise(counts = sum(accessability, na.rm = TRUE))\n",
    "\n",
    "    ### Alt CPM Calc\n",
    "    merged_meta_cpm_information_copied <- merged_meta_cpm_information\n",
    "    catch <- merged_meta_cpm_information_copied  %>%\n",
    "        group_by(!!sym(meta_slot_var)) %>%\n",
    "        group_map(~(cpm(.x$counts, log = FALSE, group = .f)), .keep = TRUE)  %>%\n",
    "        unlist()\n",
    "\n",
    "    caught_values <- as_tibble(catch)\n",
    "    see <- ungroup(merged_meta_cpm_information_copied)\n",
    "    merged_meta_cpm_information_copied<- bind_cols(merged_meta_cpm_information_copied,caught_values)  %>% \n",
    "        rename(grouped_CPM = value)\n",
    "    \n",
    "    #head(merged_meta_cpm_information_copied)\n",
    "    \n",
    "    sparse_null_dist <- convert_to_sparse_matrix(merged_meta_cpm_information_copied, meta_slot_var)\n",
    "    transposed_ACRs_by_ct <- as.matrix(t(sparse_null_dist))\n",
    "\n",
    "    message(\"Generating Null Distribution ...\")\n",
    "    #isolate_rep_values <- as.list(unique(meta_data[rep_slot_var]))\n",
    "    #print(isolate_rep_values)\n",
    "    \n",
    "    #rep_1_cpm <- subset_and_sample(null_cpm_values,isolate_rep_values[1][1])\n",
    "    calculate_specificity <- calculating_specificity(transposed_ACRs_by_ct)\n",
    "    \n",
    "    #message(\"Generating Null Distribution specificity for null rep2...\")\n",
    "    #rep_2_cpm <- subset_and_sample(null_cpm_values,isolate_rep_values[1][2])\n",
    "    #rep_2_specificity <- calculating_specificity(rep_2_cpm)\n",
    "\n",
    "    null_val_specificity <- calculate_specificity\n",
    "    \n",
    "    \n",
    "    return(null_val_specificity)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8911faa5-a932-42ae-b745-d9eff18cad39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to calculate the p-values for a single row\n",
    "calc_pvals <- function(qp, all_null_values_array) {\n",
    "  obs <- qp[is.finite(qp)]\n",
    "  ave <- mean(all_null_values_array, na.rm = TRUE)\n",
    "  sd <- sd(all_null_values_array, na.rm = TRUE)\n",
    "  pvals <- pnorm(obs, mean = ave, sd = 2*sd, lower.tail = TRUE)\n",
    "  return(pvals)\n",
    "}\n",
    "\n",
    "\n",
    "generate_specificity_pvals <- function(cell_type_specificty_scores,\n",
    "                                       all_null_values_array, \n",
    "                                       threads=5) {\n",
    "n_cores <- threads\n",
    "\n",
    "message(\"scheduling cores...\")\n",
    "# calculate the p-values for each row\n",
    "cl <- makeCluster(n_cores)\n",
    "clusterExport(cl, c(\"cell_type_specificty_scores\", \"all_null_values_array\", \"calc_pvals\"), envir=environment())\n",
    "pvals <- parLapply(cl, seq(nrow(cell_type_specificty_scores)), function(z) {\n",
    "  qp <- cell_type_specificty_scores[z,]\n",
    "  vals <- calc_pvals(qp)\n",
    "  return(vals)\n",
    "})\n",
    "stopCluster(cl)\n",
    "\n",
    "# convert the list to a matrix\n",
    "pvals <- do.call(rbind, pvals)\n",
    "\n",
    "colnames(pvals) <- colnames(cell_type_specificty_scores)\n",
    "rownames(pvals) <- rownames(cell_type_specificty_scores)\n",
    "\n",
    "return(pvals) \n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a90018b9-0e10-4cbc-8c69-d555c594c171",
   "metadata": {},
   "outputs": [],
   "source": [
    "null_distributions <- replicate(100, generate_null_distribution_v3(meta_data, \"final_annotation_n\"), simplify = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39d51ad2-3f9a-4a0b-b09b-322c651d117b",
   "metadata": {},
   "outputs": [],
   "source": [
    "look_group <- c(\"scACR_51864\", \"scACR_18990\", \"scACR_43190\", \"scACR_53777\")\n",
    "\n",
    "null_meta <- dplyr::bind_rows(null_distributions, .id = \"origin_df\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9ac302c-7422-437b-9020-8adfcefa5e01",
   "metadata": {},
   "outputs": [],
   "source": [
    "null_dist_CPMs <- lapply(null_distributions, null_CPM_gen, raw_cpm_counts_all_genes, \"final_annotation_n\")\n",
    "combined_CPMs_null <- dplyr::bind_rows(null_dist_CPMs, .id = \"origin_df\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6db5ff1-0ed1-4547-bcb3-91e1521aaf38",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(combined_CPMs_null)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f41f76-f20b-44b0-b0e6-7389cac3da81",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "options(repr.plot.width=10, repr.plot.height=10)\n",
    "combined_CPMs_null %>% \n",
    "    dplyr::ungroup() %>% \n",
    "    dplyr::filter(origin_df %in% c(1,2,3,4,5)) %>% \n",
    "ggplot(., aes(x = counts, y = grouped_CPM, color = final_annotation_n)) + \n",
    "geom_point() + facet_grid(origin_df~.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e46a3d9-94f5-4721-8605-f501bde4baff",
   "metadata": {},
   "outputs": [],
   "source": [
    "look_group <- c(\"scACR_51864\", \"scACR_18990\", \"scACR_43190\", \"scACR_53777\")\n",
    "\n",
    "combined_CPMs_null %>% \n",
    "    dplyr::filter(geneID %in% look_group) %>% \n",
    "    ggplot(., aes(x = counts)) + geom_density() + facet_grid(final_annotation_n~geneID)\n",
    "\n",
    "combined_CPMs_null %>% \n",
    "    dplyr::filter(geneID %in% look_group) %>% \n",
    "    ggplot(., aes(x = grouped_CPM)) + geom_density() + facet_grid(final_annotation_n~geneID)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "152ba6be-75ae-4011-bd5c-f8a93232b103",
   "metadata": {},
   "outputs": [],
   "source": [
    "null_dist_values <- lapply(null_distributions, generate_null_dist_values, \"final_annotation_n\", raw_cpm_counts_all_genes)\n",
    "#null_dist_values <- readRDS(\"/home/jpm73279/null_dist_gen.2023-05-17.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7af020a8-319d-4d1b-ba17-fa9a3c2b7bba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#saveRDS(null_dist_values, file = \"/home/jpm73279/null_dist_gen.2023-05-25.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c29762e2-0609-442a-abf9-62a8dfecd927",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Custom function to convert matrix to data frame and row names to column\n",
    "convert_mat_to_df <- function(mat) {\n",
    "  df <- as.data.frame(mat)\n",
    "  df <- tibble::rownames_to_column(df, \"row_name\")\n",
    "  return(df)\n",
    "}\n",
    "\n",
    "# Use lapply to apply the custom function to each matrix in the list\n",
    "dfs <- lapply(null_dist_values, convert_mat_to_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "657a7bea-c30a-473b-9e8b-809002749546",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_combined <- dplyr::bind_rows(dfs, .id = \"origin_df\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f809b86b-ce1a-4f7f-adb7-d464eeab3a9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "longer_null <- df_combined %>% \n",
    "    pivot_longer(bundle_sheath_n_cell_2878:protoderm_n_cell_2021, names_to = \"cell_type\", values_to = \"entropy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c817817-75ea-48eb-83ec-e67d17abceff",
   "metadata": {},
   "outputs": [],
   "source": [
    "look_group <- c(\"scACR_51864\", \"scACR_18990\", \"scACR_43190\", \"scACR_53777\")\n",
    "longer_null %>% \n",
    "    dplyr::filter(row_name %in% look_group) %>% \n",
    "    ggplot(., aes(x = entropy)) + geom_density() + facet_grid(cell_type~row_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02587f78-95ee-42e4-b1e9-ad3e0e8c7d51",
   "metadata": {},
   "outputs": [],
   "source": [
    "y <- bootstraps(meta_data, times = 500, strata = final_annotation)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d118e9cd-d36f-4891-a3d9-6595733036f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_pvalues_bootstraps <- function(meta_data, raw_cpm_counts_all_genes, meta_slot_var, null_values) {\n",
    "\n",
    "    merged_meta_cpm_information <- left_join(meta_data, raw_cpm_counts_all_genes, by = c(\"cellID\"))  %>%\n",
    "        group_by(!!sym(meta_slot_var), geneID)  %>%\n",
    "        summarise(counts = sum(accessability, na.rm = TRUE))\n",
    "\n",
    "    message(\"generating the CPM values\")\n",
    "    ### Alt CPM Calc\n",
    "    merged_meta_cpm_information_copied <- merged_meta_cpm_information\n",
    "    catch <- merged_meta_cpm_information_copied  %>%\n",
    "        group_by(!!sym(meta_slot_var)) %>%\n",
    "        group_map(~(cpm(.x$counts, log = FALSE, group = .f)), .keep = TRUE)  %>%\n",
    "        unlist()\n",
    "\n",
    "    caught_values <- as_tibble(catch)\n",
    "    see <- ungroup(merged_meta_cpm_information_copied)\n",
    "    merged_meta_cpm_information_copied<- bind_cols(merged_meta_cpm_information_copied,caught_values)  %>% \n",
    "        rename(grouped_CPM = value)\n",
    "    \n",
    "    #head(merged_meta_cpm_information_copied)\n",
    "    \n",
    "    sparse_null_dist <- convert_to_sparse_matrix(merged_meta_cpm_information_copied, meta_slot_var)\n",
    "    transposed_ACRs_by_ct <- as.matrix(t(sparse_null_dist))\n",
    "    \n",
    "    print(head(transposed_ACRs_by_ct))\n",
    "\n",
    "\n",
    "    #isolate_rep_values <- as.list(unique(meta_data[rep_slot_var]))\n",
    "    #print(isolate_rep_values)\n",
    "    \n",
    "    #rep_1_cpm <- subset_and_sample(null_cpm_values,isolate_rep_values[1][1])\n",
    "    message(\"Calculating Specificity ...\")\n",
    "    calculate_specificity <- calculating_specificity(transposed_ACRs_by_ct)\n",
    "    print(head(calculate_specificity))\n",
    "    \n",
    "    #message(\"Generating P values ...\")\n",
    "    #pvals <- generate_specificity_pvals_updated(calculate_specificity,\n",
    "    #                           null_values, \n",
    "    #                           threads = 5)\n",
    "    \n",
    "    return(calculate_specificity)\n",
    "\n",
    "\n",
    "}\n",
    "                                       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3a22513-c910-4789-a647-fed69a51f0ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(gplots) \n",
    "library(limma) \n",
    "library(edgeR) \n",
    "library(parallel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69fbd8bf-58b2-467c-b23b-bd6118df75b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(future.apply)\n",
    "\n",
    "generate_specificity_pvals_updated <- function(cell_type_specificty_scores,\n",
    "                                       all_null_values_array, \n",
    "                                       threads=5) {\n",
    "\n",
    "  plan(multiprocess, workers = threads)\n",
    "\n",
    "  # calculate the p-values for each row\n",
    "  pvals <- future_lapply(1:nrow(cell_type_specificty_scores), function(z) {\n",
    "    qp <- cell_type_specificty_scores[z,]\n",
    "    vals <- calc_pvals(qp)\n",
    "    return(vals)\n",
    "  })\n",
    "\n",
    "  # convert the list to a matrix\n",
    "  pvals <- do.call(rbind, pvals)\n",
    "\n",
    "  colnames(pvals) <- colnames(cell_type_specificty_scores)\n",
    "  rownames(pvals) <- rownames(cell_type_specificty_scores)\n",
    "\n",
    "  return(pvals)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de66d2e7-730d-4911-a3a5-6f0c07cf1810",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_specifivity <- generate_pvalues_bootstraps(meta_data, raw_cpm_counts_all_genes, \"final_annotation\", merged_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dad1280-4982-48f2-b03b-17d1017fb3e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(progressr)\n",
    "library(furrr)\n",
    "\n",
    "calc_pvals <- function(qp, mean_val, sd) {\n",
    "  obs <- qp[is.finite(qp)]\n",
    "  #ave <- mean(all_null_values_array, na.rm = TRUE)\n",
    "  #sd <- sd(all_null_values_array, na.rm = TRUE)\n",
    "  pvals <- pnorm(obs, mean = mean_val, sd = 2*sd, lower.tail = TRUE)\n",
    "  return(pvals)\n",
    "}\n",
    "generate_specificity_pvals_progress <- function(cell_type_specificty_scores,\n",
    "                                       mean_null,\n",
    "                                        sd_null,\n",
    "                                       threads=5) {\n",
    "  \n",
    "  plan(multisession, workers = threads)\n",
    "\n",
    "  # Set up progress reporting\n",
    "  #handlers(\"txtprogressbar\")\n",
    "  \n",
    "  message(\"scheduling cores...\")\n",
    "  \n",
    "  # calculate the p-values for each row\n",
    "  pvals <- future_map(seq(nrow(cell_type_specificty_scores)), function(z) {\n",
    "    qp <- cell_type_specificty_scores[z,]\n",
    "    vals <- calc_pvals(qp, mean_null, sd_null)\n",
    "    #p()\n",
    "    return(vals)\n",
    "  })\n",
    "    \n",
    "  # convert the list to a matrix\n",
    "  pvals <- do.call(rbind, pvals)\n",
    "\n",
    "  colnames(pvals) <- colnames(cell_type_specificty_scores)\n",
    "  rownames(pvals) <- rownames(cell_type_specificty_scores)\n",
    "\n",
    "  return(pvals)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "777c3198-38c0-41f9-8c71-0acb6d9cef58",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(future)\n",
    "library(furrr)\n",
    "\n",
    "generate_pvalues_bootstraps_fast <- function(meta_data, raw_cpm_counts_all_genes, meta_slot_var, null_mean, null_sd) {\n",
    "  \n",
    "  setDT(meta_data)\n",
    "  setDT(raw_cpm_counts_all_genes)\n",
    "  \n",
    "  merged_meta_cpm_information <- merge(meta_data, raw_cpm_counts_all_genes, by = \"cellID\", all.x = TRUE)\n",
    "  \n",
    "  merged_meta_cpm_information[, counts := sum(accessability, na.rm = TRUE), by = c(meta_slot_var, \"geneID\")]\n",
    "  \n",
    "  message(\"generating the CPM values\")\n",
    "  \n",
    "  merged_meta_cpm_information[, grouped_CPM := cpm(counts, log = FALSE, group = get(meta_slot_var)), by = meta_slot_var]\n",
    "  \n",
    "  sparse_null_dist <- convert_to_sparse_matrix(merged_meta_cpm_information, meta_slot_var)\n",
    "  transposed_ACRs_by_ct <- as.matrix(t(sparse_null_dist))\n",
    "  \n",
    "  message(\"Generating P values ...\")\n",
    "  \n",
    "  calculate_specificity <- calculating_specificity(transposed_ACRs_by_ct)\n",
    "  # 4/30/2023\n",
    "  # Replacing this approach with a KS test after generation of distributions using Bootstraps \n",
    "  #pvals <- generate_specificity_pvals_progress(calculate_specificity,\n",
    "  #                                    null_mean,\n",
    "  #                                    null_sd,\n",
    "  #                                    threads = 5)\n",
    "  \n",
    "  return(calculate_specificity)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ce29c8e-f859-4862-a4f8-6e747d2ffc2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up parallel processing\n",
    "plan(multiprocess)\n",
    "\n",
    "results <- y %>%\n",
    "  mutate(p_values = map(splits, ~ generate_pvalues_bootstraps_fast(analysis(.x), raw_cpm_counts_all_genes, \"final_annotation\", mean_null, sd_null))) %>%\n",
    "  select(id, p_values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e87158b7-f3d7-4dae-83e2-515fe911053d",
   "metadata": {},
   "outputs": [],
   "source": [
    "expaneded_bootstraps <- results %>%\n",
    "  mutate(p_values = map(p_values, ~ as_tibble(.x, rownames = \"ACR_values\"))) %>% # Convert matrices to tibbles and include row names\n",
    "  unnest(cols = p_values) %>%                          # Unnest the tibbles\n",
    "  pivot_longer(cols = -c(id, ACR_values),              # Keep 'id' and 'ACR_values' fixed\n",
    "               names_to = \"cell_type\",                 # Assign the column names to 'cell_type'\n",
    "               values_to = \"value\") %>%                # Assign the values to 'value'\n",
    "  rename(BootstrapID = id)                             # Rename the columns to the desired names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba867b7c-dcd7-4fa6-ada0-4f34cc9438a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "nested_data <- expaneded_bootstraps %>% \n",
    "    select(-BootstrapID) %>%\n",
    "    group_by(ACR_values, cell_type) %>%\n",
    "    nest() %>% \n",
    "    rename(distribution = \"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77078b38-df53-4fbd-8589-8354972640b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(nested_data$ACR_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9716e0c6-36b1-4822-a4bd-6101d54f1da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "other <- nested_data %>%\n",
    "    mutate(var_measure = sd(unlist(distribution), na.rm = TRUE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf599aad-251b-4286-8127-3b9a2a63e496",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(other, aes(var_measure)) + geom_histogram()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6562a9c5-bf35-456c-9993-b869dad1b8a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=10)\n",
    "ggplot(other, aes(var_measure)) + geom_histogram() + facet_wrap(cell_type~.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f7f3f7f-d00f-4b3a-a718-2a108458c528",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b505b847-b2d3-488a-9397-9c92487b2cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "other %>% \n",
    "    group_by(cell_type) %>% \n",
    "    summarise(mean(var_measure),\n",
    "             min(var_measure),\n",
    "             max(var_measure))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af96f92f-9636-4235-9328-a5a81888fd42",
   "metadata": {},
   "outputs": [],
   "source": [
    "df <- imap_dfr(null_dist_values, function(matrix, index) {\n",
    "  df_matrix <- reshape2::melt(matrix)\n",
    "  df_matrix <- df_matrix %>%\n",
    "    rename(\n",
    "      row_name = Var1,\n",
    "      column_name = Var2,\n",
    "      value = value\n",
    "    ) %>%\n",
    "    mutate(matrix_index = index)\n",
    "  return(df_matrix)\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7ed4d70-befa-4f72-8552-793f67c9789a",
   "metadata": {},
   "outputs": [],
   "source": [
    "null_dist_generation <- df %>% \n",
    "    rename(ACR_values = row_name) %>%\n",
    "    select(ACR_values, value) %>%\n",
    "    group_by(ACR_values) %>%\n",
    "    nest() %>%\n",
    "    rename(null_dist = data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0e17617-fe0b-430a-a4d0-cbec5b7240a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "x <- left_join(nested_data, null_dist_generation, by = c(\"ACR_values\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f78eb2f-83f3-4ad8-88e9-df9cbc1400b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#medianBootstrap(unlist(x[1,3]),unlist(x[1,4]))\n",
    "\n",
    "#y <- x[1:100,] %>% \n",
    "#    rowwise() %>%\n",
    "#    mutate(p_ci_values = pmap(list(distribution, null_dist), medianBootstrap)) %>% \n",
    "#    unnest_wider(p_ci_values, names_sep = \"_\")\n",
    "\n",
    "#y %>% \n",
    "#    dplyr::filter(p_ci_values_upper_ci < .005)%>% \n",
    "#    dplyr::select(ACR_values, cell_type, p_ci_values_1:p_ci_values_upper_ci)\n",
    "\n",
    "#y %>% \n",
    "#    ungroup() %>% \n",
    "#    dplyr::filter(ACR_values == \"scACR_1004\") %>% \n",
    "#    unnest(distribution) %>% \n",
    "#    rename(real_value = value) %>% \n",
    "#    unnest(null_dist)%>% \n",
    "#    rename(null_value = value) %>% \n",
    "#    pivot_longer(c(real_value,null_value), names_to = \"class\", values_to = \"val\") %>% \n",
    "#    ggplot(., aes(val, color = class)) + geom_density() + facet_grid(cell_type~.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dae61de-8164-4c6f-8d0f-7ecc02118add",
   "metadata": {},
   "outputs": [],
   "source": [
    " n <- x %>%\n",
    "    ungroup() %>%\n",
    "    rowwise() %>% \n",
    "    mutate(median_val = mean(unlist(distribution), na.rm = TRUE)) %>%\n",
    "    mutate(median_null= mean(unlist(null_dist), na.rm = TRUE)) %>%\n",
    "    mutate(pos_test = (sum(median_val > unlist(null_dist)))) %>% \n",
    "    mutate(list_len = lengths(null_dist)) %>%\n",
    "    mutate(pval = pos_test/list_len) %>% \n",
    "    ungroup() %>% \n",
    "    #dplyr::filter(pval < .01)%>% \n",
    "    dplyr::select(ACR_values, cell_type, median_val,median_null, pval) %>% \n",
    "    arrange(median_val, desc = TRUE )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f338540b-928f-486a-97c9-8f69af0931e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Generate the same plot looking at ACRs associated with marker genes \n",
    "options(repr.plot.width=15, repr.plot.height=15)\n",
    "#PEPC1 ME3 GL1 LRD3\n",
    "look_group <- c(\"scACR_51864\", \"scACR_18990\", \"scACR_43190\", \"scACR_53777\")\n",
    "plot_acr_null_real <- x %>% \n",
    "    ungroup() %>% \n",
    "    dplyr::filter(ACR_values %in% look_group) %>% \n",
    "    mutate(ACR_values = factor(ACR_values, levels = look_group)) %>%  # Reorder the levels of ACR_values\n",
    "    unnest(distribution) %>% \n",
    "    rename(real_value = value) %>% \n",
    "    unnest(null_dist)%>% \n",
    "    rename(null_value = value) %>% \n",
    "    pivot_longer(c(real_value,null_value), names_to = \"class\", values_to = \"val\") \n",
    "\n",
    "acr_meds <- plot_acr_null_real %>% \n",
    "    group_by(class, cell_type, ACR_values) %>% \n",
    "    summarise(median_val = mean(val))\n",
    " \n",
    "ggplot(plot_acr_null_real, aes(val, color = class)) + geom_density() + facet_grid(cell_type~ACR_values, scales=\"free_y\") +\n",
    "      geom_vline(data=acr_meds, aes(xintercept=median_val, color=class),\n",
    "             linetype=\"dashed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9372142f-e4b6-442f-b11c-76b7ab1b5940",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Generate the same plot looking at ACRs associated with marker genes \n",
    "options(repr.plot.width=15, repr.plot.height=15)\n",
    "# GOOD x3 Bad X3 \n",
    "look_group <- c(\"scACR_10800\",\n",
    "\"scACR_1159\",\n",
    "\"scACR_10096\",\n",
    "\"scACR_100\",\n",
    "\"scACR_11711\",\n",
    "\"scACR_10059\")\n",
    "\n",
    "\n",
    "plot_acr_null_real <- x %>% \n",
    "    ungroup() %>% \n",
    "    dplyr::filter(ACR_values %in% look_group) %>% \n",
    "    mutate(ACR_values = factor(ACR_values, levels = look_group)) %>%  # Reorder the levels of ACR_values\n",
    "    unnest(distribution) %>% \n",
    "    rename(real_value = value) %>% \n",
    "    unnest(null_dist)%>% \n",
    "    rename(null_value = value) %>% \n",
    "    pivot_longer(c(real_value,null_value), names_to = \"class\", values_to = \"val\") \n",
    "\n",
    "acr_meds <- plot_acr_null_real %>% \n",
    "    group_by(class, cell_type, ACR_values) %>% \n",
    "    summarise(median_val = mean(val))\n",
    " \n",
    "ggplot(plot_acr_null_real, aes(val, color = class)) + geom_density() + facet_grid(cell_type~ACR_values, scales=\"free_y\") +\n",
    "      geom_vline(data=acr_meds, aes(xintercept=median_val, color=class),\n",
    "             linetype=\"dashed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fc06838-e296-420a-9277-a1384f06c0eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_specfic <- x %>% \n",
    "    dplyr::filter(cell_type == \"mesophyll\" | cell_type == \"epidermis\") %>% \n",
    "    unnest(distribution) %>% \n",
    "    rename(real_value = value) %>%  \n",
    "    dplyr::group_by(ACR_values, cell_type) %>% \n",
    "    dplyr::summarise(mean_acr = mean(real_value, na.rm = TRUE),\n",
    "                    median_acr = median(real_value, na.rm = TRUE),\n",
    "                    var_acr = var(real_value, na.rm = TRUE)) %>% \n",
    "    dplyr::arrange(mean_acr, desc = TRUE) %>% \n",
    "    ungroup() %>% \n",
    "    dplyr::group_by(cell_type) %>% \n",
    "    dplyr::slice(1:3)\n",
    "\n",
    "\n",
    "options(repr.plot.width=15, repr.plot.height=15)\n",
    "look_group <- cell_type_specfic$ACR_values\n",
    "plot_variable_ACRs <- x %>% \n",
    "    ungroup() %>% \n",
    "    dplyr::filter(ACR_values %in% look_group) %>% \n",
    "    mutate(ACR_values = factor(ACR_values, levels = look_group)) %>%  # Reorder the levels of ACR_values\n",
    "    unnest(distribution) %>% \n",
    "    rename(real_value = value) %>% \n",
    "    unnest(null_dist)%>% \n",
    "    rename(null_value = value) %>% \n",
    "    pivot_longer(c(real_value,null_value), names_to = \"class\", values_to = \"val\") \n",
    "\n",
    "acr_meds <- plot_variable_ACRs %>% \n",
    "    group_by(class, cell_type, ACR_values) %>% \n",
    "    summarise(median_val = mean(val))\n",
    " \n",
    "ggplot(plot_variable_ACRs, aes(val, color = class)) + geom_density() + facet_grid(cell_type~ACR_values, scales=\"free_y\") +\n",
    "      geom_vline(data=acr_meds, aes(xintercept=median_val, color=class),\n",
    "             linetype=\"dashed\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73c03f07-30d1-4bd5-9f7c-06a61da5b8bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_pvals <- function(qp, mean_val, sd) {\n",
    "  obs <- qp[is.finite(qp)]\n",
    "  #ave <- mean(all_null_values_array, na.rm = TRUE)\n",
    "  #sd <- sd(all_null_values_array, na.rm = TRUE)\n",
    "  pvals <- pnorm(obs, mean = mean_val, sd = sd, lower.tail = TRUE)\n",
    "  return(pvals)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b0a7e21-d918-434d-b371-681dd06c8233",
   "metadata": {},
   "outputs": [],
   "source": [
    "pnorm_test <- x %>%\n",
    "    ungroup() %>%\n",
    "    rowwise() %>% \n",
    "    mutate(median_val = mean(unlist(distribution), na.rm = TRUE)) %>%\n",
    "    mutate(median_null= mean(unlist(null_dist), na.rm = TRUE)) %>%\n",
    "    mutate(pos_test = map(null_dist, ~calc_pvals(median_val, mean(unlist(.x), na.rm = TRUE), sd = sd(unlist(.x), na.rm = TRUE)))) %>% \n",
    "    mutate(pvalue = unlist(pos_test)) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5803379-eaa6-40f0-b6e5-eca74623f4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(pnorm_test, aes(x = pvalue)) + geom_histogram(binwidth = .005)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66dc49b0-a8eb-4ebb-bbb6-d773b917d3c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "passing <- pnorm_test %>% \n",
    "    dplyr::filter(pvalue < .01) %>% \n",
    "    dplyr::arrange(desc(pvalue)) %>% \n",
    "    dplyr::select(ACR_values, cell_type, median_val, median_null, pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "338134e6-23df-461f-886b-482944452f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "glimpse(pnorm_test)\n",
    "length(unique(passing$ACR_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53edc29a-2789-4ae6-80af-e8cd245fe200",
   "metadata": {},
   "outputs": [],
   "source": [
    "passing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e12c6eb-5654-43bf-bfa8-c4bf09d20fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Generate the same plot looking at ACRs associated with marker genes \n",
    "options(repr.plot.width=15, repr.plot.height=15)\n",
    "#Bad looking on browser - edge of Sig\n",
    "look_group <- c(\"scACR_22742\", \"scACR_31849\", \"scACR_8008\", \"scACR_32242\")\n",
    "plot_acr_null_real <- x %>% \n",
    "    ungroup() %>% \n",
    "    dplyr::filter(ACR_values %in% look_group) %>% \n",
    "    mutate(ACR_values = factor(ACR_values, levels = look_group)) %>%  # Reorder the levels of ACR_values\n",
    "    unnest(distribution) %>% \n",
    "    rename(real_value = value) %>% \n",
    "    unnest(null_dist)%>% \n",
    "    rename(null_value = value) %>% \n",
    "    pivot_longer(c(real_value,null_value), names_to = \"class\", values_to = \"val\") \n",
    "\n",
    "acr_meds <- plot_acr_null_real %>% \n",
    "    group_by(class, cell_type, ACR_values) %>% \n",
    "    summarise(median_val = mean(val))\n",
    " \n",
    "ggplot(plot_acr_null_real, aes(val, color = class)) + geom_density() + facet_grid(cell_type~ACR_values, scales=\"free_y\") +\n",
    "      geom_vline(data=acr_meds, aes(xintercept=median_val, color=class),\n",
    "             linetype=\"dashed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b0d6fab-eaab-4d73-ad6c-d1cc5763c3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "less_than_pval <- x %>%\n",
    "    ungroup() %>%\n",
    "    rowwise() %>% \n",
    "    mutate(median_val = mean(unlist(distribution), na.rm = TRUE)) %>%\n",
    "    mutate(median_null= mean(unlist(null_dist), na.rm = TRUE)) %>% \n",
    "    mutate(pos_test = (sum(unlist(distribution) > unlist(null_dist)))) %>% \n",
    "    mutate(list_len = lengths(null_dist)) %>%\n",
    "    mutate(pval = pos_test/list_len) %>% \n",
    "    ungroup() %>% \n",
    "    #dplyr::filter(pval < .001)%>% \n",
    "    dplyr::select(ACR_values, cell_type, median_val,median_null, pval) %>% \n",
    "    arrange(pval, desc = TRUE )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "538d683b-2f1b-4a8b-8e1f-32907bc9c31c",
   "metadata": {},
   "outputs": [],
   "source": [
    "passing_less <- less_than_pval %>% \n",
    "    dplyr::filter(pval < .01) %>% \n",
    "    dplyr::arrange(desc(pval)) %>% \n",
    "    dplyr::select(ACR_values, cell_type, median_val, median_null, pval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eff1faf8-0fe3-4082-b9c6-83c5aae05158",
   "metadata": {},
   "outputs": [],
   "source": [
    "glimpse(passing_less)\n",
    "length(unique(passing$ACR_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c44684e-573f-4a84-a31e-823f29cbc5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(passing_less)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b954de-1ebd-40a0-b302-7f42b7fbcba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Generate the same plot looking at ACRs associated with marker genes \n",
    "options(repr.plot.width=15, repr.plot.height=15)\n",
    "#Bad looking on browser - edge of Sig\n",
    "look_group <- c(\"scACR_10040\", \"scACR_10050\", \"scACR_10096\", \"scACR_10578\")\n",
    "plot_acr_null_real <- x %>% \n",
    "    ungroup() %>% \n",
    "    dplyr::filter(ACR_values %in% look_group) %>% \n",
    "    mutate(ACR_values = factor(ACR_values, levels = look_group)) %>%  # Reorder the levels of ACR_values\n",
    "    unnest(distribution) %>% \n",
    "    rename(real_value = value) %>% \n",
    "    unnest(null_dist)%>% \n",
    "    rename(null_value = value) %>% \n",
    "    pivot_longer(c(real_value,null_value), names_to = \"class\", values_to = \"val\") \n",
    "\n",
    "acr_meds <- plot_acr_null_real %>% \n",
    "    group_by(class, cell_type, ACR_values) %>% \n",
    "    summarise(median_val = mean(val))\n",
    " \n",
    "ggplot(plot_acr_null_real, aes(val, color = class)) + geom_density() + facet_grid(cell_type~ACR_values, scales=\"free_y\") +\n",
    "      geom_vline(data=acr_meds, aes(xintercept=median_val, color=class),\n",
    "             linetype=\"dashed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
