{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "81a65ca8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "here() starts at /home/jpm73279/Mendieta_et_al_comparative_single_cell\n",
      "\n",
      "Loading required package: usethis\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: Rcpp\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.2 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.6      \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.5 \n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.8      \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.10\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.1      \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.1 \n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.3      \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.2 \n",
      "Warning message:\n",
      "“package ‘ggplot2’ was built under R version 4.2.1”\n",
      "Warning message:\n",
      "“package ‘tibble’ was built under R version 4.2.1”\n",
      "Warning message:\n",
      "“package ‘stringr’ was built under R version 4.2.1”\n",
      "── \u001b[1mConflicts\u001b[22m ───────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mSocrates\u001b[39m\n",
      "Loading required package: cicero\n",
      "\n",
      "Loading required package: monocle\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: 'Matrix'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:tidyr':\n",
      "\n",
      "    expand, pack, unpack\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: VGAM\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: splines\n",
      "\n",
      "Loading required package: DDRTree\n",
      "\n",
      "Loading required package: irlba\n",
      "\n",
      "Loading required package: Gviz\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Matrix':\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: FNN\n",
      "\n",
      "Loading required package: doSNOW\n",
      "\n",
      "Loading required package: foreach\n",
      "\n",
      "\n",
      "Attaching package: 'foreach'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:purrr':\n",
      "\n",
      "    accumulate, when\n",
      "\n",
      "\n",
      "Loading required package: iterators\n",
      "\n",
      "Loading required package: snow\n",
      "\n",
      "Loading required package: itertools\n",
      "\n",
      "Loading required package: glmnet\n",
      "\n",
      "Loaded glmnet 4.1-6\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: 'matrixStats'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:itertools':\n",
      "\n",
      "    product\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biobase':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'DelayedArray'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    aperm, apply, rowsum, scale, sweep\n",
      "\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "\n",
      "Attaching package: 'AnnotationDbi'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "\n",
      "Attaching package: 'MASS'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Loading required package: viridis\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Warning message:\n",
      "\"package 'viridisLite' was built under R version 4.2.1\"\n",
      "Loading required package: uwot\n",
      "\n",
      "Loading required package: scales\n",
      "\n",
      "Warning message:\n",
      "\"package 'scales' was built under R version 4.2.1\"\n",
      "\n",
      "Attaching package: 'scales'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:viridis':\n",
      "\n",
      "    viridis_pal\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    discard\n",
      "\n",
      "\n",
      "The following object is masked from 'package:readr':\n",
      "\n",
      "    col_factor\n",
      "\n",
      "\n",
      "Loading required package: RColorBrewer\n",
      "\n",
      "Warning message:\n",
      "\"package 'RColorBrewer' was built under R version 4.2.1\"\n",
      "Loading required package: nabor\n",
      "\n",
      "\n",
      "Attaching package: 'nabor'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:FNN':\n",
      "\n",
      "    knn\n",
      "\n",
      "\n",
      "Loading required package: CMF\n",
      "\n",
      "Loading required package: qlcMatrix\n",
      "\n",
      "Loading required package: slam\n",
      "\n",
      "Loading required package: sparsesvd\n",
      "\n",
      "\n",
      "Attaching package: 'qlcMatrix'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biobase':\n",
      "\n",
      "    rowMax, rowMin\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    none\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"replacing previous import 'scales::viridis_pal' by 'viridis::viridis_pal' when loading 'Socrates'\"\n"
     ]
    }
   ],
   "source": [
    "library(\"here\")\n",
    "library(devtools)\n",
    "library(Seurat)\n",
    "library(harmony)\n",
    "library(RcppML)\n",
    "library(reshape2)\n",
    "library(tidyverse)\n",
    "load_all('/home/jpm73279/Socrates')\n",
    "\n",
    "\n",
    "# arguments\n",
    "#args <- commandArgs(TRUE)\n",
    "#tn5_bed <- as.character(args[1])\n",
    "#peak_bed <- as.character(args[1])\n",
    "#metadata <- as.character(args[3])\n",
    "#ann <- as.character(args[4])\n",
    "#chr <- as.character(args[4])\n",
    "#prefix <- as.character(args[5])\n",
    "\n",
    "# arguments\n",
    "args <- commandArgs(TRUE)\n",
    "#tn5_bed  <- as.character(args[1])\n",
    "#peak_bed  <- as.character(args[2])\n",
    "#metadata  <- as.character(args[3])\n",
    "#ann <-  as.character(args[4])\n",
    "#chr <-  as.character(args[5])\n",
    "#prefix <- as.character(args[6])\n",
    "\n",
    "\n",
    "tn5_bed <- \"/scratch/jpm73279/comparative_single_cell/01.alignments_annotations/sorghum_bicolor/Sorghum_leaf.sciATAC_rep2.unique.tn5.merged.V2.sorted.bed\"\n",
    "peak_bed <- \"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/Sb_peak_calls/Sb.v4.final/Sb.peaks.annot_v4.500bp_peaks.bed\"\n",
    "metadata <- \"/home/jpm73279/Mendieta_et_al_comparative_single_cell/metrics/annotations/sb_annot_v4/Sb.leaf_annot.V4.meta.final.2022-11-14.txt\"\n",
    "ann <- \"/home/jpm73279/genome_downloads/Sbicolor.v5.1.prelim.annot/Sbicolorv5.1.primaryTrs.gff3\"\n",
    "chr <- \"/home/jpm73279/genome_downloads/Sbicolor.v5.1.prelim.annot/Sorghum_bicolor_var_BTx623.chrom.size\"\n",
    "prefix <- \"Sb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "83d92bda",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running pre-check on input files and executable paths ...\n",
      "\n",
      "BED file path = /scratch/jpm73279/comparative_single_cell/01.alignments_annotations/sorghum_bicolor/Sorghum_leaf.sciATAC_rep2.unique.tn5.merged.V2.sorted.bed ... ok\n",
      "\n",
      "GFF file path = /home/jpm73279/genome_downloads/Sbicolor.v5.1.prelim.annot/Sbicolorv5.1.primaryTrs.gff3 ... ok\n",
      "\n",
      "Chromosome sizes file path = /home/jpm73279/genome_downloads/Sbicolor.v5.1.prelim.annot/Sorghum_bicolor_var_BTx623.chrom.size ... ok\n",
      "\n",
      "Macs2 is installed .... ok\n",
      "\n",
      " - loading data (this may take obj while for big BED files) ...\n",
      "\n",
      " - finished loading data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Load Object\n",
    "Zm.subcluster <- loadBEDandGenomeData(tn5_bed, ann, chr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3ed25350",
   "metadata": {},
   "outputs": [],
   "source": [
    "Zm.subcluster$meta <- read.table(metadata)\n",
    "Zm.subcluster$acr <- read.table(peak_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e473dc57",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################################################################\n",
    "###################################################################################################\n",
    "###################################################################################################\n",
    "#' generateMatrix\n",
    "#'\n",
    "#' This function generates the sparse matrix from equally sized genomic bins or ACRs.\n",
    "#'\n",
    "#' @importFrom GenomicRanges GRanges\n",
    "#' @importFrom GenomicRanges tileGenome\n",
    "#' @importFrom IRanges subsetByOverlaps\n",
    "#'\n",
    "#' @param obj Object output from findCells or buildMetaData. Required.\n",
    "#' @param filtered Logical. Whether or not to use the filtered set of cells. Defaults to TRUE.\n",
    "#' @param windows Integer. Window size to build bins. If the 'peaks' parameter is set to TRUE,\n",
    "#' this argument is over-ridden.\n",
    "#' @param peaks Logical. If TRUE, use ACRs to build sparse matrix instead of genomic tiles.\n",
    "#' Default is set to FALSE.\n",
    "#' @param blacklist in bed format. If given removes black list regions from either peaks or \n",
    "#' generated bins. Default is set to null.\n",
    "#' Default is set to FALSE.\n",
    "#' @param verbose Logical. Whether or not to print progress.\n",
    "#'\n",
    "#' @rdname generateMatrix\n",
    "#' @export\n",
    "#'\n",
    "generateMatrix_sparse <- function(obj,\n",
    "                           filtered=T,\n",
    "                           windows=1000,\n",
    "                           peaks=FALSE,\n",
    "                           blacklist=NULL,\n",
    "                           organelle_scaffolds = NULL,\n",
    "                           verbose=T){\n",
    "\n",
    "\n",
    "    # convert tn5 bed to Granges\n",
    "    tn5.gr <- GRanges(seqnames=as.character(obj$bed$V1),\n",
    "                      ranges=IRanges(start=as.numeric(obj$bed$V2),\n",
    "                                     end=as.numeric(obj$bed$V3)),\n",
    "                      strand=as.character(obj$bed$V5),\n",
    "                      names=as.character(obj$bed$V4))\n",
    "    \n",
    "    \n",
    "    # Remove Organell Scaffolds if given\n",
    "    if(is.null(organelle_scaffolds) == FALSE) {\n",
    "        \n",
    "        tn5.gr <- dropSeqlevels(tn5.gr, organelle_scaffolds)\n",
    "\n",
    "\n",
    "    } else {\n",
    "        tn5.gr <- tn5.gr\n",
    "    }\n",
    "\n",
    "\n",
    "    \n",
    "    # Read in baclist if given\n",
    "    if(is.null(blacklist) == FALSE) {\n",
    "        blacklist_r <- read.table(as.character(blacklist))\n",
    "\n",
    "\n",
    "        blacklist.gr <- GRanges(seqnames=as.character(blacklist_r$V1),\n",
    "          \n",
    "                ranges=IRanges(start=as.numeric(blacklist_r$V2),\n",
    "                                         end=as.numeric(blacklist_r$V3)),\n",
    "                          names=as.character(blacklist_r$V4))\n",
    "    } else {\n",
    "        blacklist.gr <- NULL\n",
    "    }\n",
    "\n",
    "\n",
    "    # use filtered barcodes?\n",
    "    if(filtered){\n",
    "        use <- obj$meta.v3\n",
    "    }else{\n",
    "        use <- obj$meta\n",
    "    }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # generate intervals\n",
    "    if(!peaks){\n",
    "        # build bins from specified tile length\n",
    "        chr.seq.lengths <- as.numeric(obj$chr$V2)\n",
    "        names(chr.seq.lengths) <- obj$chr$V1\n",
    "        intervals <- tileGenome(chr.seq.lengths, tilewidth=windows, cut.last.tile.in.chrom=TRUE)\n",
    "\n",
    "\n",
    "        #Remove if black list included\n",
    "        #Remove procedure learned from: https://www.biostars.org/p/263214/\n",
    "       if (is.null(blacklist.gr) == FALSE){\n",
    "\n",
    "\n",
    "            intervals <- intervals[-queryHits(findOverlaps(intervals, blacklist.gr, type=\"any\")),] \n",
    "            regions <- as.data.frame(intervals)\n",
    "            regions <- paste(regions$seqnames, regions$start, regions$end, sep=\"_\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        }else{\n",
    "            regions <- as.data.frame(intervals)\n",
    "            regions <- paste(regions$seqnames, regions$start, regions$end, sep=\"_\")\n",
    "        }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    }else{\n",
    "\n",
    "\n",
    "        # generate intervals from ACRs\n",
    "        intervals <- GRanges(seqnames=as.character(obj$acr$V1),\n",
    "                             ranges=IRanges(start=as.numeric(obj$acr$V2),\n",
    "                                            end=as.numeric(obj$acr$V3)))\n",
    "\n",
    "\n",
    "        if (is.null(blacklist.gr) == FALSE){\n",
    "            intervals <- intervals[-queryHits(findOverlaps(intervals, blacklist.gr, type=\"any\")),] \n",
    "            regions <- as.data.frame(intervals)\n",
    "            regions <- paste(regions$seqnames, regions$start, regions$end, sep=\"_\")\n",
    "        }else{\n",
    "            regions <- as.data.frame(intervals)\n",
    "            regions <- paste(regions$seqnames, regions$start, regions$end, sep=\"_\")\n",
    "        }\n",
    "    }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # get intervals overlapping Tn5 sites by barcode\n",
    "    hits <- as.data.frame(findOverlaps(tn5.gr, intervals))\n",
    "    df <- data.frame(regions=regions[hits$subjectHits], barcodes=as.character(obj$bed$V4)[hits$queryHits])\n",
    "    df <- df[!duplicated(df),]\n",
    "    df$binary <- 1\n",
    "    colnames(df) <- c(\"V1\",\"V2\",\"V3\")\n",
    "    \n",
    "    \n",
    "    #9/26/2022 include for sake of calculation of isCell \n",
    "    # make sure nSites is calculated\n",
    "    #Integration Sites\n",
    "    a <- df\n",
    "    \n",
    "    #Meta data to interset\n",
    "    b <- use\n",
    "    a$V1 <- factor(a$V1)\n",
    "    a$V2 <- factor(a$V2)\n",
    "\n",
    "\n",
    "    #Generate sparse matrix\n",
    "    a <- Matrix::sparseMatrix(i=as.numeric(a$V1),\n",
    "                              j=as.numeric(a$V2),\n",
    "                              x=as.numeric(a$V3),\n",
    "                              dimnames=list(levels(a$V1),levels(a$V2)))\n",
    "\n",
    "\n",
    "    # align barcodes\n",
    "    both <- intersect(rownames(b), colnames(a))\n",
    "    a <- a[,both]\n",
    "    b <- b[both,]\n",
    "\n",
    "\n",
    "    # make sure nSites is calculated\n",
    "    b$nSites   <- Matrix::colSums(a)\n",
    "    b$log10nSites <- log10(b$nSites)\n",
    "\n",
    "\n",
    "    # return\n",
    "    obj$counts <- df\n",
    "    obj$meta <- b \n",
    "\n",
    "\n",
    "    return(obj) \n",
    "    \n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8f3cf6f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "Zm.subcluster <- generateMatrix_sparse(Zm.subcluster, \n",
    "                                            filtered=FALSE,\n",
    "                                            peaks = TRUE,\n",
    "                                            verbose=FALSE)\n",
    "\n",
    "#Zm.subcluster.vasculature <- convertSparseData(Zm.subcluster.vasculature, verbose = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b1e5ef42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 9\n",
      " $ bed    :'data.frame':\t162461909 obs. of  5 variables:\n",
      "  ..$ V1: chr [1:162461909] \"Chr01\" \"Chr01\" \"Chr01\" \"Chr01\" ...\n",
      "  ..$ V2: int [1:162461909] 149 168 168 168 168 168 168 175 175 176 ...\n",
      "  ..$ V3: int [1:162461909] 150 169 169 169 169 169 169 176 176 177 ...\n",
      "  ..$ V4: chr [1:162461909] \"BC:Z:CAGATCAGTCAATGTCGGTAGAGACTA-Sorghum_leaf.sciATAC_rep2\" \"BC:Z:CTTGTAAGTTCCTGTTTGTCCGCTCGG-Sorghum_leaf.sciATAC_rep2\" \"BC:Z:CTTGTAATGTCAACTCGGGTAGCAGCT-Sorghum_leaf.sciATAC_rep2\" \"BC:Z:GATCAGCTTGTATGTCGATGAGCTCGG-Sorghum_leaf.sciATAC_rep2\" ...\n",
      "  ..$ V5: chr [1:162461909] \"+\" \"+\" \"+\" \"+\" ...\n",
      " $ gff    :Reference class 'TxDb' [package \"GenomicFeatures\"] with 6 fields\n",
      "  ..$ conn           :Formal class 'SQLiteConnection' [package \"RSQLite\"] with 8 slots\n",
      "  .. .. ..@ ptr                :<externalptr> \n",
      "  .. .. ..@ dbname             : chr \"\"\n",
      "  .. .. ..@ loadable.extensions: logi TRUE\n",
      "  .. .. ..@ flags              : int 70\n",
      "  .. .. ..@ vfs                : chr \"\"\n",
      "  .. .. ..@ ref                :<environment: 0x55982d1ec9a8> \n",
      "  .. .. ..@ bigint             : chr \"integer64\"\n",
      "  .. .. ..@ extended_types     : logi FALSE\n",
      "  ..$ packageName    : chr(0) \n",
      "  ..$ user2seqlevels0: int [1:47] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  ..$ user_seqlevels : chr [1:47] \"Chr01\" \"Chr02\" \"Chr03\" \"Chr04\" ...\n",
      "  ..$ user_genome    : chr [1:47] NA NA NA NA ...\n",
      "  ..$ isActiveSeq    : logi [1:47] TRUE TRUE TRUE TRUE TRUE TRUE ...\n",
      "  ..and 16 methods, of which 2 are  possibly relevant:\n",
      "  ..  finalize, initialize\n",
      " $ chr    :'data.frame':\t67 obs. of  2 variables:\n",
      "  ..$ V1: chr [1:67] \"Chr01\" \"Chr02\" \"Chr03\" \"Chr04\" ...\n",
      "  ..$ V2: int [1:67] 85112863 79114963 80873341 71215609 77058072 62713908 68911884 65779274 63277606 62870657 ...\n",
      " $ bedpath: chr \"/scratch/jpm73279/comparative_single_cell/01.alignments_annotations/sorghum_bicolor/Sorghum_leaf.sciATAC_rep2.u\"| __truncated__\n",
      " $ annpath: chr \"/home/jpm73279/genome_downloads/Sbicolor.v5.1.prelim.annot/Sbicolorv5.1.primaryTrs.gff3\"\n",
      " $ chrpath: chr \"/home/jpm73279/genome_downloads/Sbicolor.v5.1.prelim.annot/Sorghum_bicolor_var_BTx623.chrom.size\"\n",
      " $ meta   :'data.frame':\t0 obs. of  38 variables:\n",
      "  ..$ V1         : chr(0) \n",
      "  ..$ V2         : chr(0) \n",
      "  ..$ V3         : chr(0) \n",
      "  ..$ V4         : chr(0) \n",
      "  ..$ V5         : chr(0) \n",
      "  ..$ V6         : chr(0) \n",
      "  ..$ V7         : chr(0) \n",
      "  ..$ V8         : chr(0) \n",
      "  ..$ V9         : chr(0) \n",
      "  ..$ V10        : chr(0) \n",
      "  ..$ V11        : chr(0) \n",
      "  ..$ V12        : chr(0) \n",
      "  ..$ V13        : chr(0) \n",
      "  ..$ V14        : chr(0) \n",
      "  ..$ V15        : chr(0) \n",
      "  ..$ V16        : chr(0) \n",
      "  ..$ V17        : chr(0) \n",
      "  ..$ V18        : chr(0) \n",
      "  ..$ V19        : chr(0) \n",
      "  ..$ V20        : chr(0) \n",
      "  ..$ V21        : chr(0) \n",
      "  ..$ V22        : chr(0) \n",
      "  ..$ V23        : chr(0) \n",
      "  ..$ V24        : chr(0) \n",
      "  ..$ V25        : chr(0) \n",
      "  ..$ V26        : chr(0) \n",
      "  ..$ V27        : chr(0) \n",
      "  ..$ V28        : chr(0) \n",
      "  ..$ V29        : chr(0) \n",
      "  ..$ V30        : chr(0) \n",
      "  ..$ V31        : chr(0) \n",
      "  ..$ V32        : chr(0) \n",
      "  ..$ V33        : chr(0) \n",
      "  ..$ V34        : chr(0) \n",
      "  ..$ V35        : chr(0) \n",
      "  ..$ V36        : chr(0) \n",
      "  ..$ nSites     : num(0) \n",
      "  ..$ log10nSites: num(0) \n",
      " $ acr    :'data.frame':\t67338 obs. of  5 variables:\n",
      "  ..$ V1: chr [1:67338] \"Chr01\" \"Chr01\" \"Chr01\" \"Chr01\" ...\n",
      "  ..$ V2: int [1:67338] 22172 35037 41678 57764 68443 70763 83504 89684 99571 100569 ...\n",
      "  ..$ V3: int [1:67338] 22673 35538 42179 58265 68944 71264 84005 90185 100072 101070 ...\n",
      "  ..$ V4: chr [1:67338] \"scACR_1\" \"scACR_2\" \"scACR_3\" \"scACR_4\" ...\n",
      "  ..$ V5: num [1:67338] 1.86 40.54 5.2 39.17 5.77 ...\n",
      " $ counts :'data.frame':\t43963726 obs. of  3 variables:\n",
      "  ..$ V1: chr [1:43963726] \"Chr01_22172_22673\" \"Chr01_22172_22673\" \"Chr01_22172_22673\" \"Chr01_22172_22673\" ...\n",
      "  ..$ V2: chr [1:43963726] \"BC:Z:GGCTACAGTCAAGTAGTGAGTGGACTA-Sorghum_leaf.sciATAC_rep2\" \"BC:Z:GATCAGATGTCACCGAGGTAGAGAGCT-Sorghum_leaf.sciATAC_rep2\" \"BC:Z:GCCAATGGCTACGGCAGGTCCGCGATG-Sorghum_leaf.sciATAC_rep1\" \"BC:Z:CTTGTACTTGTAAAACAGTTTCGTGAC-Sorghum_leaf.sciATAC_rep2\" ...\n",
      "  ..$ V3: num [1:43963726] 1 1 1 1 1 1 1 1 1 1 ...\n"
     ]
    }
   ],
   "source": [
    "str(Zm.subcluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "580027f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "    sparse_count_matrix <- Zm.subcluster$counts\n",
    "    # make sure bins/cells are factors\n",
    "    sparse_count_matrix$V1 <- factor(sparse_count_matrix$V1)\n",
    "    sparse_count_matrix$V2 <- factor(sparse_count_matrix$V2)\n",
    "\n",
    "\n",
    "\n",
    "    # convert to sparseMatrix format\n",
    "    sparse_count_matrix <- Matrix::sparseMatrix(i=as.numeric(sparse_count_matrix$V1),\n",
    "                              j=as.numeric(sparse_count_matrix$V2),\n",
    "                              x=as.numeric(sparse_count_matrix$V3),\n",
    "                             dimnames=list(levels(sparse_count_matrix$V1),levels(sparse_count_matrix$V2)))\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "749a6789",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "saveRDS(sparse_count_matrix, file = paste0(\"Sb.test2\",\".peaks_by_intersections.rds\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
