{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e7a1d2f-24f5-4f33-a8b5-803524acd647",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.2          \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0.\u001b[31m9000\u001b[39m     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4          \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2          \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2          \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Warning message:\n",
      "“package ‘devtools’ was built under R version 4.3.1”\n",
      "Loading required package: usethis\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mBSgenome.Zm_B73\u001b[39m\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘XVector’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mBSgenome.Os\u001b[39m\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mBSgenome.Pm\u001b[39m\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mBSgenome.Sb\u001b[39m\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mBSgenome.Uf\u001b[39m\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:tidyr’:\n",
      "\n",
      "    expand, pack, unpack\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: Rsamtools\n",
      "\n",
      "\n",
      "Attaching package: ‘GenomicAlignments’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    last\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###################################################################################################\n",
    "###                             Motif Enrichment                                ###\n",
    "###################################################################################################\n",
    "\n",
    "#load libraries\n",
    "library(tidyverse)\n",
    "library(devtools)\n",
    "library(motifmatchr)\n",
    "library(BiocParallel)\n",
    "load_all('/home/jpm73279/genome_downloads/BS_genomes/BSgenome.Zm_B73')\n",
    "load_all(\"/home/jpm73279/genome_downloads/BS_genomes/BSgenome.Os\")\n",
    "load_all(\"/home/jpm73279/genome_downloads/BS_genomes/BSgenome.Pm\")\n",
    "load_all(\"/home/jpm73279/genome_downloads/BS_genomes/BSgenome.Sb\")\n",
    "load_all(\"/home/jpm73279/genome_downloads/BS_genomes/BSgenome.Uf\")\n",
    "\n",
    "library(Matrix)\n",
    "library(GenomicAlignments)\n",
    "library(dplyr)\n",
    "library(universalmotif)  # manipulating motif representations\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7636d04c-d2e6-48df-99ec-1dcf132a14c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m30\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m52\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m33\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m31\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m33\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "## Load Markers for All Species\n",
    "marker_file_paths <- c(\"/home/jpm73279/genome_downloads/C4_markers/Os.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Pm.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Sb.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Uf.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Zm.c4_markers.bed\")\n",
    "\n",
    "# Function to read and process the file\n",
    "process_file <- function(file_path) {\n",
    "  species <- substr(basename(file_path), 1, 2)\n",
    "  read_delim(file_path, delim = \"\\t\", col_names = c(\"chrom\", \"start\", \"end\", \"geneID\", \"name\", \"type\")) %>% \n",
    "    dplyr::select(\"chrom\",\"start\",\"end\",\"geneID\",\"name\",\"type\") %>%\n",
    "    dplyr::mutate(species = (species))\n",
    "}\n",
    "\n",
    "# Apply the function to each file and store results in a list\n",
    "list_of_dataframes <- lapply(marker_file_paths, process_file)\n",
    "\n",
    "# Optionally combine all dataframes into one if needed\n",
    "markers <- bind_rows(list_of_dataframes) %>% \n",
    "    dplyr::select(geneID,name,type,species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fd077dc0-42e0-4566-b3d2-3d028c01c8ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘TFBSTools’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Matrix’:\n",
      "\n",
      "    Matrix\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Load Motifs for Analysis\n",
    "core_motifs <- read_meme(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/00.data/conserved_consensus_motifs_Fig2B.meme\")\n",
    "de_novo_motifs <- read_meme(\"/home/jpm73279/genome_downloads/C4_markers/found_motifs/combined_unique_motifs.meme\")\n",
    "ms_de_novo_motifs <- read_meme(\"/home/jpm73279/genome_downloads/C4_markers/found_motifs/Ms.specific_CGTCGT_motif.comparative.meme\")\n",
    "\n",
    "all_assy_motifs <- c(core_motifs, de_novo_motifs, ms_de_novo_motifs)\n",
    "all_assy_motifs_converted <- convert_motifs(all_assy_motifs, class = \"TFBSTools-PWMatrix\")\n",
    "\n",
    "update_name <- function(pwm) {\n",
    "    if (!is.null(pwm@name)) {\n",
    "        pwm@ID <- pwm@name\n",
    "    }\n",
    "    return(pwm)\n",
    "}\n",
    "\n",
    "\n",
    "### Update the Names to be the actual IDs, as well as convert to PWM matrix List (Nasty as fuck)\n",
    "library(TFBSTools)\n",
    "pw_matrix_list_converted <- lapply(all_assy_motifs_converted, update_name)\n",
    "pw_matrix_list_converted <- lapply(pw_matrix_list_converted, PWMatrixList)\n",
    "pw_matrix_list_converted <- do.call(c, pw_matrix_list_converted)\n",
    "\n",
    "\n",
    "#Generate an Index Annotation name for later usage. Will be used because we need to rip these to annotate the Granges objects\n",
    "#later\n",
    "extractIDDataFrame <- function(pwMatrixList) {\n",
    "  ids <- sapply(pwMatrixList@listData, function(item) item@ID)\n",
    "  ids <- gsub(\"chr\", \"\", ids, fixed = TRUE)\n",
    "  indices <- seq_along(ids)\n",
    "  \n",
    "  data.frame(group = indices, TF_ID = ids)\n",
    "}\n",
    "\n",
    "# Example usage:\n",
    "# Assuming your PWMatrixList object is named pwMatrixListObject\n",
    "motif_id_index <- extractIDDataFrame(pw_matrix_list_converted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "95bed933-9813-49f5-afb4-054d3883a512",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m51387\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): chr, acr_id, score\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m65990\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): chr, acr_id, score\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m55504\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): chr, acr_id, score\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m29153\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): chr, acr_id, score\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m62669\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): chr, acr_id, score\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "sb_acrs <- read_delim(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/copy_over/sb_acr_classification.no_exons.all_ACRs.classified.sorted.bed\", col_names = c(\"chr\", \"start\", \"end\", \"acr_id\", \"score\"))\n",
    "pm_acrs <- read_delim(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/copy_over/pm_acr_classification.no_exons.all_ACRs.classified.sorted.bed\", col_names = c(\"chr\", \"start\", \"end\", \"acr_id\", \"score\"))\n",
    "zm_acrs <- read_delim(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/copy_over/zm_acr_classification.no_exons.all_ACRs.classified.sorted.bed\", col_names = c(\"chr\", \"start\", \"end\", \"acr_id\", \"score\"))\n",
    "uf_acrs <- read_delim(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/copy_over/uf_acr_classification.no_exons.all_ACRs.classified.sorted.bed\", col_names = c(\"chr\", \"start\", \"end\", \"acr_id\", \"score\"))\n",
    "os_acrs <- read_delim(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/copy_over/os_acr_classification.no_exons.all_ACRs.classified.sorted.bed\", col_names = c(\"chr\", \"start\", \"end\", \"acr_id\", \"score\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a119920a-018d-493f-bbd9-2a836c0f19dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bc07d07f-f5b2-41ab-9eab-54e42f0d153f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m146\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): chr, id, score\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "header_bed <- c('chr','start','end','id','score')\n",
    "zm_acr_intersections <- read_delim(\"/home/jpm73279/genome_downloads/C4_markers/conservation/zm.all_intersection_acr.intersection.bed\", \n",
    "            col_names = header_bed) %>% \n",
    "    separate(id, into = c(\"acr\", \"acr_cell_type_specific_class\", \"locus\"), sep = \";\") %>%\n",
    "    dplyr::mutate(acr_cell_type_specific_class = case_when(acr_cell_type_specific_class == \"bundle_sheath,procambial_meristem\" ~ \"bundle_sheath\",\n",
    "                                                          acr_cell_type_specific_class == \"bundle_sheath,procambium\" ~ \"bundle_sheath\", \n",
    "                                                          TRUE ~ acr_cell_type_specific_class)) %>% \n",
    "    tidyr::separate(locus, into = c(\"species_other\", \"loci\"), sep = \"__\",remove = FALSE) %>%\n",
    "    tidyr::separate(loci, into = c(\"gene_family\", \"number\"), sep = \"_\", remove = FALSE)\n",
    "\n",
    "combined_acrs_marker_info <- left_join(zm_acr_intersections, markers, by = c(\"locus\" = \"name\")) %>% \n",
    "    select(chr:acr_cell_type_specific_class, locus, gene_family, type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dc05be82-6a3c-4e4e-a92d-aa44696023ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "set.seed(420)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "748033ed-b1ce-487b-b7f9-6ce16f2aa04f",
   "metadata": {},
   "outputs": [],
   "source": [
    "processPeaks <- function(peaks, genome, motif_IDs) {\n",
    "    # Convert peaks to GRanges object\n",
    "    peaks_gr <- GRanges(peaks, seqinfo=seqinfo(genome))\n",
    "    names(peaks_gr) <- as.character(peaks_gr)\n",
    "    peaks_gr <- unique(peaks_gr)\n",
    "\n",
    "    # Get peak sequences\n",
    "    peak_seqs <- getSeq(genome, peaks_gr)\n",
    "\n",
    "    # Exclude peaks with any N's\n",
    "    hasNs <- (alphabetFrequency(peak_seqs)[,\"N\"] > 0)\n",
    "    peak_seqs <- peak_seqs[!hasNs]\n",
    "\n",
    "    # Run matchMotifs\n",
    "    system.time(\n",
    "        motif_matches <- matchMotifs(\n",
    "            pw_matrix_list_converted, \n",
    "            peaks_gr, \n",
    "            genome = genome, \n",
    "            p.cutoff = .0005, # using a stringent p-value\n",
    "            out=\"positions\"\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Annotate each GRanges object in the list with the corresponding TF ID\n",
    "    granges_list <- motif_matches\n",
    "    for (i in seq_along(granges_list)) {\n",
    "        granges_list[[i]]$TF_ID <- motif_IDs[i]\n",
    "        \n",
    "        overlaps <- findOverlaps(granges_list[[i]], peaks_gr)\n",
    "        ids <- peaks_gr$acr_id[subjectHits(overlaps)]\n",
    "        granges_list[[i]]$acr_id <- ids[queryHits(overlaps)]\n",
    "    }\n",
    "\n",
    "    return(granges_list)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04ca3493-c791-4c20-9248-1943478a00be",
   "metadata": {},
   "source": [
    "## Run Zea mays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ade3f2c4-76fb-40b0-bc71-85ecb58ba786",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hasNs\n",
       "FALSE   Sum \n",
       "55504 55504 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       "  5.359   0.149   5.525 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "zm_peaks_gr <- GRanges(zm_acrs, seqinfo=seqinfo(BSgenome.Zm_B73))\n",
    "names(zm_peaks_gr) <- as.character(zm_peaks_gr)\n",
    "zm_peaks_gr <- unique(zm_peaks_gr)\n",
    "\n",
    "# get peak sequences, we'll need this later\n",
    "zm_peak_seqs <- getSeq(BSgenome.Zm_B73, zm_peaks_gr) # <2s\n",
    "\n",
    "\n",
    "# exclude peaks with any N's, to avoid warnings further down\n",
    "hasNs    <- (alphabetFrequency(zm_peak_seqs)[,\"N\"]>0)\n",
    "addmargins(table(hasNs)) # => only drops 5 sequences\n",
    "\n",
    "zm_peak_seqs <- zm_peak_seqs[!hasNs]\n",
    "zm_peak_seqs  <- zm_peak_seqs[!hasNs]\n",
    "\n",
    "\n",
    "# run: 30k sequences, 100 motifs\n",
    "system.time(\n",
    "  motif_matches_zm <- matchMotifs(\n",
    "    pw_matrix_list_converted, \n",
    "    zm_peaks_gr, \n",
    "      genome = BSgenome.Zm_B73, \n",
    "    p.cutoff = .001, # using a stringent p-value\n",
    "    out=\"positions\"\n",
    "))  # <10s on my laptop\n",
    "\n",
    "motif_IDs <- motif_id_index$TF_ID\n",
    "zm_granges_list <- motif_matches_zm\n",
    "# Annotate each GRanges object in the list with the corresponding TF ID\n",
    "for (i in seq_along(zm_granges_list)) {\n",
    "    zm_granges_list[[i]]$TF_ID <- motif_IDs[i]\n",
    "    \n",
    "    overlaps <- findOverlaps(zm_granges_list[[i]] , zm_peaks_gr)\n",
    "    ids <- zm_peaks_gr$acr_id[subjectHits(overlaps)]\n",
    "    zm_granges_list[[i]]$acr_id <- ids[queryHits(overlaps)]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "1b827b30-cd96-4c0c-a5ef-33abfd64fd27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRangesList object of length 28:\n",
       "[[1]]\n",
       "GRanges object with 38201 ranges and 3 metadata columns:\n",
       "          seqnames              ranges strand |     score       TF_ID\n",
       "             <Rle>           <IRanges>  <Rle> | <numeric> <character>\n",
       "      [1]     chr1         34510-34515      + |   9.11767         ARF\n",
       "      [2]     chr1       165528-165533      - |   8.87641         ARF\n",
       "      [3]     chr1       258456-258461      + |   8.87641         ARF\n",
       "      [4]     chr1       328049-328054      - |   9.11767         ARF\n",
       "      [5]     chr1       328333-328338      - |   9.84689         ARF\n",
       "      ...      ...                 ...    ... .       ...         ...\n",
       "  [38197]     chr9 162072206-162072211      - |   9.84689         ARF\n",
       "  [38198]     chr9 162104262-162104267      + |   9.84689         ARF\n",
       "  [38199]     chr9 162232979-162232984      + |   9.11767         ARF\n",
       "  [38200]     chr9 162278311-162278316      + |   9.11767         ARF\n",
       "  [38201]     chr9 162499008-162499013      + |   8.87641         ARF\n",
       "                          acr_id\n",
       "                     <character>\n",
       "      [1] scACR_1;broadly_acce..\n",
       "      [2] scACR_5;broadly_acce..\n",
       "      [3] scACR_10;broadly_acc..\n",
       "      [4] scACR_11;broadly_acc..\n",
       "      [5] scACR_11;broadly_acc..\n",
       "      ...                    ...\n",
       "  [38197] scACR_56035;broadly_..\n",
       "  [38198] scACR_56036;broadly_..\n",
       "  [38199] scACR_56037;broadly_..\n",
       "  [38200] scACR_56040;broadly_..\n",
       "  [38201] scACR_56044;broadly_..\n",
       "  -------\n",
       "  seqinfo: 12 sequences from an unspecified genome; no seqlengths\n",
       "\n",
       "...\n",
       "<27 more elements>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "zm_granges_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "f31c91be-839c-4b95-aee7-223a4d224d57",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in left_join(., combined_acrs_marker_info_subset, by = c(\"acr\", :\n",
      "“\u001b[1m\u001b[22mDetected an unexpected many-to-many relationship between `x` and `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 17007 of `x` matches multiple rows in `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 5 of `y` matches multiple rows in `x`.\n",
      "\u001b[36mℹ\u001b[39m If a many-to-many relationship is expected, set `relationship =\n",
      "  \"many-to-many\"` to silence this warning.”\n"
     ]
    }
   ],
   "source": [
    "zm_acrs_TFs <- as_tibble(as.data.frame(zm_granges_list)) %>% \n",
    "    tidyr::separate(acr_id, into = c(\"acr\", \"acr_cell_type_specific_class\"), sep = \";\") %>%\n",
    "    dplyr::mutate(acr_cell_type_specific_class = case_when(acr_cell_type_specific_class == \"bundle_sheath,procambial_meristem\" ~ \"bundle_sheath\",\n",
    "                                                          acr_cell_type_specific_class == \"bundle_sheath,procambium\" ~ \"bundle_sheath\", \n",
    "                                                          TRUE ~ acr_cell_type_specific_class)) %>% \n",
    "    left_join(., combined_acrs_marker_info_subset, by = c(\"acr\", \"acr_cell_type_specific_class\")) %>% \n",
    "    mutate(locus = case_when(is.na(locus) == TRUE ~ \"None\",\n",
    "                            TRUE ~ locus),\n",
    "           gene_family = case_when(is.na(gene_family) == TRUE ~ \"None\",\n",
    "                            TRUE ~ locus),\n",
    "           type = case_when(is.na(type) == TRUE ~ \"None\",\n",
    "                            TRUE ~ type),\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21ed3222-ea47-44c0-b9c2-53f8231a6184",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_of_interest <- zm_acrs_TFs %>% \n",
    "    filter(type != \"None\")\n",
    "\n",
    "\n",
    "all_other_regions <- zm_acrs_TFs %>% \n",
    "    filter(type == \"None\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a26706b6-5ff6-4da6-bd0e-75e7c319568d",
   "metadata": {},
   "outputs": [],
   "source": [
    "GetGC <- function(bsgenome, gr){\n",
    "\n",
    "  seqs <- BSgenome::getSeq(bsgenome, gr)\n",
    "  return(as.numeric(Biostrings::letterFrequency(x = seqs, letters = \"GC\", as.prob = TRUE)))\n",
    "\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "7718d6d9-918d-4781-968b-7587d1cb25bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        17                 27\n",
      "  None           1670              29923\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        24                 39\n",
      "  None           2019              57950\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        51                124\n",
      "  None           6475             145427\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        27                 43\n",
      "  None           3185              57034\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll         8                 38\n",
      "  None           1826              50541\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        28                 87\n",
      "  None           2522              71300\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        15                 29\n",
      "  None           1408              42786\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        19                 29\n",
      "  None           1779              38018\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        31                 77\n",
      "  None           2015              78942\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        13                 47\n",
      "  None           1757              40336\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        14                 35\n",
      "  None           1429              34860\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        11                 42\n",
      "  None           1600              41726\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        18                 66\n",
      "  None           1984              54692\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        16                 66\n",
      "  None           2143              67641\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        21                 74\n",
      "  None           2514              70005\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        12                 88\n",
      "  None           2385              72775\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        30                 34\n",
      "  None           1948              45221\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        21                 71\n",
      "  None           2702              72439\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        24                 60\n",
      "  None           2105              43883\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        22                 27\n",
      "  None           2235              34689\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        17                 33\n",
      "  None           1293              34268\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        32                 86\n",
      "  None           2535              72622\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        16                 51\n",
      "  None           4947              73180\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        34                 43\n",
      "  None           3344              78710\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        25                 64\n",
      "  None           4186              93612\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        29                 58\n",
      "  None           3481              54394\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        32                 54\n",
      "  None           3075              57275\n",
      "           acr_cell_type_specific_class\n",
      "type        mesophyll broadly_accessible\n",
      "  mesophyll        64                140\n",
      "  None           7257             187717\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       "  0.854   0.041   0.898 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "ora_res <- list()\n",
    "system.time({\n",
    "    for (motif_id in unique(zm_acrs_TFs$TF_ID)) {\n",
    "        subsample <- zm_acrs_TFs %>% \n",
    "            dplyr::filter(TF_ID == motif_id) # Filter for each motif_id\n",
    "\n",
    "        xt <- xtabs(~ type + acr_cell_type_specific_class, data = subsample)\n",
    "\n",
    "        # Replace zeros with ones in the contingency table\n",
    "        xt <- apply(xt, c(1, 2), function(x) ifelse(x == 0, 1, x))\n",
    "        \n",
    "        print(xt[c(\"mesophyll\", \"None\"), c(\"mesophyll\", \"broadly_accessible\")])\n",
    "        ft.ms <- fisher.test(xt[c(\"mesophyll\", \"None\"), c(\"mesophyll\", \"broadly_accessible\")], alternative = \"greater\")\n",
    "        ft.bs <- fisher.test(xt[c(\"bundle_sheath\", \"None\"), c(\"bundle_sheath\", \"broadly_accessible\")], alternative = \"greater\")\n",
    "        ft.bs_vs_ms <- fisher.test(xt[c(\"bundle_sheath\", \"mesophyll\"), c(\"bundle_sheath\", \"mesophyll\")], alternative = \"greater\")\n",
    "        ft.ms_vs_bs <- fisher.test(xt[c(\"mesophyll\", \"bundle_sheath\"), c(\"mesophyll\", \"bundle_sheath\")], alternative = \"greater\")\n",
    "\n",
    "        ora_res[[motif_id]] <- data.frame(\n",
    "            \"Motif.ID\" = motif_id,\n",
    "            \"OR.ms\" = unname(ft.ms$estimate), \"pv.ms\" = ft.ms$p.value,\n",
    "            \"OR.bs\" = unname(ft.bs$estimate), \"pv.bs\" = ft.bs$p.value,\n",
    "            \"OR.bs_vs_ms\" = unname(ft.bs_vs_ms$estimate), \"pv.bs_vs_ms\" = ft.bs_vs_ms$p.value,\n",
    "            \"OR.ms_vs_bs\" = unname(ft.ms_vs_bs$estimate), \"pv.ms_vs_bs\" = ft.ms_vs_bs$p.value\n",
    "        )\n",
    "    }\n",
    "})\n",
    "\n",
    "\n",
    "ora_res <- as.data.frame(do.call(\"rbind\",ora_res))\n",
    "rownames(ora_res) <- NULL\n",
    "\n",
    "# adjust p-values for multiple testing\n",
    "ora_res$fdr.bs <- p.adjust(ora_res$pv.bs, method=\"BH\")\n",
    "ora_res$fdr.ms   <- p.adjust(ora_res$pv.ms, method=\"BH\") \n",
    "ora_res$fdr.bs_vs_ms   <- p.adjust(ora_res$pv.bs_vs_ms, method=\"BH\") \n",
    "ora_res$fdr.ms_vs_bs   <- p.adjust(ora_res$pv.ms_vs_bs, method=\"BH\") \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e16559c-defe-4b4b-849a-6f54cf3e718b",
   "metadata": {},
   "source": [
    "## Mona Lisa Method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "fbc46f99-3f9b-409f-8006-077841723a0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A spec_tbl_df: 55504 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>acr_id</th><th scope=col>score</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>  34471</td><td>  34972</td><td>scACR_1;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>  45954</td><td>  46455</td><td>scACR_2;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 123085</td><td> 123586</td><td>scACR_3;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 162544</td><td> 163045</td><td>scACR_4;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 165164</td><td> 165665</td><td>scACR_5;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 189283</td><td> 189784</td><td>scACR_6;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 199880</td><td> 200381</td><td>scACR_7;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 206262</td><td> 206763</td><td>scACR_8;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 252910</td><td> 253411</td><td>scACR_9;broadly_accessible                </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 258277</td><td> 258778</td><td>scACR_10;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 327853</td><td> 328354</td><td>scACR_11;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 328932</td><td> 329433</td><td>scACR_12;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 359989</td><td> 360490</td><td>scACR_13;bundle_sheath,procambial_meristem</td><td>0.0283323889203693,0.00636645445151828</td></tr>\n",
       "\t<tr><td>chr1</td><td> 362377</td><td> 362878</td><td>scACR_14;bundle_sheath                    </td><td>0.0057018279389569                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 457390</td><td> 457891</td><td>scACR_15;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 494213</td><td> 494714</td><td>scACR_16;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 564531</td><td> 565032</td><td>scACR_18;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 678118</td><td> 678619</td><td>scACR_19;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 680747</td><td> 681248</td><td>scACR_20;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td> 707771</td><td> 708272</td><td>scACR_21;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1002473</td><td>1002974</td><td>scACR_22;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1013849</td><td>1014350</td><td>scACR_23;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1014416</td><td>1014917</td><td>scACR_24;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1017973</td><td>1018474</td><td>scACR_25;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1022248</td><td>1022749</td><td>scACR_26;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1026915</td><td>1027416</td><td>scACR_27;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1027798</td><td>1028299</td><td>scACR_28;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1056836</td><td>1057337</td><td>scACR_29;bundle_sheath,procambial_meristem</td><td>0.0106996433452218,0.0043665211159628 </td></tr>\n",
       "\t<tr><td>chr1</td><td>1058017</td><td>1058518</td><td>scACR_30;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1063713</td><td>1064214</td><td>scACR_31;broadly_accessible               </td><td>NA                                    </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chr9</td><td>161870287</td><td>161870788</td><td>scACR_56019;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161874600</td><td>161875101</td><td>scACR_56020;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161892377</td><td>161892878</td><td>scACR_56021;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161893303</td><td>161893804</td><td>scACR_56022;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161904777</td><td>161905278</td><td>scACR_56024;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161956001</td><td>161956502</td><td>scACR_56025;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161967292</td><td>161967793</td><td>scACR_56026;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161971498</td><td>161971999</td><td>scACR_56027;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>161972275</td><td>161972776</td><td>scACR_56028;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162045573</td><td>162046074</td><td>scACR_56029;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162046375</td><td>162046876</td><td>scACR_56030;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162047423</td><td>162047924</td><td>scACR_56031;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162050722</td><td>162051223</td><td>scACR_56032;mesophyll         </td><td>0.0129778435262078 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162065379</td><td>162065880</td><td>scACR_56033;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162071240</td><td>162071741</td><td>scACR_56034;protoderm         </td><td>0.00666644445185161</td></tr>\n",
       "\t<tr><td>chr9</td><td>162071854</td><td>162072355</td><td>scACR_56035;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162103825</td><td>162104326</td><td>scACR_56036;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162232613</td><td>162233114</td><td>scACR_56037;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162236054</td><td>162236555</td><td>scACR_56038;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162237028</td><td>162237529</td><td>scACR_56039;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162278072</td><td>162278573</td><td>scACR_56040;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162280414</td><td>162280915</td><td>scACR_56041;mesophyll         </td><td>0.0267657744741842 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162355173</td><td>162355674</td><td>scACR_56042;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162362602</td><td>162363103</td><td>scACR_56043;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162498789</td><td>162499290</td><td>scACR_56044;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162575634</td><td>162576135</td><td>scACR_56045;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162759285</td><td>162759786</td><td>scACR_56046;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162766738</td><td>162767239</td><td>scACR_56047;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>162874029</td><td>162874530</td><td>scACR_56048;broadly_accessible</td><td>NA                 </td></tr>\n",
       "\t<tr><td>chr9</td><td>163004080</td><td>163004581</td><td>scACR_56049;broadly_accessible</td><td>NA                 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A spec\\_tbl\\_df: 55504 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " chr & start & end & acr\\_id & score\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t chr1 &   34471 &   34972 & scACR\\_1;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &   45954 &   46455 & scACR\\_2;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  123085 &  123586 & scACR\\_3;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  162544 &  163045 & scACR\\_4;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  165164 &  165665 & scACR\\_5;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  189283 &  189784 & scACR\\_6;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  199880 &  200381 & scACR\\_7;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  206262 &  206763 & scACR\\_8;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  252910 &  253411 & scACR\\_9;broadly\\_accessible                 & NA                                    \\\\\n",
       "\t chr1 &  258277 &  258778 & scACR\\_10;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  327853 &  328354 & scACR\\_11;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  328932 &  329433 & scACR\\_12;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  359989 &  360490 & scACR\\_13;bundle\\_sheath,procambial\\_meristem & 0.0283323889203693,0.00636645445151828\\\\\n",
       "\t chr1 &  362377 &  362878 & scACR\\_14;bundle\\_sheath                     & 0.0057018279389569                    \\\\\n",
       "\t chr1 &  457390 &  457891 & scACR\\_15;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  494213 &  494714 & scACR\\_16;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  564531 &  565032 & scACR\\_18;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  678118 &  678619 & scACR\\_19;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  680747 &  681248 & scACR\\_20;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 &  707771 &  708272 & scACR\\_21;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1002473 & 1002974 & scACR\\_22;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1013849 & 1014350 & scACR\\_23;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1014416 & 1014917 & scACR\\_24;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1017973 & 1018474 & scACR\\_25;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1022248 & 1022749 & scACR\\_26;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1026915 & 1027416 & scACR\\_27;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1027798 & 1028299 & scACR\\_28;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1056836 & 1057337 & scACR\\_29;bundle\\_sheath,procambial\\_meristem & 0.0106996433452218,0.0043665211159628 \\\\\n",
       "\t chr1 & 1058017 & 1058518 & scACR\\_30;broadly\\_accessible                & NA                                    \\\\\n",
       "\t chr1 & 1063713 & 1064214 & scACR\\_31;broadly\\_accessible                & NA                                    \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chr9 & 161870287 & 161870788 & scACR\\_56019;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161874600 & 161875101 & scACR\\_56020;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161892377 & 161892878 & scACR\\_56021;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161893303 & 161893804 & scACR\\_56022;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161904777 & 161905278 & scACR\\_56024;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161956001 & 161956502 & scACR\\_56025;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161967292 & 161967793 & scACR\\_56026;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161971498 & 161971999 & scACR\\_56027;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 161972275 & 161972776 & scACR\\_56028;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162045573 & 162046074 & scACR\\_56029;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162046375 & 162046876 & scACR\\_56030;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162047423 & 162047924 & scACR\\_56031;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162050722 & 162051223 & scACR\\_56032;mesophyll          & 0.0129778435262078 \\\\\n",
       "\t chr9 & 162065379 & 162065880 & scACR\\_56033;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162071240 & 162071741 & scACR\\_56034;protoderm          & 0.00666644445185161\\\\\n",
       "\t chr9 & 162071854 & 162072355 & scACR\\_56035;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162103825 & 162104326 & scACR\\_56036;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162232613 & 162233114 & scACR\\_56037;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162236054 & 162236555 & scACR\\_56038;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162237028 & 162237529 & scACR\\_56039;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162278072 & 162278573 & scACR\\_56040;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162280414 & 162280915 & scACR\\_56041;mesophyll          & 0.0267657744741842 \\\\\n",
       "\t chr9 & 162355173 & 162355674 & scACR\\_56042;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162362602 & 162363103 & scACR\\_56043;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162498789 & 162499290 & scACR\\_56044;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162575634 & 162576135 & scACR\\_56045;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162759285 & 162759786 & scACR\\_56046;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162766738 & 162767239 & scACR\\_56047;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 162874029 & 162874530 & scACR\\_56048;broadly\\_accessible & NA                 \\\\\n",
       "\t chr9 & 163004080 & 163004581 & scACR\\_56049;broadly\\_accessible & NA                 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A spec_tbl_df: 55504 × 5\n",
       "\n",
       "| chr &lt;chr&gt; | start &lt;dbl&gt; | end &lt;dbl&gt; | acr_id &lt;chr&gt; | score &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| chr1 |   34471 |   34972 | scACR_1;broadly_accessible                 | NA                                     |\n",
       "| chr1 |   45954 |   46455 | scACR_2;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  123085 |  123586 | scACR_3;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  162544 |  163045 | scACR_4;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  165164 |  165665 | scACR_5;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  189283 |  189784 | scACR_6;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  199880 |  200381 | scACR_7;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  206262 |  206763 | scACR_8;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  252910 |  253411 | scACR_9;broadly_accessible                 | NA                                     |\n",
       "| chr1 |  258277 |  258778 | scACR_10;broadly_accessible                | NA                                     |\n",
       "| chr1 |  327853 |  328354 | scACR_11;broadly_accessible                | NA                                     |\n",
       "| chr1 |  328932 |  329433 | scACR_12;broadly_accessible                | NA                                     |\n",
       "| chr1 |  359989 |  360490 | scACR_13;bundle_sheath,procambial_meristem | 0.0283323889203693,0.00636645445151828 |\n",
       "| chr1 |  362377 |  362878 | scACR_14;bundle_sheath                     | 0.0057018279389569                     |\n",
       "| chr1 |  457390 |  457891 | scACR_15;broadly_accessible                | NA                                     |\n",
       "| chr1 |  494213 |  494714 | scACR_16;broadly_accessible                | NA                                     |\n",
       "| chr1 |  564531 |  565032 | scACR_18;broadly_accessible                | NA                                     |\n",
       "| chr1 |  678118 |  678619 | scACR_19;broadly_accessible                | NA                                     |\n",
       "| chr1 |  680747 |  681248 | scACR_20;broadly_accessible                | NA                                     |\n",
       "| chr1 |  707771 |  708272 | scACR_21;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1002473 | 1002974 | scACR_22;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1013849 | 1014350 | scACR_23;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1014416 | 1014917 | scACR_24;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1017973 | 1018474 | scACR_25;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1022248 | 1022749 | scACR_26;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1026915 | 1027416 | scACR_27;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1027798 | 1028299 | scACR_28;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1056836 | 1057337 | scACR_29;bundle_sheath,procambial_meristem | 0.0106996433452218,0.0043665211159628  |\n",
       "| chr1 | 1058017 | 1058518 | scACR_30;broadly_accessible                | NA                                     |\n",
       "| chr1 | 1063713 | 1064214 | scACR_31;broadly_accessible                | NA                                     |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chr9 | 161870287 | 161870788 | scACR_56019;broadly_accessible | NA                  |\n",
       "| chr9 | 161874600 | 161875101 | scACR_56020;broadly_accessible | NA                  |\n",
       "| chr9 | 161892377 | 161892878 | scACR_56021;broadly_accessible | NA                  |\n",
       "| chr9 | 161893303 | 161893804 | scACR_56022;broadly_accessible | NA                  |\n",
       "| chr9 | 161904777 | 161905278 | scACR_56024;broadly_accessible | NA                  |\n",
       "| chr9 | 161956001 | 161956502 | scACR_56025;broadly_accessible | NA                  |\n",
       "| chr9 | 161967292 | 161967793 | scACR_56026;broadly_accessible | NA                  |\n",
       "| chr9 | 161971498 | 161971999 | scACR_56027;broadly_accessible | NA                  |\n",
       "| chr9 | 161972275 | 161972776 | scACR_56028;broadly_accessible | NA                  |\n",
       "| chr9 | 162045573 | 162046074 | scACR_56029;broadly_accessible | NA                  |\n",
       "| chr9 | 162046375 | 162046876 | scACR_56030;broadly_accessible | NA                  |\n",
       "| chr9 | 162047423 | 162047924 | scACR_56031;broadly_accessible | NA                  |\n",
       "| chr9 | 162050722 | 162051223 | scACR_56032;mesophyll          | 0.0129778435262078  |\n",
       "| chr9 | 162065379 | 162065880 | scACR_56033;broadly_accessible | NA                  |\n",
       "| chr9 | 162071240 | 162071741 | scACR_56034;protoderm          | 0.00666644445185161 |\n",
       "| chr9 | 162071854 | 162072355 | scACR_56035;broadly_accessible | NA                  |\n",
       "| chr9 | 162103825 | 162104326 | scACR_56036;broadly_accessible | NA                  |\n",
       "| chr9 | 162232613 | 162233114 | scACR_56037;broadly_accessible | NA                  |\n",
       "| chr9 | 162236054 | 162236555 | scACR_56038;broadly_accessible | NA                  |\n",
       "| chr9 | 162237028 | 162237529 | scACR_56039;broadly_accessible | NA                  |\n",
       "| chr9 | 162278072 | 162278573 | scACR_56040;broadly_accessible | NA                  |\n",
       "| chr9 | 162280414 | 162280915 | scACR_56041;mesophyll          | 0.0267657744741842  |\n",
       "| chr9 | 162355173 | 162355674 | scACR_56042;broadly_accessible | NA                  |\n",
       "| chr9 | 162362602 | 162363103 | scACR_56043;broadly_accessible | NA                  |\n",
       "| chr9 | 162498789 | 162499290 | scACR_56044;broadly_accessible | NA                  |\n",
       "| chr9 | 162575634 | 162576135 | scACR_56045;broadly_accessible | NA                  |\n",
       "| chr9 | 162759285 | 162759786 | scACR_56046;broadly_accessible | NA                  |\n",
       "| chr9 | 162766738 | 162767239 | scACR_56047;broadly_accessible | NA                  |\n",
       "| chr9 | 162874029 | 162874530 | scACR_56048;broadly_accessible | NA                  |\n",
       "| chr9 | 163004080 | 163004581 | scACR_56049;broadly_accessible | NA                  |\n",
       "\n"
      ],
      "text/plain": [
       "      chr  start     end       acr_id                                    \n",
       "1     chr1   34471     34972   scACR_1;broadly_accessible                \n",
       "2     chr1   45954     46455   scACR_2;broadly_accessible                \n",
       "3     chr1  123085    123586   scACR_3;broadly_accessible                \n",
       "4     chr1  162544    163045   scACR_4;broadly_accessible                \n",
       "5     chr1  165164    165665   scACR_5;broadly_accessible                \n",
       "6     chr1  189283    189784   scACR_6;broadly_accessible                \n",
       "7     chr1  199880    200381   scACR_7;broadly_accessible                \n",
       "8     chr1  206262    206763   scACR_8;broadly_accessible                \n",
       "9     chr1  252910    253411   scACR_9;broadly_accessible                \n",
       "10    chr1  258277    258778   scACR_10;broadly_accessible               \n",
       "11    chr1  327853    328354   scACR_11;broadly_accessible               \n",
       "12    chr1  328932    329433   scACR_12;broadly_accessible               \n",
       "13    chr1  359989    360490   scACR_13;bundle_sheath,procambial_meristem\n",
       "14    chr1  362377    362878   scACR_14;bundle_sheath                    \n",
       "15    chr1  457390    457891   scACR_15;broadly_accessible               \n",
       "16    chr1  494213    494714   scACR_16;broadly_accessible               \n",
       "17    chr1  564531    565032   scACR_18;broadly_accessible               \n",
       "18    chr1  678118    678619   scACR_19;broadly_accessible               \n",
       "19    chr1  680747    681248   scACR_20;broadly_accessible               \n",
       "20    chr1  707771    708272   scACR_21;broadly_accessible               \n",
       "21    chr1 1002473   1002974   scACR_22;broadly_accessible               \n",
       "22    chr1 1013849   1014350   scACR_23;broadly_accessible               \n",
       "23    chr1 1014416   1014917   scACR_24;broadly_accessible               \n",
       "24    chr1 1017973   1018474   scACR_25;broadly_accessible               \n",
       "25    chr1 1022248   1022749   scACR_26;broadly_accessible               \n",
       "26    chr1 1026915   1027416   scACR_27;broadly_accessible               \n",
       "27    chr1 1027798   1028299   scACR_28;broadly_accessible               \n",
       "28    chr1 1056836   1057337   scACR_29;bundle_sheath,procambial_meristem\n",
       "29    chr1 1058017   1058518   scACR_30;broadly_accessible               \n",
       "30    chr1 1063713   1064214   scACR_31;broadly_accessible               \n",
       "⋮     ⋮    ⋮         ⋮         ⋮                                         \n",
       "55475 chr9 161870287 161870788 scACR_56019;broadly_accessible            \n",
       "55476 chr9 161874600 161875101 scACR_56020;broadly_accessible            \n",
       "55477 chr9 161892377 161892878 scACR_56021;broadly_accessible            \n",
       "55478 chr9 161893303 161893804 scACR_56022;broadly_accessible            \n",
       "55479 chr9 161904777 161905278 scACR_56024;broadly_accessible            \n",
       "55480 chr9 161956001 161956502 scACR_56025;broadly_accessible            \n",
       "55481 chr9 161967292 161967793 scACR_56026;broadly_accessible            \n",
       "55482 chr9 161971498 161971999 scACR_56027;broadly_accessible            \n",
       "55483 chr9 161972275 161972776 scACR_56028;broadly_accessible            \n",
       "55484 chr9 162045573 162046074 scACR_56029;broadly_accessible            \n",
       "55485 chr9 162046375 162046876 scACR_56030;broadly_accessible            \n",
       "55486 chr9 162047423 162047924 scACR_56031;broadly_accessible            \n",
       "55487 chr9 162050722 162051223 scACR_56032;mesophyll                     \n",
       "55488 chr9 162065379 162065880 scACR_56033;broadly_accessible            \n",
       "55489 chr9 162071240 162071741 scACR_56034;protoderm                     \n",
       "55490 chr9 162071854 162072355 scACR_56035;broadly_accessible            \n",
       "55491 chr9 162103825 162104326 scACR_56036;broadly_accessible            \n",
       "55492 chr9 162232613 162233114 scACR_56037;broadly_accessible            \n",
       "55493 chr9 162236054 162236555 scACR_56038;broadly_accessible            \n",
       "55494 chr9 162237028 162237529 scACR_56039;broadly_accessible            \n",
       "55495 chr9 162278072 162278573 scACR_56040;broadly_accessible            \n",
       "55496 chr9 162280414 162280915 scACR_56041;mesophyll                     \n",
       "55497 chr9 162355173 162355674 scACR_56042;broadly_accessible            \n",
       "55498 chr9 162362602 162363103 scACR_56043;broadly_accessible            \n",
       "55499 chr9 162498789 162499290 scACR_56044;broadly_accessible            \n",
       "55500 chr9 162575634 162576135 scACR_56045;broadly_accessible            \n",
       "55501 chr9 162759285 162759786 scACR_56046;broadly_accessible            \n",
       "55502 chr9 162766738 162767239 scACR_56047;broadly_accessible            \n",
       "55503 chr9 162874029 162874530 scACR_56048;broadly_accessible            \n",
       "55504 chr9 163004080 163004581 scACR_56049;broadly_accessible            \n",
       "      score                                 \n",
       "1     NA                                    \n",
       "2     NA                                    \n",
       "3     NA                                    \n",
       "4     NA                                    \n",
       "5     NA                                    \n",
       "6     NA                                    \n",
       "7     NA                                    \n",
       "8     NA                                    \n",
       "9     NA                                    \n",
       "10    NA                                    \n",
       "11    NA                                    \n",
       "12    NA                                    \n",
       "13    0.0283323889203693,0.00636645445151828\n",
       "14    0.0057018279389569                    \n",
       "15    NA                                    \n",
       "16    NA                                    \n",
       "17    NA                                    \n",
       "18    NA                                    \n",
       "19    NA                                    \n",
       "20    NA                                    \n",
       "21    NA                                    \n",
       "22    NA                                    \n",
       "23    NA                                    \n",
       "24    NA                                    \n",
       "25    NA                                    \n",
       "26    NA                                    \n",
       "27    NA                                    \n",
       "28    0.0106996433452218,0.0043665211159628 \n",
       "29    NA                                    \n",
       "30    NA                                    \n",
       "⋮     ⋮                                     \n",
       "55475 NA                                    \n",
       "55476 NA                                    \n",
       "55477 NA                                    \n",
       "55478 NA                                    \n",
       "55479 NA                                    \n",
       "55480 NA                                    \n",
       "55481 NA                                    \n",
       "55482 NA                                    \n",
       "55483 NA                                    \n",
       "55484 NA                                    \n",
       "55485 NA                                    \n",
       "55486 NA                                    \n",
       "55487 0.0129778435262078                    \n",
       "55488 NA                                    \n",
       "55489 0.00666644445185161                   \n",
       "55490 NA                                    \n",
       "55491 NA                                    \n",
       "55492 NA                                    \n",
       "55493 NA                                    \n",
       "55494 NA                                    \n",
       "55495 NA                                    \n",
       "55496 0.0267657744741842                    \n",
       "55497 NA                                    \n",
       "55498 NA                                    \n",
       "55499 NA                                    \n",
       "55500 NA                                    \n",
       "55501 NA                                    \n",
       "55502 NA                                    \n",
       "55503 NA                                    \n",
       "55504 NA                                    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "zm_acrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "d08916b1-1088-4ea0-91d1-56d6682c74fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hasNs\n",
       "FALSE   Sum \n",
       "47983 47983 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Using MonaLisa for testing Motif Enrichment\n",
    "zm_acrs_annotated <- as_tibble(as.data.frame(zm_acrs)) %>% \n",
    "    tidyr::separate(acr_id, into = c(\"acr\", \"acr_cell_type_specific_class\"), sep = \";\") %>%\n",
    "    dplyr::mutate(acr_cell_type_specific_class = case_when(acr_cell_type_specific_class == \"bundle_sheath,procambial_meristem\" ~ \"bundle_sheath\",\n",
    "                                                          acr_cell_type_specific_class == \"bundle_sheath,procambium\" ~ \"bundle_sheath\", \n",
    "                                                          TRUE ~ acr_cell_type_specific_class)) %>% \n",
    "    left_join(., combined_acrs_marker_info_subset, by = c(\"acr\", \"acr_cell_type_specific_class\")) %>% \n",
    "    mutate(locus = case_when(is.na(locus) == TRUE ~ \"None\",\n",
    "                            TRUE ~ locus),\n",
    "           gene_family = case_when(is.na(gene_family) == TRUE ~ \"None\",\n",
    "                            TRUE ~ locus),\n",
    "           type = case_when(is.na(type) == TRUE ~ \"None\",\n",
    "                            TRUE ~ type),\n",
    "          ) %>% \n",
    "    dplyr::filter(acr_cell_type_specific_class %in% c(\"bundle_sheath\", \"broadly_accessible\", \"mesophyll\", \"bundle_sheath,mesophyll\"))\n",
    "\n",
    "zm_peaks_gr_annotated <- GRanges(zm_acrs_annotated, seqinfo=seqinfo(BSgenome.Zm_B73))\n",
    "names(zm_peaks_gr_annotated) <- as.character(zm_peaks_gr_annotated)\n",
    "zm_peaks_gr_annotated <- unique(zm_peaks_gr_annotated)\n",
    "\n",
    "\n",
    "# get peak sequences, we'll need this later\n",
    "zm_peak_seqs_annotated <- getSeq(BSgenome.Zm_B73, zm_peaks_gr_annotated) # <2s\n",
    "\n",
    "\n",
    "# exclude peaks with any N's, to avoid warnings further down\n",
    "hasNs    <- (alphabetFrequency(zm_peak_seqs_annotated)[,\"N\"]>0)\n",
    "addmargins(table(hasNs)) # => only drops 5 sequences\n",
    "\n",
    "zm_peak_seqs_annotated <- zm_peak_seqs_annotated[!hasNs]\n",
    "zm_peak_seqs_annotated  <- zm_peak_seqs_annotated[!hasNs]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "4741a3ff-099b-4c2a-b6f0-e5156afedcdc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       " 35.883   7.787  19.118 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(monaLisa)\n",
    "# we first have to set up a way to define the \"common\" state as the background\n",
    "zm_peaks_gr_annotated$acr_cell_type_specific_class <- setZeroBin(factor(zm_peaks_gr_annotated$acr_cell_type_specific_class), \"broadly_accessible\")\n",
    "\n",
    "## Testing if the ACR class if signifigant in any way\n",
    "\n",
    "# now run monaLisa\n",
    "system.time( check_res_acr_type <- calcBinnedMotifEnrR(\n",
    "  zm_peak_seqs_annotated,\n",
    "  bins = zm_peaks_gr_annotated$acr_cell_type_specific_class,\n",
    "  pwmL = pw_matrix_list_converted,\n",
    "  background = \"otherBins\", \n",
    "  min.score = 3,\n",
    "  BPPARAM =  BiocParallel::MulticoreParam(16)\n",
    ") ) # ~40s on my laptop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "0b60b8b2-74ea-402f-9628-3f3747d12b4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 8 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif.id</th><th scope=col>motif.name</th><th scope=col>OR.bs</th><th scope=col>OR.ms</th><th scope=col>fdr.bs</th><th scope=col>fdr.ms</th><th scope=col>pval.bs</th><th scope=col>pval.ms</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BES1    </td><td>BES1    </td><td>0.9910325</td><td>1.0291767</td><td>1.000000e+00</td><td>1.899897e-02</td><td>7.404542e-01</td><td>3.166495e-03</td></tr>\n",
       "\t<tr><td>C2C2_Dof</td><td>C2C2_Dof</td><td>1.0026643</td><td>0.9998445</td><td>1.706511e-01</td><td>1.000000e+00</td><td>4.875746e-02</td><td>6.318872e-01</td></tr>\n",
       "\t<tr><td>bs_DOF  </td><td>bs_DOF  </td><td>1.0143999</td><td>0.9965168</td><td>6.151562e-06</td><td>1.000000e+00</td><td>5.948632e-07</td><td>8.644863e-01</td></tr>\n",
       "\t<tr><td>ms_AGCTA</td><td>ms_AGCTA</td><td>1.0264024</td><td>1.0246645</td><td>9.960413e-02</td><td>3.865624e-02</td><td>2.371527e-02</td><td>7.823286e-03</td></tr>\n",
       "\t<tr><td>ms_GACA </td><td>ms_GACA </td><td>1.0284745</td><td>1.0412858</td><td>9.401424e-02</td><td>1.808378e-04</td><td>2.126513e-02</td><td>2.152831e-05</td></tr>\n",
       "\t<tr><td>BBRBPC  </td><td>BBRBPC  </td><td>1.0370805</td><td>1.0065790</td><td>5.626082e-04</td><td>6.427478e-01</td><td>7.367488e-05</td><td>2.325162e-01</td></tr>\n",
       "\t<tr><td>NAC     </td><td>NAC     </td><td>1.0567678</td><td>1.0125019</td><td>2.365128e-02</td><td>6.427478e-01</td><td>4.505005e-03</td><td>2.414374e-01</td></tr>\n",
       "\t<tr><td>bs_CATG </td><td>bs_CATG </td><td>1.1723881</td><td>1.1110174</td><td>6.265750e-08</td><td>6.151562e-06</td><td>2.983690e-09</td><td>4.128956e-07</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 8 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " motif.id & motif.name & OR.bs & OR.ms & fdr.bs & fdr.ms & pval.bs & pval.ms\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t BES1     & BES1     & 0.9910325 & 1.0291767 & 1.000000e+00 & 1.899897e-02 & 7.404542e-01 & 3.166495e-03\\\\\n",
       "\t C2C2\\_Dof & C2C2\\_Dof & 1.0026643 & 0.9998445 & 1.706511e-01 & 1.000000e+00 & 4.875746e-02 & 6.318872e-01\\\\\n",
       "\t bs\\_DOF   & bs\\_DOF   & 1.0143999 & 0.9965168 & 6.151562e-06 & 1.000000e+00 & 5.948632e-07 & 8.644863e-01\\\\\n",
       "\t ms\\_AGCTA & ms\\_AGCTA & 1.0264024 & 1.0246645 & 9.960413e-02 & 3.865624e-02 & 2.371527e-02 & 7.823286e-03\\\\\n",
       "\t ms\\_GACA  & ms\\_GACA  & 1.0284745 & 1.0412858 & 9.401424e-02 & 1.808378e-04 & 2.126513e-02 & 2.152831e-05\\\\\n",
       "\t BBRBPC   & BBRBPC   & 1.0370805 & 1.0065790 & 5.626082e-04 & 6.427478e-01 & 7.367488e-05 & 2.325162e-01\\\\\n",
       "\t NAC      & NAC      & 1.0567678 & 1.0125019 & 2.365128e-02 & 6.427478e-01 & 4.505005e-03 & 2.414374e-01\\\\\n",
       "\t bs\\_CATG  & bs\\_CATG  & 1.1723881 & 1.1110174 & 6.265750e-08 & 6.151562e-06 & 2.983690e-09 & 4.128956e-07\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 8 × 8\n",
       "\n",
       "| motif.id &lt;chr&gt; | motif.name &lt;chr&gt; | OR.bs &lt;dbl&gt; | OR.ms &lt;dbl&gt; | fdr.bs &lt;dbl&gt; | fdr.ms &lt;dbl&gt; | pval.bs &lt;dbl&gt; | pval.ms &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| BES1     | BES1     | 0.9910325 | 1.0291767 | 1.000000e+00 | 1.899897e-02 | 7.404542e-01 | 3.166495e-03 |\n",
       "| C2C2_Dof | C2C2_Dof | 1.0026643 | 0.9998445 | 1.706511e-01 | 1.000000e+00 | 4.875746e-02 | 6.318872e-01 |\n",
       "| bs_DOF   | bs_DOF   | 1.0143999 | 0.9965168 | 6.151562e-06 | 1.000000e+00 | 5.948632e-07 | 8.644863e-01 |\n",
       "| ms_AGCTA | ms_AGCTA | 1.0264024 | 1.0246645 | 9.960413e-02 | 3.865624e-02 | 2.371527e-02 | 7.823286e-03 |\n",
       "| ms_GACA  | ms_GACA  | 1.0284745 | 1.0412858 | 9.401424e-02 | 1.808378e-04 | 2.126513e-02 | 2.152831e-05 |\n",
       "| BBRBPC   | BBRBPC   | 1.0370805 | 1.0065790 | 5.626082e-04 | 6.427478e-01 | 7.367488e-05 | 2.325162e-01 |\n",
       "| NAC      | NAC      | 1.0567678 | 1.0125019 | 2.365128e-02 | 6.427478e-01 | 4.505005e-03 | 2.414374e-01 |\n",
       "| bs_CATG  | bs_CATG  | 1.1723881 | 1.1110174 | 6.265750e-08 | 6.151562e-06 | 2.983690e-09 | 4.128956e-07 |\n",
       "\n"
      ],
      "text/plain": [
       "  motif.id motif.name OR.bs     OR.ms     fdr.bs       fdr.ms      \n",
       "1 BES1     BES1       0.9910325 1.0291767 1.000000e+00 1.899897e-02\n",
       "2 C2C2_Dof C2C2_Dof   1.0026643 0.9998445 1.706511e-01 1.000000e+00\n",
       "3 bs_DOF   bs_DOF     1.0143999 0.9965168 6.151562e-06 1.000000e+00\n",
       "4 ms_AGCTA ms_AGCTA   1.0264024 1.0246645 9.960413e-02 3.865624e-02\n",
       "5 ms_GACA  ms_GACA    1.0284745 1.0412858 9.401424e-02 1.808378e-04\n",
       "6 BBRBPC   BBRBPC     1.0370805 1.0065790 5.626082e-04 6.427478e-01\n",
       "7 NAC      NAC        1.0567678 1.0125019 2.365128e-02 6.427478e-01\n",
       "8 bs_CATG  bs_CATG    1.1723881 1.1110174 6.265750e-08 6.151562e-06\n",
       "  pval.bs      pval.ms     \n",
       "1 7.404542e-01 3.166495e-03\n",
       "2 4.875746e-02 6.318872e-01\n",
       "3 5.948632e-07 8.644863e-01\n",
       "4 2.371527e-02 7.823286e-03\n",
       "5 2.126513e-02 2.152831e-05\n",
       "6 7.367488e-05 2.325162e-01\n",
       "7 4.505005e-03 2.414374e-01\n",
       "8 2.983690e-09 4.128956e-07"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# reformat the results for convenience\n",
    "ora_res2_consider_acr_type <- cbind(\n",
    "  rowData(check_res_acr_type)[,c(\"motif.id\",\"motif.name\")],\n",
    "  2^assay(check_res_acr_type, \"log2enr\")[,c(\"bundle_sheath\",\"mesophyll\")],  # convert to OR\n",
    "  10^(-1 * assay(check_res_acr_type, \"negLog10Padj\")[,c(\"bundle_sheath\",\"mesophyll\")]), # convert to p-val\n",
    "  10^(-1 * assay(check_res_acr_type, \"negLog10P\")[,c(\"bundle_sheath\",\"mesophyll\")]) # convert to p-val\n",
    ") \n",
    "colnames(ora_res2_consider_acr_type)[3:8] <- c(\"OR.bs\", \"OR.ms\", \n",
    "                             \"fdr.bs\", \"fdr.ms\", \n",
    "                             \"pval.bs\", \"pval.ms\")\n",
    "\n",
    "\n",
    "### Yuuuuup - very nice, aligns with expectations\n",
    "as_tibble(ora_res2_consider_acr_type) %>% \n",
    "    arrange(OR.bs, OR.ms) %>% \n",
    "    dplyr::filter(pval.bs < .05 | pval.ms < .05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "fc1fae52-98ef-4531-847e-a3acd18776eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       " 48.783   9.051  19.589 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(monaLisa)\n",
    "# we first have to set up a way to define the \"common\" state as the background\n",
    "zm_peaks_gr_annotated$type <- setZeroBin(factor(zm_peaks_gr_annotated$type), \"None\")\n",
    "\n",
    "# now run monaLisa\n",
    "system.time( check_res_type <- calcBinnedMotifEnrR(\n",
    "  zm_peak_seqs_annotated,\n",
    "  bins = zm_peaks_gr_annotated$type,\n",
    "  pwmL = pw_matrix_list_converted,\n",
    "  background = \"otherBins\", \n",
    "  min.score = 3,\n",
    "  BPPARAM =  BiocParallel::MulticoreParam(16)\n",
    ") ) # ~40s on my laptop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "5794a040-05c4-468a-994e-a92cbbb2a8a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 28 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif.id</th><th scope=col>motif.name</th><th scope=col>OR.bs</th><th scope=col>OR.ms</th><th scope=col>OR.ms_bs</th><th scope=col>fdr.bs</th><th scope=col>fdr.ms</th><th scope=col>fdr.ms_bs</th><th scope=col>pval.bs</th><th scope=col>pval.ms</th><th scope=col>pval.ms_bs</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB      </td><td>AS2LOB      </td><td>0.9351690</td><td>0.9279546</td><td>1.0236580</td><td>1</td><td>1</td><td>1</td><td>0.9740776</td><td>0.98202857</td><td>0.49448293</td></tr>\n",
       "\t<tr><td>HD-Zip_IV   </td><td>HD-Zip_IV   </td><td>0.9535660</td><td>0.8861025</td><td>1.0850942</td><td>1</td><td>1</td><td>1</td><td>0.8681568</td><td>0.99622827</td><td>0.18088532</td></tr>\n",
       "\t<tr><td>HSF         </td><td>HSF         </td><td>0.9556872</td><td>1.0059945</td><td>1.0160428</td><td>1</td><td>1</td><td>1</td><td>0.7527697</td><td>0.53226094</td><td>0.53329654</td></tr>\n",
       "\t<tr><td>CPP         </td><td>CPP         </td><td>0.9591760</td><td>0.9348583</td><td>1.1619745</td><td>1</td><td>1</td><td>1</td><td>0.8064409</td><td>0.92111661</td><td>0.04159156</td></tr>\n",
       "\t<tr><td>ms_AGCTA    </td><td>ms_AGCTA    </td><td>0.9595508</td><td>1.0212335</td><td>1.0574985</td><td>1</td><td>1</td><td>1</td><td>0.8206621</td><td>0.42591781</td><td>0.29257470</td></tr>\n",
       "\t<tr><td>bs_GCCGCC   </td><td>bs_GCCGCC   </td><td>0.9644805</td><td>1.0463189</td><td>1.0032460</td><td>1</td><td>1</td><td>1</td><td>0.7258991</td><td>0.33855958</td><td>0.61128779</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>bHLH_TCP    </td><td>0.9741285</td><td>1.0215303</td><td>0.9747614</td><td>1</td><td>1</td><td>1</td><td>0.9640359</td><td>0.25557447</td><td>0.97448692</td></tr>\n",
       "\t<tr><td>SBP         </td><td>SBP         </td><td>0.9765223</td><td>0.9732956</td><td>0.9603793</td><td>1</td><td>1</td><td>1</td><td>0.8061510</td><td>0.82718454</td><td>0.88576430</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>MADS_MIKC   </td><td>0.9890780</td><td>0.9934723</td><td>0.9848422</td><td>1</td><td>1</td><td>1</td><td>0.9227107</td><td>0.88316849</td><td>0.89497279</td></tr>\n",
       "\t<tr><td>E2FDP       </td><td>E2FDP       </td><td>0.9909936</td><td>1.0064495</td><td>1.0329472</td><td>1</td><td>1</td><td>1</td><td>0.7136156</td><td>0.54681696</td><td>0.37607269</td></tr>\n",
       "\t<tr><td>HD          </td><td>HD          </td><td>0.9956306</td><td>1.0002314</td><td>0.9354426</td><td>1</td><td>1</td><td>1</td><td>0.6215830</td><td>0.58597529</td><td>0.84925557</td></tr>\n",
       "\t<tr><td>ARF         </td><td>ARF         </td><td>1.0001096</td><td>1.0002851</td><td>1.0000560</td><td>1</td><td>1</td><td>1</td><td>0.9945826</td><td>0.98225011</td><td>0.99850411</td></tr>\n",
       "\t<tr><td>GARP_G2-like</td><td>GARP_G2-like</td><td>1.0013522</td><td>1.0021630</td><td>1.0026539</td><td>1</td><td>1</td><td>1</td><td>0.9343063</td><td>0.87053700</td><td>0.92359462</td></tr>\n",
       "\t<tr><td>HD_PLINC    </td><td>HD_PLINC    </td><td>1.0017065</td><td>0.9726633</td><td>1.0231815</td><td>1</td><td>1</td><td>1</td><td>0.6018928</td><td>0.83182329</td><td>0.50118864</td></tr>\n",
       "\t<tr><td>C2C2_Dof    </td><td>C2C2_Dof    </td><td>1.0017222</td><td>1.0019794</td><td>1.0029666</td><td>1</td><td>1</td><td>1</td><td>0.9175119</td><td>0.88101615</td><td>0.91484639</td></tr>\n",
       "\t<tr><td>ms_GACGA    </td><td>ms_GACGA    </td><td>1.0026893</td><td>1.0048496</td><td>1.0006329</td><td>1</td><td>1</td><td>1</td><td>0.8744165</td><td>0.73342158</td><td>0.98120918</td></tr>\n",
       "\t<tr><td>WRKY        </td><td>WRKY        </td><td>1.0028250</td><td>0.9830634</td><td>1.0613005</td><td>1</td><td>1</td><td>1</td><td>0.6391319</td><td>0.82563639</td><td>0.16456400</td></tr>\n",
       "\t<tr><td>BES1        </td><td>BES1        </td><td>1.0054051</td><td>1.0556240</td><td>1.0701391</td><td>1</td><td>1</td><td>1</td><td>0.5552891</td><td>0.23225436</td><td>0.23576572</td></tr>\n",
       "\t<tr><td>CAMTA       </td><td>CAMTA       </td><td>1.0098625</td><td>0.9813595</td><td>1.0023857</td><td>1</td><td>1</td><td>1</td><td>0.6121285</td><td>0.95058773</td><td>0.93102344</td></tr>\n",
       "\t<tr><td>GARP_ARR-B  </td><td>GARP_ARR-B  </td><td>1.0106659</td><td>1.0101870</td><td>1.0148945</td><td>1</td><td>1</td><td>1</td><td>0.5879448</td><td>0.52289695</td><td>0.64110214</td></tr>\n",
       "\t<tr><td>bs_DOF      </td><td>bs_DOF      </td><td>1.0115313</td><td>0.9955115</td><td>1.0176103</td><td>1</td><td>1</td><td>1</td><td>0.5636461</td><td>0.83938205</td><td>0.59145898</td></tr>\n",
       "\t<tr><td>NAC         </td><td>NAC         </td><td>1.0127675</td><td>1.0700564</td><td>0.9035369</td><td>1</td><td>1</td><td>1</td><td>0.5127927</td><td>0.22220035</td><td>0.86325391</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II </td><td>HD-Zip_I_II </td><td>1.0161510</td><td>1.0018182</td><td>0.9961367</td><td>1</td><td>1</td><td>1</td><td>0.4772459</td><td>0.60685031</td><td>0.67275770</td></tr>\n",
       "\t<tr><td>C2C2_GATA   </td><td>C2C2_GATA   </td><td>1.0305666</td><td>1.0370713</td><td>1.0536538</td><td>1</td><td>1</td><td>1</td><td>0.2879184</td><td>0.15628244</td><td>0.20536499</td></tr>\n",
       "\t<tr><td>bs_CATG     </td><td>bs_CATG     </td><td>1.0313836</td><td>0.9459646</td><td>0.9210256</td><td>1</td><td>1</td><td>1</td><td>0.4549189</td><td>0.72803751</td><td>0.79769442</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>BBRBPC      </td><td>1.0337345</td><td>0.9869805</td><td>0.9847691</td><td>1</td><td>1</td><td>1</td><td>0.3220876</td><td>0.71581069</td><td>0.74596800</td></tr>\n",
       "\t<tr><td>ms_CGTCGTCGT</td><td>ms_CGTCGTCGT</td><td>1.0341785</td><td>1.0008447</td><td>0.9397306</td><td>1</td><td>1</td><td>1</td><td>0.4034011</td><td>0.55904092</td><td>0.88859017</td></tr>\n",
       "\t<tr><td>ms_GACA     </td><td>ms_GACA     </td><td>1.0439849</td><td>1.0918212</td><td>1.0084923</td><td>1</td><td>1</td><td>1</td><td>0.3071890</td><td>0.06280049</td><td>0.58207351</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 28 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " motif.id & motif.name & OR.bs & OR.ms & OR.ms\\_bs & fdr.bs & fdr.ms & fdr.ms\\_bs & pval.bs & pval.ms & pval.ms\\_bs\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AS2LOB       & AS2LOB       & 0.9351690 & 0.9279546 & 1.0236580 & 1 & 1 & 1 & 0.9740776 & 0.98202857 & 0.49448293\\\\\n",
       "\t HD-Zip\\_IV    & HD-Zip\\_IV    & 0.9535660 & 0.8861025 & 1.0850942 & 1 & 1 & 1 & 0.8681568 & 0.99622827 & 0.18088532\\\\\n",
       "\t HSF          & HSF          & 0.9556872 & 1.0059945 & 1.0160428 & 1 & 1 & 1 & 0.7527697 & 0.53226094 & 0.53329654\\\\\n",
       "\t CPP          & CPP          & 0.9591760 & 0.9348583 & 1.1619745 & 1 & 1 & 1 & 0.8064409 & 0.92111661 & 0.04159156\\\\\n",
       "\t ms\\_AGCTA     & ms\\_AGCTA     & 0.9595508 & 1.0212335 & 1.0574985 & 1 & 1 & 1 & 0.8206621 & 0.42591781 & 0.29257470\\\\\n",
       "\t bs\\_GCCGCC    & bs\\_GCCGCC    & 0.9644805 & 1.0463189 & 1.0032460 & 1 & 1 & 1 & 0.7258991 & 0.33855958 & 0.61128779\\\\\n",
       "\t bHLH\\_TCP     & bHLH\\_TCP     & 0.9741285 & 1.0215303 & 0.9747614 & 1 & 1 & 1 & 0.9640359 & 0.25557447 & 0.97448692\\\\\n",
       "\t SBP          & SBP          & 0.9765223 & 0.9732956 & 0.9603793 & 1 & 1 & 1 & 0.8061510 & 0.82718454 & 0.88576430\\\\\n",
       "\t MADS\\_MIKC    & MADS\\_MIKC    & 0.9890780 & 0.9934723 & 0.9848422 & 1 & 1 & 1 & 0.9227107 & 0.88316849 & 0.89497279\\\\\n",
       "\t E2FDP        & E2FDP        & 0.9909936 & 1.0064495 & 1.0329472 & 1 & 1 & 1 & 0.7136156 & 0.54681696 & 0.37607269\\\\\n",
       "\t HD           & HD           & 0.9956306 & 1.0002314 & 0.9354426 & 1 & 1 & 1 & 0.6215830 & 0.58597529 & 0.84925557\\\\\n",
       "\t ARF          & ARF          & 1.0001096 & 1.0002851 & 1.0000560 & 1 & 1 & 1 & 0.9945826 & 0.98225011 & 0.99850411\\\\\n",
       "\t GARP\\_G2-like & GARP\\_G2-like & 1.0013522 & 1.0021630 & 1.0026539 & 1 & 1 & 1 & 0.9343063 & 0.87053700 & 0.92359462\\\\\n",
       "\t HD\\_PLINC     & HD\\_PLINC     & 1.0017065 & 0.9726633 & 1.0231815 & 1 & 1 & 1 & 0.6018928 & 0.83182329 & 0.50118864\\\\\n",
       "\t C2C2\\_Dof     & C2C2\\_Dof     & 1.0017222 & 1.0019794 & 1.0029666 & 1 & 1 & 1 & 0.9175119 & 0.88101615 & 0.91484639\\\\\n",
       "\t ms\\_GACGA     & ms\\_GACGA     & 1.0026893 & 1.0048496 & 1.0006329 & 1 & 1 & 1 & 0.8744165 & 0.73342158 & 0.98120918\\\\\n",
       "\t WRKY         & WRKY         & 1.0028250 & 0.9830634 & 1.0613005 & 1 & 1 & 1 & 0.6391319 & 0.82563639 & 0.16456400\\\\\n",
       "\t BES1         & BES1         & 1.0054051 & 1.0556240 & 1.0701391 & 1 & 1 & 1 & 0.5552891 & 0.23225436 & 0.23576572\\\\\n",
       "\t CAMTA        & CAMTA        & 1.0098625 & 0.9813595 & 1.0023857 & 1 & 1 & 1 & 0.6121285 & 0.95058773 & 0.93102344\\\\\n",
       "\t GARP\\_ARR-B   & GARP\\_ARR-B   & 1.0106659 & 1.0101870 & 1.0148945 & 1 & 1 & 1 & 0.5879448 & 0.52289695 & 0.64110214\\\\\n",
       "\t bs\\_DOF       & bs\\_DOF       & 1.0115313 & 0.9955115 & 1.0176103 & 1 & 1 & 1 & 0.5636461 & 0.83938205 & 0.59145898\\\\\n",
       "\t NAC          & NAC          & 1.0127675 & 1.0700564 & 0.9035369 & 1 & 1 & 1 & 0.5127927 & 0.22220035 & 0.86325391\\\\\n",
       "\t HD-Zip\\_I\\_II  & HD-Zip\\_I\\_II  & 1.0161510 & 1.0018182 & 0.9961367 & 1 & 1 & 1 & 0.4772459 & 0.60685031 & 0.67275770\\\\\n",
       "\t C2C2\\_GATA    & C2C2\\_GATA    & 1.0305666 & 1.0370713 & 1.0536538 & 1 & 1 & 1 & 0.2879184 & 0.15628244 & 0.20536499\\\\\n",
       "\t bs\\_CATG      & bs\\_CATG      & 1.0313836 & 0.9459646 & 0.9210256 & 1 & 1 & 1 & 0.4549189 & 0.72803751 & 0.79769442\\\\\n",
       "\t BBRBPC       & BBRBPC       & 1.0337345 & 0.9869805 & 0.9847691 & 1 & 1 & 1 & 0.3220876 & 0.71581069 & 0.74596800\\\\\n",
       "\t ms\\_CGTCGTCGT & ms\\_CGTCGTCGT & 1.0341785 & 1.0008447 & 0.9397306 & 1 & 1 & 1 & 0.4034011 & 0.55904092 & 0.88859017\\\\\n",
       "\t ms\\_GACA      & ms\\_GACA      & 1.0439849 & 1.0918212 & 1.0084923 & 1 & 1 & 1 & 0.3071890 & 0.06280049 & 0.58207351\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 28 × 11\n",
       "\n",
       "| motif.id &lt;chr&gt; | motif.name &lt;chr&gt; | OR.bs &lt;dbl&gt; | OR.ms &lt;dbl&gt; | OR.ms_bs &lt;dbl&gt; | fdr.bs &lt;dbl&gt; | fdr.ms &lt;dbl&gt; | fdr.ms_bs &lt;dbl&gt; | pval.bs &lt;dbl&gt; | pval.ms &lt;dbl&gt; | pval.ms_bs &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| AS2LOB       | AS2LOB       | 0.9351690 | 0.9279546 | 1.0236580 | 1 | 1 | 1 | 0.9740776 | 0.98202857 | 0.49448293 |\n",
       "| HD-Zip_IV    | HD-Zip_IV    | 0.9535660 | 0.8861025 | 1.0850942 | 1 | 1 | 1 | 0.8681568 | 0.99622827 | 0.18088532 |\n",
       "| HSF          | HSF          | 0.9556872 | 1.0059945 | 1.0160428 | 1 | 1 | 1 | 0.7527697 | 0.53226094 | 0.53329654 |\n",
       "| CPP          | CPP          | 0.9591760 | 0.9348583 | 1.1619745 | 1 | 1 | 1 | 0.8064409 | 0.92111661 | 0.04159156 |\n",
       "| ms_AGCTA     | ms_AGCTA     | 0.9595508 | 1.0212335 | 1.0574985 | 1 | 1 | 1 | 0.8206621 | 0.42591781 | 0.29257470 |\n",
       "| bs_GCCGCC    | bs_GCCGCC    | 0.9644805 | 1.0463189 | 1.0032460 | 1 | 1 | 1 | 0.7258991 | 0.33855958 | 0.61128779 |\n",
       "| bHLH_TCP     | bHLH_TCP     | 0.9741285 | 1.0215303 | 0.9747614 | 1 | 1 | 1 | 0.9640359 | 0.25557447 | 0.97448692 |\n",
       "| SBP          | SBP          | 0.9765223 | 0.9732956 | 0.9603793 | 1 | 1 | 1 | 0.8061510 | 0.82718454 | 0.88576430 |\n",
       "| MADS_MIKC    | MADS_MIKC    | 0.9890780 | 0.9934723 | 0.9848422 | 1 | 1 | 1 | 0.9227107 | 0.88316849 | 0.89497279 |\n",
       "| E2FDP        | E2FDP        | 0.9909936 | 1.0064495 | 1.0329472 | 1 | 1 | 1 | 0.7136156 | 0.54681696 | 0.37607269 |\n",
       "| HD           | HD           | 0.9956306 | 1.0002314 | 0.9354426 | 1 | 1 | 1 | 0.6215830 | 0.58597529 | 0.84925557 |\n",
       "| ARF          | ARF          | 1.0001096 | 1.0002851 | 1.0000560 | 1 | 1 | 1 | 0.9945826 | 0.98225011 | 0.99850411 |\n",
       "| GARP_G2-like | GARP_G2-like | 1.0013522 | 1.0021630 | 1.0026539 | 1 | 1 | 1 | 0.9343063 | 0.87053700 | 0.92359462 |\n",
       "| HD_PLINC     | HD_PLINC     | 1.0017065 | 0.9726633 | 1.0231815 | 1 | 1 | 1 | 0.6018928 | 0.83182329 | 0.50118864 |\n",
       "| C2C2_Dof     | C2C2_Dof     | 1.0017222 | 1.0019794 | 1.0029666 | 1 | 1 | 1 | 0.9175119 | 0.88101615 | 0.91484639 |\n",
       "| ms_GACGA     | ms_GACGA     | 1.0026893 | 1.0048496 | 1.0006329 | 1 | 1 | 1 | 0.8744165 | 0.73342158 | 0.98120918 |\n",
       "| WRKY         | WRKY         | 1.0028250 | 0.9830634 | 1.0613005 | 1 | 1 | 1 | 0.6391319 | 0.82563639 | 0.16456400 |\n",
       "| BES1         | BES1         | 1.0054051 | 1.0556240 | 1.0701391 | 1 | 1 | 1 | 0.5552891 | 0.23225436 | 0.23576572 |\n",
       "| CAMTA        | CAMTA        | 1.0098625 | 0.9813595 | 1.0023857 | 1 | 1 | 1 | 0.6121285 | 0.95058773 | 0.93102344 |\n",
       "| GARP_ARR-B   | GARP_ARR-B   | 1.0106659 | 1.0101870 | 1.0148945 | 1 | 1 | 1 | 0.5879448 | 0.52289695 | 0.64110214 |\n",
       "| bs_DOF       | bs_DOF       | 1.0115313 | 0.9955115 | 1.0176103 | 1 | 1 | 1 | 0.5636461 | 0.83938205 | 0.59145898 |\n",
       "| NAC          | NAC          | 1.0127675 | 1.0700564 | 0.9035369 | 1 | 1 | 1 | 0.5127927 | 0.22220035 | 0.86325391 |\n",
       "| HD-Zip_I_II  | HD-Zip_I_II  | 1.0161510 | 1.0018182 | 0.9961367 | 1 | 1 | 1 | 0.4772459 | 0.60685031 | 0.67275770 |\n",
       "| C2C2_GATA    | C2C2_GATA    | 1.0305666 | 1.0370713 | 1.0536538 | 1 | 1 | 1 | 0.2879184 | 0.15628244 | 0.20536499 |\n",
       "| bs_CATG      | bs_CATG      | 1.0313836 | 0.9459646 | 0.9210256 | 1 | 1 | 1 | 0.4549189 | 0.72803751 | 0.79769442 |\n",
       "| BBRBPC       | BBRBPC       | 1.0337345 | 0.9869805 | 0.9847691 | 1 | 1 | 1 | 0.3220876 | 0.71581069 | 0.74596800 |\n",
       "| ms_CGTCGTCGT | ms_CGTCGTCGT | 1.0341785 | 1.0008447 | 0.9397306 | 1 | 1 | 1 | 0.4034011 | 0.55904092 | 0.88859017 |\n",
       "| ms_GACA      | ms_GACA      | 1.0439849 | 1.0918212 | 1.0084923 | 1 | 1 | 1 | 0.3071890 | 0.06280049 | 0.58207351 |\n",
       "\n"
      ],
      "text/plain": [
       "   motif.id     motif.name   OR.bs     OR.ms     OR.ms_bs  fdr.bs fdr.ms\n",
       "1  AS2LOB       AS2LOB       0.9351690 0.9279546 1.0236580 1      1     \n",
       "2  HD-Zip_IV    HD-Zip_IV    0.9535660 0.8861025 1.0850942 1      1     \n",
       "3  HSF          HSF          0.9556872 1.0059945 1.0160428 1      1     \n",
       "4  CPP          CPP          0.9591760 0.9348583 1.1619745 1      1     \n",
       "5  ms_AGCTA     ms_AGCTA     0.9595508 1.0212335 1.0574985 1      1     \n",
       "6  bs_GCCGCC    bs_GCCGCC    0.9644805 1.0463189 1.0032460 1      1     \n",
       "7  bHLH_TCP     bHLH_TCP     0.9741285 1.0215303 0.9747614 1      1     \n",
       "8  SBP          SBP          0.9765223 0.9732956 0.9603793 1      1     \n",
       "9  MADS_MIKC    MADS_MIKC    0.9890780 0.9934723 0.9848422 1      1     \n",
       "10 E2FDP        E2FDP        0.9909936 1.0064495 1.0329472 1      1     \n",
       "11 HD           HD           0.9956306 1.0002314 0.9354426 1      1     \n",
       "12 ARF          ARF          1.0001096 1.0002851 1.0000560 1      1     \n",
       "13 GARP_G2-like GARP_G2-like 1.0013522 1.0021630 1.0026539 1      1     \n",
       "14 HD_PLINC     HD_PLINC     1.0017065 0.9726633 1.0231815 1      1     \n",
       "15 C2C2_Dof     C2C2_Dof     1.0017222 1.0019794 1.0029666 1      1     \n",
       "16 ms_GACGA     ms_GACGA     1.0026893 1.0048496 1.0006329 1      1     \n",
       "17 WRKY         WRKY         1.0028250 0.9830634 1.0613005 1      1     \n",
       "18 BES1         BES1         1.0054051 1.0556240 1.0701391 1      1     \n",
       "19 CAMTA        CAMTA        1.0098625 0.9813595 1.0023857 1      1     \n",
       "20 GARP_ARR-B   GARP_ARR-B   1.0106659 1.0101870 1.0148945 1      1     \n",
       "21 bs_DOF       bs_DOF       1.0115313 0.9955115 1.0176103 1      1     \n",
       "22 NAC          NAC          1.0127675 1.0700564 0.9035369 1      1     \n",
       "23 HD-Zip_I_II  HD-Zip_I_II  1.0161510 1.0018182 0.9961367 1      1     \n",
       "24 C2C2_GATA    C2C2_GATA    1.0305666 1.0370713 1.0536538 1      1     \n",
       "25 bs_CATG      bs_CATG      1.0313836 0.9459646 0.9210256 1      1     \n",
       "26 BBRBPC       BBRBPC       1.0337345 0.9869805 0.9847691 1      1     \n",
       "27 ms_CGTCGTCGT ms_CGTCGTCGT 1.0341785 1.0008447 0.9397306 1      1     \n",
       "28 ms_GACA      ms_GACA      1.0439849 1.0918212 1.0084923 1      1     \n",
       "   fdr.ms_bs pval.bs   pval.ms    pval.ms_bs\n",
       "1  1         0.9740776 0.98202857 0.49448293\n",
       "2  1         0.8681568 0.99622827 0.18088532\n",
       "3  1         0.7527697 0.53226094 0.53329654\n",
       "4  1         0.8064409 0.92111661 0.04159156\n",
       "5  1         0.8206621 0.42591781 0.29257470\n",
       "6  1         0.7258991 0.33855958 0.61128779\n",
       "7  1         0.9640359 0.25557447 0.97448692\n",
       "8  1         0.8061510 0.82718454 0.88576430\n",
       "9  1         0.9227107 0.88316849 0.89497279\n",
       "10 1         0.7136156 0.54681696 0.37607269\n",
       "11 1         0.6215830 0.58597529 0.84925557\n",
       "12 1         0.9945826 0.98225011 0.99850411\n",
       "13 1         0.9343063 0.87053700 0.92359462\n",
       "14 1         0.6018928 0.83182329 0.50118864\n",
       "15 1         0.9175119 0.88101615 0.91484639\n",
       "16 1         0.8744165 0.73342158 0.98120918\n",
       "17 1         0.6391319 0.82563639 0.16456400\n",
       "18 1         0.5552891 0.23225436 0.23576572\n",
       "19 1         0.6121285 0.95058773 0.93102344\n",
       "20 1         0.5879448 0.52289695 0.64110214\n",
       "21 1         0.5636461 0.83938205 0.59145898\n",
       "22 1         0.5127927 0.22220035 0.86325391\n",
       "23 1         0.4772459 0.60685031 0.67275770\n",
       "24 1         0.2879184 0.15628244 0.20536499\n",
       "25 1         0.4549189 0.72803751 0.79769442\n",
       "26 1         0.3220876 0.71581069 0.74596800\n",
       "27 1         0.4034011 0.55904092 0.88859017\n",
       "28 1         0.3071890 0.06280049 0.58207351"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# reformat the results for convenience\n",
    "ora_res2_consider_type <- cbind(\n",
    "  rowData(check_res_type)[,c(\"motif.id\",\"motif.name\")],\n",
    "  2^assay(check_res_type, \"log2enr\")[,c(\"bundle_sheath\",\"mesophyll\", \"mesophyll,bundle_sheath\")],  # convert to OR\n",
    "  10^(-1 * assay(check_res_type, \"negLog10Padj\")[,c(\"bundle_sheath\",\"mesophyll\",\"mesophyll,bundle_sheath\")]), # convert to p-val\n",
    "  10^(-1 * assay(check_res_type, \"negLog10P\")[,c(\"bundle_sheath\",\"mesophyll\", \"mesophyll,bundle_sheath\")]) # convert to p-val\n",
    ") \n",
    "colnames(ora_res2_consider_type)[3:11] <- c(\"OR.bs\", \"OR.ms\", \"OR.ms_bs\", \n",
    "                             \"fdr.bs\", \"fdr.ms\", \"fdr.ms_bs\", \n",
    "                             \"pval.bs\", \"pval.ms\", \"pval.ms_bs\")\n",
    "\n",
    "as_tibble(ora_res2_consider_type) %>% \n",
    "    arrange(OR.bs, OR.ms) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b79b474-d497-43e6-8bef-006fce0b52fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Trying to Run MonaLisa on just Bs and Ms regions surrounding C4 genes,\n",
    "## Asking if we see bias in some way shape or form. Clearly the signal gets swamped out otherwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "cfc86c75-b4f6-46ac-a1a8-d706a22db278",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'None'</li><li>'bundle_sheath'</li><li>'mesophyll'</li><li>'mesophyll,bundle_sheath'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'None'\n",
       "\\item 'bundle\\_sheath'\n",
       "\\item 'mesophyll'\n",
       "\\item 'mesophyll,bundle\\_sheath'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'None'\n",
       "2. 'bundle_sheath'\n",
       "3. 'mesophyll'\n",
       "4. 'mesophyll,bundle_sheath'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"None\"                    \"bundle_sheath\"          \n",
       "[3] \"mesophyll\"               \"mesophyll,bundle_sheath\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(zm_acrs_annotated$type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f5ce9a3f-db3e-4229-9911-e70c80e9091d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hasNs\n",
       "FALSE   Sum \n",
       "47756 47756 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Using MonaLisa for testing Motif Enrichment\n",
    "zm_acrs_in_c4 <- as_tibble(as.data.frame(zm_acrs)) %>% \n",
    "    tidyr::separate(acr_id, into = c(\"acr\", \"acr_cell_type_specific_class\"), sep = \";\") %>%\n",
    "    dplyr::mutate(acr_cell_type_specific_class = case_when(acr_cell_type_specific_class == \"bundle_sheath,procambial_meristem\" ~ \"bundle_sheath\",\n",
    "                                                          acr_cell_type_specific_class == \"bundle_sheath,procambium\" ~ \"bundle_sheath\", \n",
    "                                                          TRUE ~ acr_cell_type_specific_class)) %>% \n",
    "    left_join(., combined_acrs_marker_info_subset, by = c(\"acr\", \"acr_cell_type_specific_class\")) %>% \n",
    "    mutate(locus = case_when(is.na(locus) == TRUE ~ \"None\",\n",
    "                            TRUE ~ locus),\n",
    "           gene_family = case_when(is.na(gene_family) == TRUE ~ \"None\",\n",
    "                            TRUE ~ locus),\n",
    "           type = case_when(is.na(type) == TRUE ~ \"None\",\n",
    "                            TRUE ~ type),\n",
    "          ) %>% \n",
    "    dplyr::filter(acr_cell_type_specific_class %in% c(\"bundle_sheath\", \"broadly_accessible\", \"mesophyll\")) %>% \n",
    "    dplyr::filter(type != \"None\")\n",
    "\n",
    "zm_acrs_in_c4_gr <- GRanges(zm_acrs_in_c4, seqinfo=seqinfo(BSgenome.Zm_B73))\n",
    "names(zm_acrs_in_c4_gr) <- as.character(zm_acrs_in_c4_gr)\n",
    "zm_acrs_in_c4_gr <- unique(zm_peaks_gr_annotated)\n",
    "\n",
    "\n",
    "# get peak sequences, we'll need this later\n",
    "zm_acrs_in_c4_sequences <- getSeq(BSgenome.Zm_B73, zm_acrs_in_c4_gr) # <2s\n",
    "\n",
    "\n",
    "# exclude peaks with any N's, to avoid warnings further down\n",
    "hasNs    <- (alphabetFrequency(zm_acrs_in_c4_sequences)[,\"N\"]>0)\n",
    "addmargins(table(hasNs)) # => only drops 5 sequences\n",
    "\n",
    "zm_acrs_in_c4_sequences <- zm_acrs_in_c4_sequences[!hasNs]\n",
    "zm_acrs_in_c4_sequences  <- zm_acrs_in_c4_sequences[!hasNs]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "cacfec68-74f3-47d2-905c-e9807b69fabd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       " 36.599   8.800  20.165 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(monaLisa)\n",
    "\n",
    "# now run monaLisa\n",
    "system.time( check_motif_c4_type <- calcBinnedMotifEnrR(\n",
    "  zm_acrs_in_c4_sequences,\n",
    "  bins = zm_acrs_in_c4_gr$type,\n",
    "  pwmL = pw_matrix_list_converted,\n",
    "  min.score = 3,\n",
    "  BPPARAM =  BiocParallel::MulticoreParam(16)\n",
    ") ) # ~40s on my laptop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "7a91a4f8-e752-436f-9db8-360610a6c8ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 28 rows and 8 columns\n",
       "                 motif.id   motif.name     OR.bs     OR.ms    fdr.bs    fdr.ms\n",
       "              <character>  <character> <numeric> <numeric> <numeric> <numeric>\n",
       "ARF                   ARF          ARF  1.000106  1.000297         1         1\n",
       "AS2LOB             AS2LOB       AS2LOB  0.936416  0.928427         1         1\n",
       "BBRBPC             BBRBPC       BBRBPC  1.036072  0.974336         1         1\n",
       "BES1                 BES1         BES1  1.007741  1.038767         1         1\n",
       "bHLH_TCP         bHLH_TCP     bHLH_TCP  0.974697  1.021001         1         1\n",
       "...                   ...          ...       ...       ...       ...       ...\n",
       "bs_GCCGCC       bs_GCCGCC    bs_GCCGCC  0.967986  1.039354         1         1\n",
       "ms_GACGA         ms_GACGA     ms_GACGA  1.002611  1.004775         1         1\n",
       "ms_AGCTA         ms_AGCTA     ms_AGCTA  0.962549  1.023548         1         1\n",
       "ms_GACA           ms_GACA      ms_GACA  1.045623  1.094194         1         1\n",
       "ms_CGTCGTCGT ms_CGTCGTCGT ms_CGTCGTCGT  1.033308  0.998082         1         1\n",
       "               pval.bs   pval.ms\n",
       "             <numeric> <numeric>\n",
       "ARF           0.994481  0.980902\n",
       "AS2LOB        0.974174  0.983047\n",
       "BBRBPC        0.299080  0.802808\n",
       "BES1          0.540117  0.324777\n",
       "bHLH_TCP      0.963674  0.258656\n",
       "...                ...       ...\n",
       "bs_GCCGCC     0.713939  0.365877\n",
       "ms_GACGA      0.875441  0.733948\n",
       "ms_AGCTA      0.809516  0.406694\n",
       "ms_GACA       0.292974  0.055616\n",
       "ms_CGTCGTCGT  0.404109  0.573549"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# reformat the results for convenience\n",
    "c4_type_only_ora <- cbind(\n",
    "  rowData(check_motif_c4_type)[,c(\"motif.id\",\"motif.name\")],\n",
    "  2^assay(check_motif_c4_type, \"log2enr\")[,c(\"bundle_sheath\",\"mesophyll\")],  # convert to OR\n",
    "  10^(-1 * assay(check_motif_c4_type, \"negLog10Padj\")[,c(\"bundle_sheath\",\"mesophyll\")]), # convert to p-val\n",
    "  10^(-1 * assay(check_motif_c4_type, \"negLog10P\")[,c(\"bundle_sheath\",\"mesophyll\")]) # convert to p-val\n",
    ") \n",
    "colnames(c4_type_only_ora)[3:8] <- c(\"OR.bs\", \"OR.ms\", \n",
    "                             \"fdr.bs\", \"fdr.ms\", \n",
    "                             \"pval.bs\", \"pval.ms\")\n",
    "\n",
    "c4_type_only_ora"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f64ef4a3-cd34-4e1d-8f22-d131387da3f6",
   "metadata": {},
   "source": [
    "### Running on Cell type Specific ACRs around C4 Genes Only, and some Braodly ACC ACRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "c8d75531-42aa-427d-b3b7-2a807f1ec71c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hasNs\n",
       "FALSE   Sum \n",
       "  537   537 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Using MonaLisa for testing Motif Enrichment\n",
    "zm_acrs_in_c4 <- zm_acrs_annotated %>% \n",
    "    dplyr::filter(type != \"None\" & acr_cell_type_specific_class != \"broadly_accessible\") %>% \n",
    "    dplyr::filter(acr_cell_type_specific_class %in% c(\"bundle_sheath\", \"broadly_accessible\", \"mesophyll\"))  %>% \n",
    "    dplyr::filter(type == acr_cell_type_specific_class )\n",
    "\n",
    "# zm_acrs_in_c4 %>% \n",
    "#     dplyr::group_by(type, acr_cell_type_specific_class )%>% \n",
    "#     summarise(counts = n())\n",
    "\n",
    "all_null_acrs <- zm_acrs_annotated %>% \n",
    "    dplyr::filter(type == \"None\" & acr_cell_type_specific_class == \"broadly_accessible\") %>% \n",
    "    sample_n(500)\n",
    "\n",
    "combined_vals <- bind_rows(zm_acrs_in_c4,all_null_acrs) %>% \n",
    "  mutate_at(vars(type, acr_cell_type_specific_class), factor)\n",
    "\n",
    "x <- GRanges(combined_vals, seqinfo=seqinfo(BSgenome.Zm_B73))\n",
    "names(x) <- as.character(x)\n",
    "x <- unique(x)\n",
    "\n",
    "# Convert 'type' to a factor\n",
    "mcols(x)$acr_cell_type_specific_class <- factor(mcols(x)$acr_cell_type_specific_class)\n",
    "mcols(x)$type <- factor(mcols(x)$type)\n",
    "\n",
    "\n",
    "# get peak sequences, we'll need this later\n",
    "zm_acrs_in_c4_sequences <- getSeq(BSgenome.Zm_B73, x) # <2s\n",
    "\n",
    "\n",
    "# exclude peaks with any N's, to avoid warnings further down\n",
    "hasNs    <- (alphabetFrequency(zm_acrs_in_c4_sequences)[,\"N\"]>0)\n",
    "addmargins(table(hasNs)) # => only drops 5 sequences\n",
    "\n",
    "zm_acrs_in_c4_sequences <- zm_acrs_in_c4_sequences[!hasNs]\n",
    "zm_acrs_in_c4_sequences  <- zm_acrs_in_c4_sequences[!hasNs]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "7b04b197-c390-4377-8b2a-cb20c04604d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       "  3.089   2.315   1.552 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(monaLisa)\n",
    "x$acr_cell_type_specific_class <- setZeroBin(factor(x$acr_cell_type_specific_class), \"broadly_accessible\")\n",
    "\n",
    "# now run monaLisa\n",
    "system.time( check_motif_c4_type <- calcBinnedMotifEnrR(\n",
    "  zm_acrs_in_c4_sequences,\n",
    "  bins = x$acr_cell_type_specific_class,\n",
    "  pwmL = pw_matrix_list_converted,\n",
    "  min.score = 3,\n",
    "  background = \"otherBins\", \n",
    "\n",
    "  BPPARAM =  BiocParallel::MulticoreParam(16)\n",
    ") ) # ~40s on my laptop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "6ef404d6-dfb6-4747-b25b-c7b0b214659b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformat the results for convenience\n",
    "c4_type_only_ora <- cbind(\n",
    "  rowData(check_motif_c4_type)[,c(\"motif.id\",\"motif.name\")],\n",
    "  2^assay(check_motif_c4_type, \"log2enr\")[,c(\"bundle_sheath\",\"mesophyll\")],  # convert to OR\n",
    "  10^(-1 * assay(check_motif_c4_type, \"negLog10Padj\")[,c(\"bundle_sheath\",\"mesophyll\")]), # convert to p-val\n",
    "  10^(-1 * assay(check_motif_c4_type, \"negLog10P\")[,c(\"bundle_sheath\",\"mesophyll\")]) # convert to p-val\n",
    ") \n",
    "colnames(c4_type_only_ora)[3:8] <- c(\"OR.bs\", \"OR.ms\", \n",
    "                             \"fdr.bs\", \"fdr.ms\", \n",
    "                             \"pval.bs\", \"pval.ms\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "7a7c8781-838b-4707-975a-51944d491946",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 1 row and 8 columns\n",
       "           motif.id  motif.name     OR.bs     OR.ms    fdr.bs    fdr.ms\n",
       "        <character> <character> <numeric> <numeric> <numeric> <numeric>\n",
       "ms_GACA     ms_GACA     ms_GACA   1.14611   1.08443         1         1\n",
       "          pval.bs   pval.ms\n",
       "        <numeric> <numeric>\n",
       "ms_GACA 0.0586323  0.232239"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "DataFrame with 1 row and 8 columns\n",
       "       motif.id  motif.name     OR.bs     OR.ms    fdr.bs    fdr.ms   pval.bs\n",
       "    <character> <character> <numeric> <numeric> <numeric> <numeric> <numeric>\n",
       "HSF         HSF         HSF    0.8349   1.18133         1         1  0.970034\n",
       "      pval.ms\n",
       "    <numeric>\n",
       "HSF 0.0688643"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "c4_type_only_ora[c4_type_only_ora$pval.bs < .1, ]\n",
    "c4_type_only_ora[c4_type_only_ora$pval.ms < .1, ]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
