{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f5a7a57-51ec-40e4-ae2d-d676047867fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(devtools)\n",
    "library(tidyverse)\n",
    "library(cicero)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5f3795ac-3e00-4d0d-9851-d1a3798ab146",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m65990\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): chrom, acr_class, pval\n",
      "\u001b[32mdbl\u001b[39m (2): start, stop\n",
      "\n",
      "\u001b[36mi\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mi\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "peaks_intersections <- read.table(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/02.peak_accessability/pm.peaks_accessability.txt\")\n",
    "all_peaks <- read_delim(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/03.peak_assignment/pm_dir/pm_acr_classification.no_exons.all_ACRs.classified.sorted.bed\", delim = \"\\t\",\n",
    "                          col_names = c(\"chrom\", \"start\", \"stop\", \"acr_class\", \"pval\"))\n",
    "\n",
    "read_meta_data <- read.delim(\"/home/jpm73279/Mendieta_et_al_comparative_single_cell/metrics/annotations/pm_annot/Pm.leaf_annotation.V1.meta.txt\") \n",
    "genome <- read.table(\"/home/jpm73279/genome_downloads/Pmiliaceum/pm_genmap_output.chrom.sizes\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "generate_cicero_format <- all_peaks %>% \n",
    "    dplyr::mutate(combined_vals = str_c(chrom, start,stop, sep = \"_\")) %>% \n",
    "    tidyr::separate(acr_class, c(\"ACR\", \"category\"), \";\")\n",
    "\n",
    "\n",
    "final_cicero_format <- left_join(peaks_intersections, generate_cicero_format, by = c(\"V1\" = \"ACR\")) %>% \n",
    "              dplyr::select(combined_vals, V2, V3)\n",
    "\n",
    "\n",
    "reduced_dim <- read_meta_data %>% \n",
    "    dplyr::select(cellID, umap1, umap2) %>% \n",
    "    dplyr::rename(umap_coord1 = umap1) %>% \n",
    "    dplyr::rename(umap_coord2 = umap2)\n",
    "\n",
    "row.names(reduced_dim) <- reduced_dim$cellID\n",
    "reduced_dim <- reduced_dim[-1]\n",
    "\n",
    "final_cicero_format_filtered <- final_cicero_format %>% \n",
    "    dplyr::filter(V2 %in% read_meta_data$cellID)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "90c48518-ec9e-4837-8c21-afb312ed1bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/scratch/jpm73279/comparative_single_cell/dev_location/cicero\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b4fe90f-18df-4df0-a391-d9b9c5366fd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overlap QC metrics:\n",
      "Cells per bin: 50\n",
      "Maximum shared cells bin-bin: 44\n",
      "Mean shared cells bin-bin: 4.86847990765457\n",
      "Median shared cells bin-bin: 0\n",
      "\n",
      "Warning message in if (isSparseMatrix(counts)) {:\n",
      "\"the condition has length > 1 and only the first element will be used\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Starting Cicero\"\n",
      "[1] \"Calculating distance_parameter value\"\n",
      "[1] \"Running models\"\n",
      "[1] \"Assembling connections\"\n",
      "[1] \"Successful cicero models:  3130\"\n",
      "[1] \"Other models: \"\n",
      "\n",
      "Zero or one element in range \n",
      "                        1458 \n",
      "[1] \"Models with errors:  0\"\n",
      "[1] \"Done\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overlap QC metrics:\n",
      "Cells per bin: 50\n",
      "Maximum shared cells bin-bin: 44\n",
      "Mean shared cells bin-bin: 6.41020361473347\n",
      "Median shared cells bin-bin: 0\n",
      "\n",
      "Warning message in make_cicero_cds(input_cds, reduced_coordinates = reduced_dim):\n",
      "\"On average, more than 10% of cells are shared between paired bins.\"\n",
      "Warning message in if (isSparseMatrix(counts)) {:\n",
      "\"the condition has length > 1 and only the first element will be used\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Starting Cicero\"\n",
      "[1] \"Calculating distance_parameter value\"\n",
      "[1] \"Running models\"\n",
      "[1] \"Assembling connections\"\n",
      "[1] \"Successful cicero models:  3127\"\n",
      "[1] \"Other models: \"\n",
      "\n",
      "Zero or one element in range \n",
      "                        1461 \n",
      "[1] \"Models with errors:  0\"\n",
      "[1] \"Done\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overlap QC metrics:\n",
      "Cells per bin: 50\n",
      "Maximum shared cells bin-bin: 44\n",
      "Mean shared cells bin-bin: 2.00590863236778\n",
      "Median shared cells bin-bin: 0\n",
      "\n",
      "Warning message in if (isSparseMatrix(counts)) {:\n",
      "\"the condition has length > 1 and only the first element will be used\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Starting Cicero\"\n",
      "[1] \"Calculating distance_parameter value\"\n",
      "[1] \"Running models\"\n",
      "[1] \"Assembling connections\"\n",
      "[1] \"Successful cicero models:  3143\"\n",
      "[1] \"Other models: \"\n",
      "\n",
      "Zero or one element in range \n",
      "                        1445 \n",
      "[1] \"Models with errors:  0\"\n",
      "[1] \"Done\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overlap QC metrics:\n",
      "Cells per bin: 50\n",
      "Maximum shared cells bin-bin: 44\n",
      "Mean shared cells bin-bin: 0.89310920040257\n",
      "Median shared cells bin-bin: 0\n",
      "\n",
      "Warning message in if (isSparseMatrix(counts)) {:\n",
      "\"the condition has length > 1 and only the first element will be used\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Starting Cicero\"\n",
      "[1] \"Calculating distance_parameter value\"\n",
      "[1] \"Running models\"\n",
      "[1] \"Assembling connections\"\n",
      "[1] \"Successful cicero models:  3144\"\n",
      "[1] \"Other models: \"\n",
      "\n",
      "Zero or one element in range \n",
      "                        1444 \n",
      "[1] \"Models with errors:  0\"\n",
      "[1] \"Done\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overlap QC metrics:\n",
      "Cells per bin: 50\n",
      "Maximum shared cells bin-bin: 44\n",
      "Mean shared cells bin-bin: 3.02285331322778\n",
      "Median shared cells bin-bin: 0\n",
      "\n",
      "Warning message in if (isSparseMatrix(counts)) {:\n",
      "\"the condition has length > 1 and only the first element will be used\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Starting Cicero\"\n",
      "[1] \"Calculating distance_parameter value\"\n",
      "[1] \"Running models\"\n",
      "[1] \"Assembling connections\"\n",
      "[1] \"Successful cicero models:  3138\"\n",
      "[1] \"Other models: \"\n",
      "\n",
      "Zero or one element in range \n",
      "                        1450 \n",
      "[1] \"Models with errors:  0\"\n",
      "[1] \"Done\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overlap QC metrics:\n",
      "Cells per bin: 50\n",
      "Maximum shared cells bin-bin: 44\n",
      "Mean shared cells bin-bin: 2.87087134154624\n",
      "Median shared cells bin-bin: 0\n",
      "\n",
      "Warning message in if (isSparseMatrix(counts)) {:\n",
      "\"the condition has length > 1 and only the first element will be used\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Starting Cicero\"\n",
      "[1] \"Calculating distance_parameter value\"\n",
      "[1] \"Running models\"\n",
      "[1] \"Assembling connections\"\n",
      "[1] \"Successful cicero models:  3143\"\n",
      "[1] \"Other models: \"\n",
      "\n",
      "Zero or one element in range \n",
      "                        1445 \n",
      "[1] \"Models with errors:  0\"\n",
      "[1] \"Done\"\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for(COL_SLOT_ID in unique(read_meta_data$reduce_resolution_annotation)) {\n",
    "  \n",
    "  # Subset meta data based on the unique value of COLUMN_NAME\n",
    "  subset_meta_data <- read_meta_data[read_meta_data$reduce_resolution_annotation == COL_SLOT_ID, ]\n",
    "  print(COL_SLOT_ID)\n",
    "  # Update reduced_dim based on subset\n",
    "  reduced_dim <- subset_meta_data %>% \n",
    "    dplyr::select(cellID, umap1, umap2) %>% \n",
    "    dplyr::rename(umap_coord1 = umap1) %>% \n",
    "    dplyr::rename(umap_coord2 = umap2)\n",
    "\n",
    "  row.names(reduced_dim) <- reduced_dim$cellID\n",
    "  reduced_dim <- reduced_dim[-1]\n",
    "  \n",
    "  # Update final_cicero_format_filtered based on subset\n",
    "  final_cicero_format_filtered <- final_cicero_format %>% \n",
    "    dplyr::filter(V2 %in% subset_meta_data$cellID)\n",
    "  \n",
    "  input_cds <- make_atac_cds(final_cicero_format_filtered, binarize = TRUE)\n",
    "  \n",
    "  # Make cicero CDS\n",
    "  cicero_cds <- make_cicero_cds(input_cds, reduced_coordinates = reduced_dim)\n",
    "  \n",
    "  # Run cicero and save output\n",
    "  running_cic <- run_cicero(cicero_cds, genome)\n",
    "  write.csv(running_cic, paste0(\"pm.cicero.\", COL_SLOT_ID, \".output.csv\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "583f5d5a-21c1-40a9-8f17-13209bcc2166",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for plot_connections {cicero}\"><tr><td>plot_connections {cicero}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Plot connections</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Plotting function for Cicero connections. Uses <code>plotTracks</code>\n",
       "as its basis\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "plot_connections(\n",
       "  connection_df,\n",
       "  chr,\n",
       "  minbp,\n",
       "  maxbp,\n",
       "  coaccess_cutoff = 0,\n",
       "  peak_color = \"#B4656F\",\n",
       "  connection_color = \"#7F7CAF\",\n",
       "  connection_color_legend = TRUE,\n",
       "  alpha_by_coaccess = FALSE,\n",
       "  connection_width = 2,\n",
       "  connection_ymax = NULL,\n",
       "  gene_model = NULL,\n",
       "  gene_model_color = \"#81D2C7\",\n",
       "  gene_model_shape = c(\"smallArrow\", \"box\"),\n",
       "  comparison_track = NULL,\n",
       "  comparison_coaccess_cutoff = 0,\n",
       "  comparison_peak_color = \"#B4656F\",\n",
       "  comparison_connection_color = \"#7F7CAF\",\n",
       "  comparison_connection_color_legend = TRUE,\n",
       "  comparison_connection_width = 2,\n",
       "  comparison_ymax = NULL,\n",
       "  collapseTranscripts = FALSE,\n",
       "  include_axis_track = TRUE,\n",
       "  return_as_list = FALSE,\n",
       "  viewpoint = NULL,\n",
       "  comparison_viewpoint = TRUE,\n",
       "  viewpoint_color = \"#F0544F\",\n",
       "  viewpoint_fill = \"#EFD8D7\",\n",
       "  viewpoint_alpha = 0.5\n",
       ")\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>connection_df</code></td>\n",
       "<td>\n",
       "<p>Data frame of connections, which must include the\n",
       "columns 'Peak1', 'Peak2', and 'coaccess'. Generally, the output of\n",
       "run_cicero or assemble_connections.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>chr</code></td>\n",
       "<td>\n",
       "<p>The chromosome of the region you would like to plot in the form\n",
       "'chr10'.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>minbp</code></td>\n",
       "<td>\n",
       "<p>The base pair coordinate of the start of the region to be\n",
       "plotted.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>maxbp</code></td>\n",
       "<td>\n",
       "<p>The base pair coordinate of the end of the region to be plotted.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>coaccess_cutoff</code></td>\n",
       "<td>\n",
       "<p>The minimum cicero co-accessibility score you would\n",
       "like to be plotted. Default is 0.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>peak_color</code></td>\n",
       "<td>\n",
       "<p>Color for peak annotations - a single color, the name of a\n",
       "column containing color values that correspond to Peak1, or the name of\n",
       "column containing a character or factor to base peak colors on.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>connection_color</code></td>\n",
       "<td>\n",
       "<p>Color for connection lines. A single color, the name\n",
       "of a column containing color values, or the name of a column containing a\n",
       "character or factor to base connection colors on.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>connection_color_legend</code></td>\n",
       "<td>\n",
       "<p>Logical, should connection color legend be\n",
       "shown?</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>alpha_by_coaccess</code></td>\n",
       "<td>\n",
       "<p>Logical, should the transparency of connection\n",
       "lines be scaled based on co-accessibility score?</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>connection_width</code></td>\n",
       "<td>\n",
       "<p>Width of connection lines.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>connection_ymax</code></td>\n",
       "<td>\n",
       "<p>Connection y-axis height. If <code>NULL</code>, chosen\n",
       "automatically.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene_model</code></td>\n",
       "<td>\n",
       "<p>Either <code>NULL</code> or a data.frame. The data.frame should\n",
       "be in a form compatible with the Gviz function\n",
       "<code>GeneRegionTrack-class</code> (cannot have NA as column names).</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene_model_color</code></td>\n",
       "<td>\n",
       "<p>Color for gene annotations.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene_model_shape</code></td>\n",
       "<td>\n",
       "<p>Shape for gene models, passed to\n",
       "<code>GeneRegionTrack-class</code>. Options described at\n",
       "<code>GeneRegionTrack-class</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_track</code></td>\n",
       "<td>\n",
       "<p>Either <code>NULL</code> or a data frame. If a data frame,\n",
       "a second track of connections will be plotted based on this data. This\n",
       "data frame has the same requirements as connection_df (Peak1, Peak2 and\n",
       "coaccess columns).</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_coaccess_cutoff</code></td>\n",
       "<td>\n",
       "<p>The minimum cicero co-accessibility score\n",
       "you would like to be plotted for the comparison dataset. Default = 0.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_peak_color</code></td>\n",
       "<td>\n",
       "<p>Color for comparison peak annotations - a\n",
       "single color, the name of a column containing color values that correspond\n",
       "to Peak1, or the name of a column containing a character or factor to base\n",
       "peak colors on.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_connection_color</code></td>\n",
       "<td>\n",
       "<p>Color for comparison connection lines. A\n",
       "single color, the name of a column containing color values, or the name of\n",
       "a column containing a character or factor to base connection colors on.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_connection_color_legend</code></td>\n",
       "<td>\n",
       "<p>Logical, should comparison\n",
       "connection color legend be shown?</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_connection_width</code></td>\n",
       "<td>\n",
       "<p>Width of comparison connection lines.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_ymax</code></td>\n",
       "<td>\n",
       "<p>Connection y-axis height for comparison track. If\n",
       "<code>NULL</code>, chosen automatically.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>collapseTranscripts</code></td>\n",
       "<td>\n",
       "<p>Logical or character scalar. Can be one in\n",
       "<code>gene</code>, <code>longest</code>, <code>shortest</code> or <code>meta</code>. Variable is\n",
       "passed to the <code>GeneRegionTrack-class</code> function of Gviz.\n",
       "Determines whether and how to collapse related transcripts. See Gviz\n",
       "documentation for details.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>include_axis_track</code></td>\n",
       "<td>\n",
       "<p>Logical, should a genomic axis be plotted?</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>return_as_list</code></td>\n",
       "<td>\n",
       "<p>Logical, if TRUE, the function will not plot, but will\n",
       "return the plot components as a list. Allows user to add/customize Gviz\n",
       "components and plot them separately using <code>plotTracks</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>viewpoint</code></td>\n",
       "<td>\n",
       "<p><code>NULL</code> or Coordinates in form &quot;chr1_10000_10020&quot;. Use\n",
       "viewpoint if you would like to plot cicero connections &quot;4C-seq style&quot;.\n",
       "Only connections originating in the viewpoint will be shown. Ideal for\n",
       "comparisons with 4C-seq data. If comparison_viewpoint is <code>TRUE</code>, any\n",
       "comparison track will be subsetted as well.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comparison_viewpoint</code></td>\n",
       "<td>\n",
       "<p>Logical, should viewpoint apply to comparison\n",
       "track as well?</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>viewpoint_color</code></td>\n",
       "<td>\n",
       "<p>Color for the highlight border.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>viewpoint_fill</code></td>\n",
       "<td>\n",
       "<p>Color for the highlight fill.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>viewpoint_alpha</code></td>\n",
       "<td>\n",
       "<p>Alpha value for the highlight fill.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A gene region plot, or list of components if return_as_list is\n",
       "<code>TRUE</code>.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "  cicero_cons &lt;- data.frame(\n",
       "             Peak1 = c(\"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10087586_10087901\", \"chr18_10120685_10127115\",\n",
       "                       \"chr18_10097718_10097934\", \"chr18_10087586_10087901\",\n",
       "                       \"chr18_10154818_10155215\", \"chr18_10238762_10238983\",\n",
       "                       \"chr18_10198959_10199183\", \"chr18_10250985_10251585\"),\n",
       "             Peak2 = c(\"chr18_10097718_10097934\", \"chr18_10087586_10087901\",\n",
       "                       \"chr18_10154818_10155215\", \"chr18_10238762_10238983\",\n",
       "                       \"chr18_10198959_10199183\", \"chr18_10250985_10251585\",\n",
       "                       \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10087586_10087901\", \"chr18_10120685_10127115\"),\n",
       "             coaccess = c(0.0051121787, 0.0016698617, 0.0006570246,\n",
       "                          0.0013466927, 0.0737935011, 0.3264019452,\n",
       "                          0.0051121787, 0.0016698617, 0.0006570246,\n",
       "                          0.0013466927, 0.0737935011, 0.3264019452))\n",
       "  plot_connections(cicero_cons, chr = \"chr18\",\n",
       "                   minbp = 10034652,\n",
       "                   maxbp = 10251585,\n",
       "                   peak_color = \"purple\")\n",
       "\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>cicero</em> version 1.12.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{plot\\_connections}{Plot connections}{plot.Rul.connections}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Plotting function for Cicero connections. Uses \\code{\\LinkA{plotTracks}{plotTracks}}\n",
       "as its basis\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "plot_connections(\n",
       "  connection_df,\n",
       "  chr,\n",
       "  minbp,\n",
       "  maxbp,\n",
       "  coaccess_cutoff = 0,\n",
       "  peak_color = \"#B4656F\",\n",
       "  connection_color = \"#7F7CAF\",\n",
       "  connection_color_legend = TRUE,\n",
       "  alpha_by_coaccess = FALSE,\n",
       "  connection_width = 2,\n",
       "  connection_ymax = NULL,\n",
       "  gene_model = NULL,\n",
       "  gene_model_color = \"#81D2C7\",\n",
       "  gene_model_shape = c(\"smallArrow\", \"box\"),\n",
       "  comparison_track = NULL,\n",
       "  comparison_coaccess_cutoff = 0,\n",
       "  comparison_peak_color = \"#B4656F\",\n",
       "  comparison_connection_color = \"#7F7CAF\",\n",
       "  comparison_connection_color_legend = TRUE,\n",
       "  comparison_connection_width = 2,\n",
       "  comparison_ymax = NULL,\n",
       "  collapseTranscripts = FALSE,\n",
       "  include_axis_track = TRUE,\n",
       "  return_as_list = FALSE,\n",
       "  viewpoint = NULL,\n",
       "  comparison_viewpoint = TRUE,\n",
       "  viewpoint_color = \"#F0544F\",\n",
       "  viewpoint_fill = \"#EFD8D7\",\n",
       "  viewpoint_alpha = 0.5\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{connection\\_df}] Data frame of connections, which must include the\n",
       "columns 'Peak1', 'Peak2', and 'coaccess'. Generally, the output of\n",
       "run\\_cicero or assemble\\_connections.\n",
       "\n",
       "\\item[\\code{chr}] The chromosome of the region you would like to plot in the form\n",
       "'chr10'.\n",
       "\n",
       "\\item[\\code{minbp}] The base pair coordinate of the start of the region to be\n",
       "plotted.\n",
       "\n",
       "\\item[\\code{maxbp}] The base pair coordinate of the end of the region to be plotted.\n",
       "\n",
       "\\item[\\code{coaccess\\_cutoff}] The minimum cicero co-accessibility score you would\n",
       "like to be plotted. Default is 0.\n",
       "\n",
       "\\item[\\code{peak\\_color}] Color for peak annotations - a single color, the name of a\n",
       "column containing color values that correspond to Peak1, or the name of\n",
       "column containing a character or factor to base peak colors on.\n",
       "\n",
       "\\item[\\code{connection\\_color}] Color for connection lines. A single color, the name\n",
       "of a column containing color values, or the name of a column containing a\n",
       "character or factor to base connection colors on.\n",
       "\n",
       "\\item[\\code{connection\\_color\\_legend}] Logical, should connection color legend be\n",
       "shown?\n",
       "\n",
       "\\item[\\code{alpha\\_by\\_coaccess}] Logical, should the transparency of connection\n",
       "lines be scaled based on co-accessibility score?\n",
       "\n",
       "\\item[\\code{connection\\_width}] Width of connection lines.\n",
       "\n",
       "\\item[\\code{connection\\_ymax}] Connection y-axis height. If \\code{NULL}, chosen\n",
       "automatically.\n",
       "\n",
       "\\item[\\code{gene\\_model}] Either \\code{NULL} or a data.frame. The data.frame should\n",
       "be in a form compatible with the Gviz function\n",
       "\\code{\\LinkA{GeneRegionTrack-class}{GeneRegionTrack.Rdash.class}} (cannot have NA as column names).\n",
       "\n",
       "\\item[\\code{gene\\_model\\_color}] Color for gene annotations.\n",
       "\n",
       "\\item[\\code{gene\\_model\\_shape}] Shape for gene models, passed to\n",
       "\\code{\\LinkA{GeneRegionTrack-class}{GeneRegionTrack.Rdash.class}}. Options described at\n",
       "\\code{\\LinkA{GeneRegionTrack-class}{GeneRegionTrack.Rdash.class}}.\n",
       "\n",
       "\\item[\\code{comparison\\_track}] Either \\code{NULL} or a data frame. If a data frame,\n",
       "a second track of connections will be plotted based on this data. This\n",
       "data frame has the same requirements as connection\\_df (Peak1, Peak2 and\n",
       "coaccess columns).\n",
       "\n",
       "\\item[\\code{comparison\\_coaccess\\_cutoff}] The minimum cicero co-accessibility score\n",
       "you would like to be plotted for the comparison dataset. Default = 0.\n",
       "\n",
       "\\item[\\code{comparison\\_peak\\_color}] Color for comparison peak annotations - a\n",
       "single color, the name of a column containing color values that correspond\n",
       "to Peak1, or the name of a column containing a character or factor to base\n",
       "peak colors on.\n",
       "\n",
       "\\item[\\code{comparison\\_connection\\_color}] Color for comparison connection lines. A\n",
       "single color, the name of a column containing color values, or the name of\n",
       "a column containing a character or factor to base connection colors on.\n",
       "\n",
       "\\item[\\code{comparison\\_connection\\_color\\_legend}] Logical, should comparison\n",
       "connection color legend be shown?\n",
       "\n",
       "\\item[\\code{comparison\\_connection\\_width}] Width of comparison connection lines.\n",
       "\n",
       "\\item[\\code{comparison\\_ymax}] Connection y-axis height for comparison track. If\n",
       "\\code{NULL}, chosen automatically.\n",
       "\n",
       "\\item[\\code{collapseTranscripts}] Logical or character scalar. Can be one in\n",
       "\\code{gene}, \\code{longest}, \\code{shortest} or \\code{meta}. Variable is\n",
       "passed to the \\code{\\LinkA{GeneRegionTrack-class}{GeneRegionTrack.Rdash.class}} function of Gviz.\n",
       "Determines whether and how to collapse related transcripts. See Gviz\n",
       "documentation for details.\n",
       "\n",
       "\\item[\\code{include\\_axis\\_track}] Logical, should a genomic axis be plotted?\n",
       "\n",
       "\\item[\\code{return\\_as\\_list}] Logical, if TRUE, the function will not plot, but will\n",
       "return the plot components as a list. Allows user to add/customize Gviz\n",
       "components and plot them separately using \\code{\\LinkA{plotTracks}{plotTracks}}.\n",
       "\n",
       "\\item[\\code{viewpoint}] \\code{NULL} or Coordinates in form \"chr1\\_10000\\_10020\". Use\n",
       "viewpoint if you would like to plot cicero connections \"4C-seq style\".\n",
       "Only connections originating in the viewpoint will be shown. Ideal for\n",
       "comparisons with 4C-seq data. If comparison\\_viewpoint is \\code{TRUE}, any\n",
       "comparison track will be subsetted as well.\n",
       "\n",
       "\\item[\\code{comparison\\_viewpoint}] Logical, should viewpoint apply to comparison\n",
       "track as well?\n",
       "\n",
       "\\item[\\code{viewpoint\\_color}] Color for the highlight border.\n",
       "\n",
       "\\item[\\code{viewpoint\\_fill}] Color for the highlight fill.\n",
       "\n",
       "\\item[\\code{viewpoint\\_alpha}] Alpha value for the highlight fill.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Value}\n",
       "A gene region plot, or list of components if return\\_as\\_list is\n",
       "\\code{TRUE}.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "  cicero_cons <- data.frame(\n",
       "             Peak1 = c(\"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10087586_10087901\", \"chr18_10120685_10127115\",\n",
       "                       \"chr18_10097718_10097934\", \"chr18_10087586_10087901\",\n",
       "                       \"chr18_10154818_10155215\", \"chr18_10238762_10238983\",\n",
       "                       \"chr18_10198959_10199183\", \"chr18_10250985_10251585\"),\n",
       "             Peak2 = c(\"chr18_10097718_10097934\", \"chr18_10087586_10087901\",\n",
       "                       \"chr18_10154818_10155215\", \"chr18_10238762_10238983\",\n",
       "                       \"chr18_10198959_10199183\", \"chr18_10250985_10251585\",\n",
       "                       \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                       \"chr18_10087586_10087901\", \"chr18_10120685_10127115\"),\n",
       "             coaccess = c(0.0051121787, 0.0016698617, 0.0006570246,\n",
       "                          0.0013466927, 0.0737935011, 0.3264019452,\n",
       "                          0.0051121787, 0.0016698617, 0.0006570246,\n",
       "                          0.0013466927, 0.0737935011, 0.3264019452))\n",
       "  plot_connections(cicero_cons, chr = \"chr18\",\n",
       "                   minbp = 10034652,\n",
       "                   maxbp = 10251585,\n",
       "                   peak_color = \"purple\")\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "plot_connections            package:cicero             R Documentation\n",
       "\n",
       "_\bP_\bl_\bo_\bt _\bc_\bo_\bn_\bn_\be_\bc_\bt_\bi_\bo_\bn_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Plotting function for Cicero connections. Uses 'plotTracks' as its\n",
       "     basis\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     plot_connections(\n",
       "       connection_df,\n",
       "       chr,\n",
       "       minbp,\n",
       "       maxbp,\n",
       "       coaccess_cutoff = 0,\n",
       "       peak_color = \"#B4656F\",\n",
       "       connection_color = \"#7F7CAF\",\n",
       "       connection_color_legend = TRUE,\n",
       "       alpha_by_coaccess = FALSE,\n",
       "       connection_width = 2,\n",
       "       connection_ymax = NULL,\n",
       "       gene_model = NULL,\n",
       "       gene_model_color = \"#81D2C7\",\n",
       "       gene_model_shape = c(\"smallArrow\", \"box\"),\n",
       "       comparison_track = NULL,\n",
       "       comparison_coaccess_cutoff = 0,\n",
       "       comparison_peak_color = \"#B4656F\",\n",
       "       comparison_connection_color = \"#7F7CAF\",\n",
       "       comparison_connection_color_legend = TRUE,\n",
       "       comparison_connection_width = 2,\n",
       "       comparison_ymax = NULL,\n",
       "       collapseTranscripts = FALSE,\n",
       "       include_axis_track = TRUE,\n",
       "       return_as_list = FALSE,\n",
       "       viewpoint = NULL,\n",
       "       comparison_viewpoint = TRUE,\n",
       "       viewpoint_color = \"#F0544F\",\n",
       "       viewpoint_fill = \"#EFD8D7\",\n",
       "       viewpoint_alpha = 0.5\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "connection_df: Data frame of connections, which must include the\n",
       "          columns 'Peak1', 'Peak2', and 'coaccess'. Generally, the\n",
       "          output of run_cicero or assemble_connections.\n",
       "\n",
       "     chr: The chromosome of the region you would like to plot in the\n",
       "          form 'chr10'.\n",
       "\n",
       "   minbp: The base pair coordinate of the start of the region to be\n",
       "          plotted.\n",
       "\n",
       "   maxbp: The base pair coordinate of the end of the region to be\n",
       "          plotted.\n",
       "\n",
       "coaccess_cutoff: The minimum cicero co-accessibility score you would\n",
       "          like to be plotted. Default is 0.\n",
       "\n",
       "peak_color: Color for peak annotations - a single color, the name of a\n",
       "          column containing color values that correspond to Peak1, or\n",
       "          the name of column containing a character or factor to base\n",
       "          peak colors on.\n",
       "\n",
       "connection_color: Color for connection lines. A single color, the name\n",
       "          of a column containing color values, or the name of a column\n",
       "          containing a character or factor to base connection colors\n",
       "          on.\n",
       "\n",
       "connection_color_legend: Logical, should connection color legend be\n",
       "          shown?\n",
       "\n",
       "alpha_by_coaccess: Logical, should the transparency of connection lines\n",
       "          be scaled based on co-accessibility score?\n",
       "\n",
       "connection_width: Width of connection lines.\n",
       "\n",
       "connection_ymax: Connection y-axis height. If 'NULL', chosen\n",
       "          automatically.\n",
       "\n",
       "gene_model: Either 'NULL' or a data.frame. The data.frame should be in\n",
       "          a form compatible with the Gviz function\n",
       "          'GeneRegionTrack-class' (cannot have NA as column names).\n",
       "\n",
       "gene_model_color: Color for gene annotations.\n",
       "\n",
       "gene_model_shape: Shape for gene models, passed to\n",
       "          'GeneRegionTrack-class'. Options described at\n",
       "          'GeneRegionTrack-class'.\n",
       "\n",
       "comparison_track: Either 'NULL' or a data frame. If a data frame, a\n",
       "          second track of connections will be plotted based on this\n",
       "          data. This data frame has the same requirements as\n",
       "          connection_df (Peak1, Peak2 and coaccess columns).\n",
       "\n",
       "comparison_coaccess_cutoff: The minimum cicero co-accessibility score\n",
       "          you would like to be plotted for the comparison dataset.\n",
       "          Default = 0.\n",
       "\n",
       "comparison_peak_color: Color for comparison peak annotations - a single\n",
       "          color, the name of a column containing color values that\n",
       "          correspond to Peak1, or the name of a column containing a\n",
       "          character or factor to base peak colors on.\n",
       "\n",
       "comparison_connection_color: Color for comparison connection lines. A\n",
       "          single color, the name of a column containing color values,\n",
       "          or the name of a column containing a character or factor to\n",
       "          base connection colors on.\n",
       "\n",
       "comparison_connection_color_legend: Logical, should comparison\n",
       "          connection color legend be shown?\n",
       "\n",
       "comparison_connection_width: Width of comparison connection lines.\n",
       "\n",
       "comparison_ymax: Connection y-axis height for comparison track. If\n",
       "          'NULL', chosen automatically.\n",
       "\n",
       "collapseTranscripts: Logical or character scalar. Can be one in 'gene',\n",
       "          'longest', 'shortest' or 'meta'. Variable is passed to the\n",
       "          'GeneRegionTrack-class' function of Gviz. Determines whether\n",
       "          and how to collapse related transcripts. See Gviz\n",
       "          documentation for details.\n",
       "\n",
       "include_axis_track: Logical, should a genomic axis be plotted?\n",
       "\n",
       "return_as_list: Logical, if TRUE, the function will not plot, but will\n",
       "          return the plot components as a list. Allows user to\n",
       "          add/customize Gviz components and plot them separately using\n",
       "          'plotTracks'.\n",
       "\n",
       "viewpoint: 'NULL' or Coordinates in form \"chr1_10000_10020\". Use\n",
       "          viewpoint if you would like to plot cicero connections\n",
       "          \"4C-seq style\". Only connections originating in the viewpoint\n",
       "          will be shown. Ideal for comparisons with 4C-seq data. If\n",
       "          comparison_viewpoint is 'TRUE', any comparison track will be\n",
       "          subsetted as well.\n",
       "\n",
       "comparison_viewpoint: Logical, should viewpoint apply to comparison\n",
       "          track as well?\n",
       "\n",
       "viewpoint_color: Color for the highlight border.\n",
       "\n",
       "viewpoint_fill: Color for the highlight fill.\n",
       "\n",
       "viewpoint_alpha: Alpha value for the highlight fill.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A gene region plot, or list of components if return_as_list is\n",
       "     'TRUE'.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "       cicero_cons <- data.frame(\n",
       "                  Peak1 = c(\"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                            \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                            \"chr18_10087586_10087901\", \"chr18_10120685_10127115\",\n",
       "                            \"chr18_10097718_10097934\", \"chr18_10087586_10087901\",\n",
       "                            \"chr18_10154818_10155215\", \"chr18_10238762_10238983\",\n",
       "                            \"chr18_10198959_10199183\", \"chr18_10250985_10251585\"),\n",
       "                  Peak2 = c(\"chr18_10097718_10097934\", \"chr18_10087586_10087901\",\n",
       "                            \"chr18_10154818_10155215\", \"chr18_10238762_10238983\",\n",
       "                            \"chr18_10198959_10199183\", \"chr18_10250985_10251585\",\n",
       "                            \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                            \"chr18_10034652_10034983\", \"chr18_10034652_10034983\",\n",
       "                            \"chr18_10087586_10087901\", \"chr18_10120685_10127115\"),\n",
       "                  coaccess = c(0.0051121787, 0.0016698617, 0.0006570246,\n",
       "                               0.0013466927, 0.0737935011, 0.3264019452,\n",
       "                               0.0051121787, 0.0016698617, 0.0006570246,\n",
       "                               0.0013466927, 0.0737935011, 0.3264019452))\n",
       "       plot_connections(cicero_cons, chr = \"chr18\",\n",
       "                        minbp = 10034652,\n",
       "                        maxbp = 10251585,\n",
       "                        peak_color = \"purple\")\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?plot_connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "bffb3a47-632c-434c-8018-58528f9c49b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 350114 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Peak1</th><th scope=col>Peak2</th><th scope=col>coaccess</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr10_100241725_100242226</td><td>chr10_100242357_100242858</td><td>0.0906766262</td></tr>\n",
       "\t<tr><td>chr10_100242357_100242858</td><td>chr10_100241725_100242226</td><td>0.0906766262</td></tr>\n",
       "\t<tr><td>chr10_100484245_100484746</td><td>chr10_100485037_100485538</td><td>0.1135825904</td></tr>\n",
       "\t<tr><td>chr10_100485037_100485538</td><td>chr10_100484245_100484746</td><td>0.1135825904</td></tr>\n",
       "\t<tr><td>chr10_10048691_10049192  </td><td>chr10_9768945_9769446    </td><td>0.1164981762</td></tr>\n",
       "\t<tr><td>chr10_10048691_10049192  </td><td>chr10_10140631_10141132  </td><td>0.1547400509</td></tr>\n",
       "\t<tr><td>chr10_10048691_10049192  </td><td>chr10_10146015_10146516  </td><td>0.1702093118</td></tr>\n",
       "\t<tr><td>chr10_10048691_10049192  </td><td>chr10_10236626_10237127  </td><td>0.0288173343</td></tr>\n",
       "\t<tr><td>chr10_10048691_10049192  </td><td>chr10_10241427_10241928  </td><td>0.0188652891</td></tr>\n",
       "\t<tr><td>chr10_1009083_1009584    </td><td>chr10_1463639_1464140    </td><td>0.0005210603</td></tr>\n",
       "\t<tr><td>chr10_101335383_101335884</td><td>chr10_101698337_101698838</td><td>0.0069869046</td></tr>\n",
       "\t<tr><td>chr10_101338451_101338952</td><td>chr10_101506152_101506653</td><td>0.0410098745</td></tr>\n",
       "\t<tr><td>chr10_101338451_101338952</td><td>chr10_101512935_101513436</td><td>0.0157884476</td></tr>\n",
       "\t<tr><td>chr10_10140631_10141132  </td><td>chr10_9768945_9769446    </td><td>0.0180269337</td></tr>\n",
       "\t<tr><td>chr10_10140631_10141132  </td><td>chr10_10048691_10049192  </td><td>0.1547400509</td></tr>\n",
       "\t<tr><td>chr10_10140631_10141132  </td><td>chr10_10146015_10146516  </td><td>0.0263605137</td></tr>\n",
       "\t<tr><td>chr10_10140631_10141132  </td><td>chr10_10236626_10237127  </td><td>0.0044593474</td></tr>\n",
       "\t<tr><td>chr10_10140631_10141132  </td><td>chr10_10241427_10241928  </td><td>0.0029191281</td></tr>\n",
       "\t<tr><td>chr10_10146015_10146516  </td><td>chr10_9768945_9769446    </td><td>0.0198290744</td></tr>\n",
       "\t<tr><td>chr10_10146015_10146516  </td><td>chr10_10048691_10049192  </td><td>0.1702093118</td></tr>\n",
       "\t<tr><td>chr10_10146015_10146516  </td><td>chr10_10140631_10141132  </td><td>0.0263605137</td></tr>\n",
       "\t<tr><td>chr10_10146015_10146516  </td><td>chr10_10236626_10237127  </td><td>0.0049059460</td></tr>\n",
       "\t<tr><td>chr10_10146015_10146516  </td><td>chr10_10241427_10241928  </td><td>0.0032104878</td></tr>\n",
       "\t<tr><td>chr10_101506152_101506653</td><td>chr10_101338451_101338952</td><td>0.0410098745</td></tr>\n",
       "\t<tr><td>chr10_101506152_101506653</td><td>chr10_101512935_101513436</td><td>0.0003237411</td></tr>\n",
       "\t<tr><td>chr10_101512935_101513436</td><td>chr10_101338451_101338952</td><td>0.0157884476</td></tr>\n",
       "\t<tr><td>chr10_101512935_101513436</td><td>chr10_101506152_101506653</td><td>0.0003237411</td></tr>\n",
       "\t<tr><td>chr10_101520326_101520827</td><td>chr10_101521026_101521527</td><td>0.1838308949</td></tr>\n",
       "\t<tr><td>chr10_101520326_101520827</td><td>chr10_101698337_101698838</td><td>0.0082854999</td></tr>\n",
       "\t<tr><td>chr10_101521026_101521527</td><td>chr10_101520326_101520827</td><td>0.1838308949</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chr9_99903993_99904494</td><td>chr9_99715259_99715760</td><td>6.791758e-03</td></tr>\n",
       "\t<tr><td>chr9_99903993_99904494</td><td>chr9_99848477_99848978</td><td>1.026813e-01</td></tr>\n",
       "\t<tr><td>chr9_99903993_99904494</td><td>chr9_99951231_99951732</td><td>1.441456e-03</td></tr>\n",
       "\t<tr><td>chr9_99903993_99904494</td><td>chr9_99952043_99952544</td><td>5.885417e-03</td></tr>\n",
       "\t<tr><td>chr9_99906655_99907156</td><td>chr9_99501516_99502017</td><td>1.672635e-03</td></tr>\n",
       "\t<tr><td>chr9_99906655_99907156</td><td>chr9_99502368_99502869</td><td>3.049706e-04</td></tr>\n",
       "\t<tr><td>chr9_99906655_99907156</td><td>chr9_99714500_99715001</td><td>1.296867e-03</td></tr>\n",
       "\t<tr><td>chr9_99906655_99907156</td><td>chr9_99909926_99910427</td><td>2.295261e-03</td></tr>\n",
       "\t<tr><td>chr9_99909926_99910427</td><td>chr9_99501516_99502017</td><td>1.325823e-02</td></tr>\n",
       "\t<tr><td>chr9_99909926_99910427</td><td>chr9_99502368_99502869</td><td>1.530700e-03</td></tr>\n",
       "\t<tr><td>chr9_99909926_99910427</td><td>chr9_99714500_99715001</td><td>2.296504e-03</td></tr>\n",
       "\t<tr><td>chr9_99909926_99910427</td><td>chr9_99906655_99907156</td><td>2.295261e-03</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99502368_99502869</td><td>7.341571e-04</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99509643_99510144</td><td>1.929738e-04</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99510426_99510927</td><td>8.087583e-04</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99580155_99580656</td><td>4.885006e-02</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99667197_99667698</td><td>6.808478e-03</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99714500_99715001</td><td>3.173281e-02</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99715259_99715760</td><td>3.601647e-02</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99903993_99904494</td><td>1.441456e-03</td></tr>\n",
       "\t<tr><td>chr9_99951231_99951732</td><td>chr9_99952043_99952544</td><td>4.174951e-02</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99502368_99502869</td><td>2.172401e-03</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99509643_99510144</td><td>5.519552e-05</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99510426_99510927</td><td>2.313259e-04</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99580155_99580656</td><td>1.541104e-02</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99667197_99667698</td><td>4.599720e-02</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99714500_99715001</td><td>1.002892e-01</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99715259_99715760</td><td>9.383894e-03</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99903993_99904494</td><td>5.885417e-03</td></tr>\n",
       "\t<tr><td>chr9_99952043_99952544</td><td>chr9_99951231_99951732</td><td>4.174951e-02</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 350114 × 3\n",
       "\\begin{tabular}{lll}\n",
       " Peak1 & Peak2 & coaccess\\\\\n",
       " <chr> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr10\\_100241725\\_100242226 & chr10\\_100242357\\_100242858 & 0.0906766262\\\\\n",
       "\t chr10\\_100242357\\_100242858 & chr10\\_100241725\\_100242226 & 0.0906766262\\\\\n",
       "\t chr10\\_100484245\\_100484746 & chr10\\_100485037\\_100485538 & 0.1135825904\\\\\n",
       "\t chr10\\_100485037\\_100485538 & chr10\\_100484245\\_100484746 & 0.1135825904\\\\\n",
       "\t chr10\\_10048691\\_10049192   & chr10\\_9768945\\_9769446     & 0.1164981762\\\\\n",
       "\t chr10\\_10048691\\_10049192   & chr10\\_10140631\\_10141132   & 0.1547400509\\\\\n",
       "\t chr10\\_10048691\\_10049192   & chr10\\_10146015\\_10146516   & 0.1702093118\\\\\n",
       "\t chr10\\_10048691\\_10049192   & chr10\\_10236626\\_10237127   & 0.0288173343\\\\\n",
       "\t chr10\\_10048691\\_10049192   & chr10\\_10241427\\_10241928   & 0.0188652891\\\\\n",
       "\t chr10\\_1009083\\_1009584     & chr10\\_1463639\\_1464140     & 0.0005210603\\\\\n",
       "\t chr10\\_101335383\\_101335884 & chr10\\_101698337\\_101698838 & 0.0069869046\\\\\n",
       "\t chr10\\_101338451\\_101338952 & chr10\\_101506152\\_101506653 & 0.0410098745\\\\\n",
       "\t chr10\\_101338451\\_101338952 & chr10\\_101512935\\_101513436 & 0.0157884476\\\\\n",
       "\t chr10\\_10140631\\_10141132   & chr10\\_9768945\\_9769446     & 0.0180269337\\\\\n",
       "\t chr10\\_10140631\\_10141132   & chr10\\_10048691\\_10049192   & 0.1547400509\\\\\n",
       "\t chr10\\_10140631\\_10141132   & chr10\\_10146015\\_10146516   & 0.0263605137\\\\\n",
       "\t chr10\\_10140631\\_10141132   & chr10\\_10236626\\_10237127   & 0.0044593474\\\\\n",
       "\t chr10\\_10140631\\_10141132   & chr10\\_10241427\\_10241928   & 0.0029191281\\\\\n",
       "\t chr10\\_10146015\\_10146516   & chr10\\_9768945\\_9769446     & 0.0198290744\\\\\n",
       "\t chr10\\_10146015\\_10146516   & chr10\\_10048691\\_10049192   & 0.1702093118\\\\\n",
       "\t chr10\\_10146015\\_10146516   & chr10\\_10140631\\_10141132   & 0.0263605137\\\\\n",
       "\t chr10\\_10146015\\_10146516   & chr10\\_10236626\\_10237127   & 0.0049059460\\\\\n",
       "\t chr10\\_10146015\\_10146516   & chr10\\_10241427\\_10241928   & 0.0032104878\\\\\n",
       "\t chr10\\_101506152\\_101506653 & chr10\\_101338451\\_101338952 & 0.0410098745\\\\\n",
       "\t chr10\\_101506152\\_101506653 & chr10\\_101512935\\_101513436 & 0.0003237411\\\\\n",
       "\t chr10\\_101512935\\_101513436 & chr10\\_101338451\\_101338952 & 0.0157884476\\\\\n",
       "\t chr10\\_101512935\\_101513436 & chr10\\_101506152\\_101506653 & 0.0003237411\\\\\n",
       "\t chr10\\_101520326\\_101520827 & chr10\\_101521026\\_101521527 & 0.1838308949\\\\\n",
       "\t chr10\\_101520326\\_101520827 & chr10\\_101698337\\_101698838 & 0.0082854999\\\\\n",
       "\t chr10\\_101521026\\_101521527 & chr10\\_101520326\\_101520827 & 0.1838308949\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t chr9\\_99903993\\_99904494 & chr9\\_99715259\\_99715760 & 6.791758e-03\\\\\n",
       "\t chr9\\_99903993\\_99904494 & chr9\\_99848477\\_99848978 & 1.026813e-01\\\\\n",
       "\t chr9\\_99903993\\_99904494 & chr9\\_99951231\\_99951732 & 1.441456e-03\\\\\n",
       "\t chr9\\_99903993\\_99904494 & chr9\\_99952043\\_99952544 & 5.885417e-03\\\\\n",
       "\t chr9\\_99906655\\_99907156 & chr9\\_99501516\\_99502017 & 1.672635e-03\\\\\n",
       "\t chr9\\_99906655\\_99907156 & chr9\\_99502368\\_99502869 & 3.049706e-04\\\\\n",
       "\t chr9\\_99906655\\_99907156 & chr9\\_99714500\\_99715001 & 1.296867e-03\\\\\n",
       "\t chr9\\_99906655\\_99907156 & chr9\\_99909926\\_99910427 & 2.295261e-03\\\\\n",
       "\t chr9\\_99909926\\_99910427 & chr9\\_99501516\\_99502017 & 1.325823e-02\\\\\n",
       "\t chr9\\_99909926\\_99910427 & chr9\\_99502368\\_99502869 & 1.530700e-03\\\\\n",
       "\t chr9\\_99909926\\_99910427 & chr9\\_99714500\\_99715001 & 2.296504e-03\\\\\n",
       "\t chr9\\_99909926\\_99910427 & chr9\\_99906655\\_99907156 & 2.295261e-03\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99502368\\_99502869 & 7.341571e-04\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99509643\\_99510144 & 1.929738e-04\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99510426\\_99510927 & 8.087583e-04\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99580155\\_99580656 & 4.885006e-02\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99667197\\_99667698 & 6.808478e-03\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99714500\\_99715001 & 3.173281e-02\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99715259\\_99715760 & 3.601647e-02\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99903993\\_99904494 & 1.441456e-03\\\\\n",
       "\t chr9\\_99951231\\_99951732 & chr9\\_99952043\\_99952544 & 4.174951e-02\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99502368\\_99502869 & 2.172401e-03\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99509643\\_99510144 & 5.519552e-05\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99510426\\_99510927 & 2.313259e-04\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99580155\\_99580656 & 1.541104e-02\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99667197\\_99667698 & 4.599720e-02\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99714500\\_99715001 & 1.002892e-01\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99715259\\_99715760 & 9.383894e-03\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99903993\\_99904494 & 5.885417e-03\\\\\n",
       "\t chr9\\_99952043\\_99952544 & chr9\\_99951231\\_99951732 & 4.174951e-02\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 350114 × 3\n",
       "\n",
       "| Peak1 &lt;chr&gt; | Peak2 &lt;fct&gt; | coaccess &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| chr10_100241725_100242226 | chr10_100242357_100242858 | 0.0906766262 |\n",
       "| chr10_100242357_100242858 | chr10_100241725_100242226 | 0.0906766262 |\n",
       "| chr10_100484245_100484746 | chr10_100485037_100485538 | 0.1135825904 |\n",
       "| chr10_100485037_100485538 | chr10_100484245_100484746 | 0.1135825904 |\n",
       "| chr10_10048691_10049192   | chr10_9768945_9769446     | 0.1164981762 |\n",
       "| chr10_10048691_10049192   | chr10_10140631_10141132   | 0.1547400509 |\n",
       "| chr10_10048691_10049192   | chr10_10146015_10146516   | 0.1702093118 |\n",
       "| chr10_10048691_10049192   | chr10_10236626_10237127   | 0.0288173343 |\n",
       "| chr10_10048691_10049192   | chr10_10241427_10241928   | 0.0188652891 |\n",
       "| chr10_1009083_1009584     | chr10_1463639_1464140     | 0.0005210603 |\n",
       "| chr10_101335383_101335884 | chr10_101698337_101698838 | 0.0069869046 |\n",
       "| chr10_101338451_101338952 | chr10_101506152_101506653 | 0.0410098745 |\n",
       "| chr10_101338451_101338952 | chr10_101512935_101513436 | 0.0157884476 |\n",
       "| chr10_10140631_10141132   | chr10_9768945_9769446     | 0.0180269337 |\n",
       "| chr10_10140631_10141132   | chr10_10048691_10049192   | 0.1547400509 |\n",
       "| chr10_10140631_10141132   | chr10_10146015_10146516   | 0.0263605137 |\n",
       "| chr10_10140631_10141132   | chr10_10236626_10237127   | 0.0044593474 |\n",
       "| chr10_10140631_10141132   | chr10_10241427_10241928   | 0.0029191281 |\n",
       "| chr10_10146015_10146516   | chr10_9768945_9769446     | 0.0198290744 |\n",
       "| chr10_10146015_10146516   | chr10_10048691_10049192   | 0.1702093118 |\n",
       "| chr10_10146015_10146516   | chr10_10140631_10141132   | 0.0263605137 |\n",
       "| chr10_10146015_10146516   | chr10_10236626_10237127   | 0.0049059460 |\n",
       "| chr10_10146015_10146516   | chr10_10241427_10241928   | 0.0032104878 |\n",
       "| chr10_101506152_101506653 | chr10_101338451_101338952 | 0.0410098745 |\n",
       "| chr10_101506152_101506653 | chr10_101512935_101513436 | 0.0003237411 |\n",
       "| chr10_101512935_101513436 | chr10_101338451_101338952 | 0.0157884476 |\n",
       "| chr10_101512935_101513436 | chr10_101506152_101506653 | 0.0003237411 |\n",
       "| chr10_101520326_101520827 | chr10_101521026_101521527 | 0.1838308949 |\n",
       "| chr10_101520326_101520827 | chr10_101698337_101698838 | 0.0082854999 |\n",
       "| chr10_101521026_101521527 | chr10_101520326_101520827 | 0.1838308949 |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| chr9_99903993_99904494 | chr9_99715259_99715760 | 6.791758e-03 |\n",
       "| chr9_99903993_99904494 | chr9_99848477_99848978 | 1.026813e-01 |\n",
       "| chr9_99903993_99904494 | chr9_99951231_99951732 | 1.441456e-03 |\n",
       "| chr9_99903993_99904494 | chr9_99952043_99952544 | 5.885417e-03 |\n",
       "| chr9_99906655_99907156 | chr9_99501516_99502017 | 1.672635e-03 |\n",
       "| chr9_99906655_99907156 | chr9_99502368_99502869 | 3.049706e-04 |\n",
       "| chr9_99906655_99907156 | chr9_99714500_99715001 | 1.296867e-03 |\n",
       "| chr9_99906655_99907156 | chr9_99909926_99910427 | 2.295261e-03 |\n",
       "| chr9_99909926_99910427 | chr9_99501516_99502017 | 1.325823e-02 |\n",
       "| chr9_99909926_99910427 | chr9_99502368_99502869 | 1.530700e-03 |\n",
       "| chr9_99909926_99910427 | chr9_99714500_99715001 | 2.296504e-03 |\n",
       "| chr9_99909926_99910427 | chr9_99906655_99907156 | 2.295261e-03 |\n",
       "| chr9_99951231_99951732 | chr9_99502368_99502869 | 7.341571e-04 |\n",
       "| chr9_99951231_99951732 | chr9_99509643_99510144 | 1.929738e-04 |\n",
       "| chr9_99951231_99951732 | chr9_99510426_99510927 | 8.087583e-04 |\n",
       "| chr9_99951231_99951732 | chr9_99580155_99580656 | 4.885006e-02 |\n",
       "| chr9_99951231_99951732 | chr9_99667197_99667698 | 6.808478e-03 |\n",
       "| chr9_99951231_99951732 | chr9_99714500_99715001 | 3.173281e-02 |\n",
       "| chr9_99951231_99951732 | chr9_99715259_99715760 | 3.601647e-02 |\n",
       "| chr9_99951231_99951732 | chr9_99903993_99904494 | 1.441456e-03 |\n",
       "| chr9_99951231_99951732 | chr9_99952043_99952544 | 4.174951e-02 |\n",
       "| chr9_99952043_99952544 | chr9_99502368_99502869 | 2.172401e-03 |\n",
       "| chr9_99952043_99952544 | chr9_99509643_99510144 | 5.519552e-05 |\n",
       "| chr9_99952043_99952544 | chr9_99510426_99510927 | 2.313259e-04 |\n",
       "| chr9_99952043_99952544 | chr9_99580155_99580656 | 1.541104e-02 |\n",
       "| chr9_99952043_99952544 | chr9_99667197_99667698 | 4.599720e-02 |\n",
       "| chr9_99952043_99952544 | chr9_99714500_99715001 | 1.002892e-01 |\n",
       "| chr9_99952043_99952544 | chr9_99715259_99715760 | 9.383894e-03 |\n",
       "| chr9_99952043_99952544 | chr9_99903993_99904494 | 5.885417e-03 |\n",
       "| chr9_99952043_99952544 | chr9_99951231_99951732 | 4.174951e-02 |\n",
       "\n"
      ],
      "text/plain": [
       "       Peak1                     Peak2                     coaccess    \n",
       "1      chr10_100241725_100242226 chr10_100242357_100242858 0.0906766262\n",
       "2      chr10_100242357_100242858 chr10_100241725_100242226 0.0906766262\n",
       "3      chr10_100484245_100484746 chr10_100485037_100485538 0.1135825904\n",
       "4      chr10_100485037_100485538 chr10_100484245_100484746 0.1135825904\n",
       "5      chr10_10048691_10049192   chr10_9768945_9769446     0.1164981762\n",
       "6      chr10_10048691_10049192   chr10_10140631_10141132   0.1547400509\n",
       "7      chr10_10048691_10049192   chr10_10146015_10146516   0.1702093118\n",
       "8      chr10_10048691_10049192   chr10_10236626_10237127   0.0288173343\n",
       "9      chr10_10048691_10049192   chr10_10241427_10241928   0.0188652891\n",
       "10     chr10_1009083_1009584     chr10_1463639_1464140     0.0005210603\n",
       "11     chr10_101335383_101335884 chr10_101698337_101698838 0.0069869046\n",
       "12     chr10_101338451_101338952 chr10_101506152_101506653 0.0410098745\n",
       "13     chr10_101338451_101338952 chr10_101512935_101513436 0.0157884476\n",
       "14     chr10_10140631_10141132   chr10_9768945_9769446     0.0180269337\n",
       "15     chr10_10140631_10141132   chr10_10048691_10049192   0.1547400509\n",
       "16     chr10_10140631_10141132   chr10_10146015_10146516   0.0263605137\n",
       "17     chr10_10140631_10141132   chr10_10236626_10237127   0.0044593474\n",
       "18     chr10_10140631_10141132   chr10_10241427_10241928   0.0029191281\n",
       "19     chr10_10146015_10146516   chr10_9768945_9769446     0.0198290744\n",
       "20     chr10_10146015_10146516   chr10_10048691_10049192   0.1702093118\n",
       "21     chr10_10146015_10146516   chr10_10140631_10141132   0.0263605137\n",
       "22     chr10_10146015_10146516   chr10_10236626_10237127   0.0049059460\n",
       "23     chr10_10146015_10146516   chr10_10241427_10241928   0.0032104878\n",
       "24     chr10_101506152_101506653 chr10_101338451_101338952 0.0410098745\n",
       "25     chr10_101506152_101506653 chr10_101512935_101513436 0.0003237411\n",
       "26     chr10_101512935_101513436 chr10_101338451_101338952 0.0157884476\n",
       "27     chr10_101512935_101513436 chr10_101506152_101506653 0.0003237411\n",
       "28     chr10_101520326_101520827 chr10_101521026_101521527 0.1838308949\n",
       "29     chr10_101520326_101520827 chr10_101698337_101698838 0.0082854999\n",
       "30     chr10_101521026_101521527 chr10_101520326_101520827 0.1838308949\n",
       "<U+22EE> <U+22EE>                  <U+22EE>                  <U+22EE>    \n",
       "350085 chr9_99903993_99904494    chr9_99715259_99715760    6.791758e-03\n",
       "350086 chr9_99903993_99904494    chr9_99848477_99848978    1.026813e-01\n",
       "350087 chr9_99903993_99904494    chr9_99951231_99951732    1.441456e-03\n",
       "350088 chr9_99903993_99904494    chr9_99952043_99952544    5.885417e-03\n",
       "350089 chr9_99906655_99907156    chr9_99501516_99502017    1.672635e-03\n",
       "350090 chr9_99906655_99907156    chr9_99502368_99502869    3.049706e-04\n",
       "350091 chr9_99906655_99907156    chr9_99714500_99715001    1.296867e-03\n",
       "350092 chr9_99906655_99907156    chr9_99909926_99910427    2.295261e-03\n",
       "350093 chr9_99909926_99910427    chr9_99501516_99502017    1.325823e-02\n",
       "350094 chr9_99909926_99910427    chr9_99502368_99502869    1.530700e-03\n",
       "350095 chr9_99909926_99910427    chr9_99714500_99715001    2.296504e-03\n",
       "350096 chr9_99909926_99910427    chr9_99906655_99907156    2.295261e-03\n",
       "350097 chr9_99951231_99951732    chr9_99502368_99502869    7.341571e-04\n",
       "350098 chr9_99951231_99951732    chr9_99509643_99510144    1.929738e-04\n",
       "350099 chr9_99951231_99951732    chr9_99510426_99510927    8.087583e-04\n",
       "350100 chr9_99951231_99951732    chr9_99580155_99580656    4.885006e-02\n",
       "350101 chr9_99951231_99951732    chr9_99667197_99667698    6.808478e-03\n",
       "350102 chr9_99951231_99951732    chr9_99714500_99715001    3.173281e-02\n",
       "350103 chr9_99951231_99951732    chr9_99715259_99715760    3.601647e-02\n",
       "350104 chr9_99951231_99951732    chr9_99903993_99904494    1.441456e-03\n",
       "350105 chr9_99951231_99951732    chr9_99952043_99952544    4.174951e-02\n",
       "350106 chr9_99952043_99952544    chr9_99502368_99502869    2.172401e-03\n",
       "350107 chr9_99952043_99952544    chr9_99509643_99510144    5.519552e-05\n",
       "350108 chr9_99952043_99952544    chr9_99510426_99510927    2.313259e-04\n",
       "350109 chr9_99952043_99952544    chr9_99580155_99580656    1.541104e-02\n",
       "350110 chr9_99952043_99952544    chr9_99667197_99667698    4.599720e-02\n",
       "350111 chr9_99952043_99952544    chr9_99714500_99715001    1.002892e-01\n",
       "350112 chr9_99952043_99952544    chr9_99715259_99715760    9.383894e-03\n",
       "350113 chr9_99952043_99952544    chr9_99903993_99904494    5.885417e-03\n",
       "350114 chr9_99952043_99952544    chr9_99951231_99951732    4.174951e-02"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "running_cic %>% \n",
    "    dplyr::filter(coaccess > 0)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
