{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "377b2940-c3ef-4d5e-b30b-44aacadb0799",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.2          \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0.\u001b[31m9000\u001b[39m     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.3          \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2          \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0     \n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2          \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "01a84a85-0773-450e-8e71-c4065205837c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m33\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m62\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m31\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m28\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m31\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): chrom, geneID, name, type\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "marker_file_paths <- c(\"/home/jpm73279/genome_downloads/C4_markers/Os.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Pm.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Sb.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Uf.c4_markers.bed\",\n",
    "                \"/home/jpm73279/genome_downloads/C4_markers/Zm.c4_markers.bed\")\n",
    "\n",
    "# Function to read and process the file\n",
    "process_file <- function(file_path) {\n",
    "  species <- substr(basename(file_path), 1, 2)\n",
    "  read_delim(file_path, delim = \"\\t\", col_names = c(\"chrom\", \"start\", \"end\", \"geneID\", \"name\", \"type\")) %>% \n",
    "    dplyr::select(\"chrom\",\"start\",\"end\",\"geneID\",\"name\",\"type\") %>%\n",
    "    dplyr::mutate(species = tolower(species))\n",
    "}\n",
    "\n",
    "# Apply the function to each file and store results in a list\n",
    "list_of_dataframes <- lapply(marker_file_paths, process_file)\n",
    "\n",
    "# Optionally combine all dataframes into one if needed\n",
    "markers <- bind_rows(list_of_dataframes) %>% \n",
    "    dplyr::select(species, name, type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aaa00fa-1f70-4db2-af07-6e0327130ece",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "11bacf64-807a-4308-a669-414688b19b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# zm_fimo_output <- read_tsv(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/06.motif_enrichment/zm_dir/all_acrs/zm_fimo_output/fimo.tsv\")\n",
    "# col_names <- c(\n",
    "#   \"chromosome_1\", \n",
    "#   \"start_1\", \n",
    "#   \"end_1\", \n",
    "#   \"gene_id\", \n",
    "#   \"gene_name\", \n",
    "#   \"chromosome_2\", \n",
    "#   \"start_2\", \n",
    "#   \"end_2\", \n",
    "#   \"acrs_id\", \n",
    "#   \"p_value\"\n",
    "# )\n",
    "# zm_ms_intersect_file <- \"/home/jpm73279/genome_downloads/C4_markers/Zm_ms_acrs_intersect.bed\"\n",
    "# zm_bs_intersect_file <- \"/home/jpm73279/genome_downloads/C4_markers/Zm_bs_acrs_intersect.bed\"\n",
    "# zm_ms <- read_tsv(zm_ms_intersect_file, col_names = col_names) %>% \n",
    "#     mutate(type = \"ms\")\n",
    "\n",
    "# zm_bs <- read_tsv(zm_bs_intersect_file, col_names = col_names) %>% \n",
    "#     mutate(type = \"bs\")\n",
    "\n",
    "# combined_rows <- bind_rows(zm_ms, zm_bs) %>% \n",
    "#     dplyr::mutate(species = \"zm\")\n",
    "\n",
    "\n",
    "# number_acrs_per_type <- combined_rows %>% \n",
    "#     dplyr::group_by(species, type) %>% \n",
    "#     dplyr::summarise(counts = n())\n",
    "\n",
    "\n",
    "# zm_fimo_output <- zm_fimo_output %>% \n",
    "#     separate(sequence_name, c(\"acrID\", \"region\"), sep = \"::\") %>% \n",
    "#     dplyr::filter(acrID %in% combined_rows$acrs_id)\n",
    "\n",
    "# number_acrs_per_type <- combined_rows %>% \n",
    "#     dplyr::group_by(species, type) %>% \n",
    "#     dplyr::summarise(counts = n())\n",
    "\n",
    "# intersection_combined <- combined_rows %>% \n",
    "#     dplyr::select(gene_id, gene_name, acrs_id, type, species)\n",
    "\n",
    "\n",
    "# combined_fimo_intersect <- left_join(zm_fimo_output, intersection_combined, by = c(\"acrID\" = \"acrs_id\"), relationship =  \"many-to-many\") %>% \n",
    "#     dplyr::select(motif_id, acrID, gene_id,gene_name,type,species) %>% \n",
    "#     unique()\n",
    "#     #left_join(., number_acrs_per_type, by = c(\"type\", \"species\"))\n",
    "\n",
    "# contingency_start <- combined_fimo_intersect %>% \n",
    "#     dplyr::group_by(motif_id, type, species) %>% \n",
    "#     dplyr::summarise(count_motif_occurence = n()) %>% \n",
    "#     left_join(., number_acrs_per_type, by = c(\"type\", \"species\")) %>% \n",
    "#     dplyr::rowwise() %>% \n",
    "#     mutate(non_motif_occurence = counts - count_motif_occurence)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c3ab6908-a663-4795-945d-be905ae10aac",
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_counts_contingency_table <- function(intersection_files, species_name) {\n",
    "  col_names <- c(\n",
    "    \"chromosome_1\", \n",
    "    \"start_1\", \n",
    "    \"end_1\", \n",
    "    \"gene_id\", \n",
    "    \"gene_name\", \n",
    "    \"chromosome_2\", \n",
    "    \"start_2\", \n",
    "    \"end_2\", \n",
    "    \"acrs_id\", \n",
    "    \"p_value\"\n",
    "  )\n",
    "  # Read the ms intersect file and mutate type\n",
    "  zm_ms <- readr::read_tsv(intersection_files[[1]], col_names = col_names) %>% \n",
    "    dplyr::mutate(type = \"bs\")\n",
    "\n",
    "  # Read the bs intersect file and mutate type\n",
    "  zm_bs <- readr::read_tsv(intersection_files[[2]], col_names = col_names) %>% \n",
    "    dplyr::mutate(type = \"ms\")\n",
    "\n",
    "  # Combine rows and mutate species\n",
    "  combined_rows <- dplyr::bind_rows(zm_ms, zm_bs) %>% \n",
    "    dplyr::mutate(species = species_name)\n",
    "\n",
    "  # Calculate number of acrs per type\n",
    "  number_acrs_per_type <- combined_rows %>% \n",
    "    dplyr::group_by(species, type) %>% \n",
    "    dplyr::summarise(counts = n())\n",
    "  \n",
    "  return(number_acrs_per_type)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "process_data <- function(fimo_output, intersection_files, species_name) {\n",
    "  \n",
    "  # Read the FIMO output\n",
    "  zm_fimo_output <- read_tsv(fimo_output)\n",
    "  \n",
    "  col_names <- c(\n",
    "    \"chromosome_1\", \n",
    "    \"start_1\", \n",
    "    \"end_1\", \n",
    "    \"gene_id\", \n",
    "    \"gene_name\", \n",
    "    \"chromosome_2\", \n",
    "    \"start_2\", \n",
    "    \"end_2\", \n",
    "    \"acrs_id\", \n",
    "    \"p_value\"\n",
    "  )\n",
    "  \n",
    "  # Read the intersection files\n",
    "  zm_ms <- read_tsv(intersection_files[[1]], col_names = col_names) %>% \n",
    "        mutate(type = \"bs\") %>% \n",
    "        dplyr::mutate(species = species_name)\n",
    "  zm_bs <- read_tsv(intersection_files[[2]], col_names = col_names) %>% \n",
    "        mutate(type = \"ms\") %>% \n",
    "        dplyr::mutate(species = species_name)\n",
    "\n",
    "        \n",
    "  \n",
    "  # Combine rows from both intersection files\n",
    "  combined_rows <- bind_rows(zm_ms, zm_bs)\n",
    "  \n",
    "  # Process the FIMO output\n",
    "  zm_fimo_filtered_output <- zm_fimo_output %>% \n",
    "    separate(sequence_name, c(\"acrID\", \"region\"), sep = \"::\") %>% \n",
    "    dplyr::filter(acrID %in% combined_rows$acrs_id)\n",
    "  \n",
    "  # Extract selected columns from combined_rows\n",
    "  intersection_combined <- combined_rows %>% \n",
    "    dplyr::select(gene_id, gene_name, acrs_id, type)\n",
    "  \n",
    "  # Combine the FIMO output with the intersection data\n",
    "  combined_fimo_intersect <- left_join(zm_fimo_filtered_output, intersection_combined, by = c(\"acrID\" = \"acrs_id\"), relationship = \"many-to-many\") %>% \n",
    "        dplyr::mutate(species = species_name)\n",
    "    \n",
    "    rm(zm_fimo_output)\n",
    "      rm(zm_ms)\n",
    "    rm(zm_bs)\n",
    "    gc()\n",
    "  return(combined_fimo_intersect)\n",
    "}\n",
    "\n",
    "\n",
    "process_data_null_files <- function(fimo_output, intersection_files, species_name) {\n",
    "  \n",
    "  # Read the FIMO output\n",
    "  zm_fimo_output <- read_tsv(fimo_output)\n",
    "  \n",
    "  col_names <- c(\n",
    "    \"chromosome\", \n",
    "    \"start\", \n",
    "    \"end\", \n",
    "    \"acrs_id\", \n",
    "    \"p_value\"\n",
    "  )\n",
    "  \n",
    "  # Read the intersection files\n",
    "  zm_ms <- read_tsv(intersection_files[[1]], col_names = col_names) %>% \n",
    "        mutate(type = \"bs\") %>% \n",
    "        mutate(gene_name = \"null\")\n",
    "  zm_bs <- read_tsv(intersection_files[[2]], col_names = col_names) %>% \n",
    "        mutate(type = \"ms\") %>% \n",
    "        mutate(gene_name = \"null\")\n",
    "        \n",
    "  \n",
    "  # Combine rows from both intersection files\n",
    "  combined_rows <- bind_rows(zm_ms, zm_bs)\n",
    "  \n",
    "  # Process the FIMO output\n",
    "  zm_fimo_filtered_output <- zm_fimo_output %>% \n",
    "    separate(sequence_name, c(\"acrID\", \"region\"), sep = \"::\") %>% \n",
    "    dplyr::filter(acrID %in% combined_rows$acrs_id)\n",
    "  \n",
    "  # Extract selected columns from combined_rows\n",
    "  intersection_combined <- combined_rows %>% \n",
    "    dplyr::select(gene_name, acrs_id, type)\n",
    "  \n",
    "  # Combine the FIMO output with the intersection data\n",
    "  combined_fimo_intersect <- left_join(zm_fimo_filtered_output, intersection_combined, by = c(\"acrID\" = \"acrs_id\"), relationship = \"many-to-many\") %>% \n",
    "        dplyr::mutate(species = species_name)\n",
    "    \n",
    "    rm(zm_fimo_output)\n",
    "      rm(zm_ms)\n",
    "    rm(zm_bs)\n",
    "    gc()\n",
    "  return(combined_fimo_intersect)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "15a36d79-f9c7-4d48-a39b-40da8cad0ef8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "base_path <- \"/home/jpm73279/genome_downloads/C4_markers/\"\n",
    "intersect_list <- list(\n",
    "  Os = list(\n",
    "    bs = paste0(base_path, \"Os_bs_acrs_intersect.bed\"),\n",
    "    ms = paste0(base_path, \"Os_ms_acrs_intersect.bed\")\n",
    "  ),\n",
    "  Pm = list(\n",
    "    bs = paste0(base_path, \"Pm_bs_acrs_intersect.bed\"),\n",
    "    ms = paste0(base_path, \"Pm_ms_acrs_intersect.bed\")\n",
    "  ),\n",
    "  Sb = list(\n",
    "    bs = paste0(base_path, \"Sb_bs_acrs_intersect.bed\"),\n",
    "    ms = paste0(base_path, \"Sb_ms_acrs_intersect.bed\")\n",
    "  ),\n",
    "  Uf = list(\n",
    "    bs = paste0(base_path, \"Uf_bs_acrs_intersect.bed\"),\n",
    "    ms = paste0(base_path, \"Uf_ms_acrs_intersect.bed\")\n",
    "  ),\n",
    "  Zm = list(\n",
    "    bs = paste0(base_path, \"Zm_bs_acrs_intersect.bed\"),\n",
    "    ms = paste0(base_path, \"Zm_ms_acrs_intersect.bed\")\n",
    "  )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "551b0ebb-d277-4864-9702-129108f27eb7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "64a5dd47-8a8e-4412-9fcd-eb6609f67de4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m13\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m29\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'species'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m54\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m38\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'species'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m14\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m25\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'species'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m35\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m33\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'species'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m14\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m28\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): chromosome_1, gene_id, gene_name, chromosome_2, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (5): start_1, end_1, start_2, end_2, p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'species'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "zm_fimo_counts_all <- generate_counts_contingency_table(c(intersect_list$Zm$bs, intersect_list$Zm$ms), \"zm\")       \n",
    "pm_fimo_counts_all <- generate_counts_contingency_table(c(intersect_list$Pm$bs, intersect_list$Pm$ms), \"pm\")       \n",
    "os_fimo_counts_all <- generate_counts_contingency_table(c(intersect_list$Os$bs, intersect_list$Os$ms), \"os\")       \n",
    "uf_fimo_counts_all <- generate_counts_contingency_table(c(intersect_list$Uf$bs, intersect_list$Uf$ms), \"uf\")       \n",
    "sb_fimo_counts_all <- generate_counts_contingency_table(c(intersect_list$Sb$bs, intersect_list$Sb$ms), \"sb\")       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2f36b0c1-fda6-470c-a7de-85b9f30b0355",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>species</th><th scope=col>type</th><th scope=col>counts</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>sb</td><td>bs</td><td>14</td></tr>\n",
       "\t<tr><td>sb</td><td>ms</td><td>28</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " species & type & counts\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t sb & bs & 14\\\\\n",
       "\t sb & ms & 28\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 3\n",
       "\n",
       "| species &lt;chr&gt; | type &lt;chr&gt; | counts &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| sb | bs | 14 |\n",
       "| sb | ms | 28 |\n",
       "\n"
      ],
      "text/plain": [
       "  species type counts\n",
       "1 sb      bs   14    \n",
       "2 sb      ms   28    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sb_fimo_counts_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dde4c939-df3b-42f1-a733-77ba56e2c413",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_species_fimo_counts <- bind_rows(zm_fimo_counts, pm_fimo_counts, os_fimo_counts,sb_fimo_counts,uf_fimo_counts)\n",
    "all_species_total_counts <- bind_rows(zm_fimo_counts_all,pm_fimo_counts_all, os_fimo_counts_all, uf_fimo_counts_all, sb_fimo_counts_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1eb011d2-b67e-45c7-9d21-39db2759659d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 707 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>motif_alt_id</th><th scope=col>acrID</th><th scope=col>region</th><th scope=col>start</th><th scope=col>stop</th><th scope=col>strand</th><th scope=col>score</th><th scope=col>p-value</th><th scope=col>q-value</th><th scope=col>matched_sequence</th><th scope=col>gene_id</th><th scope=col>gene_name</th><th scope=col>type</th><th scope=col>species</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>HSF        </td><td>NA</td><td>scACR_5926;mesophyll     </td><td>chr1:226422323-226422824</td><td>450</td><td>463</td><td>-</td><td>21.5688</td><td>1.46e-08</td><td>0.06570</td><td>AGAAGGTTCTAGAA </td><td>Zm00001eb042970</td><td>PPT              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_42073;mesophyll    </td><td>chr6:177097950-177098451</td><td>349</td><td>362</td><td>-</td><td>16.5854</td><td>9.56e-07</td><td>0.00282</td><td>AAGAGAGAGAAGGA </td><td>Zm00001eb296550</td><td>NAD-MDH2         </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>MADS_MIKC  </td><td>NA</td><td>scACR_1560;mesophyll     </td><td>chr1:35695232-35695733  </td><td>428</td><td>440</td><td>+</td><td>14.9458</td><td>1.22e-06</td><td>0.56700</td><td>CCCAAAATGGTAA  </td><td>Zm00001eb011080</td><td>PEPCK            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_42073;mesophyll    </td><td>chr6:177097950-177098451</td><td>353</td><td>366</td><td>-</td><td>16.1829</td><td>1.56e-06</td><td>0.00431</td><td>AAGAAAGAGAGAGA </td><td>Zm00001eb296550</td><td>NAD-MDH2         </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_42073;mesophyll    </td><td>chr6:177097950-177098451</td><td>347</td><td>360</td><td>-</td><td>16.1098</td><td>1.70e-06</td><td>0.00465</td><td>GAGAGAGAAGGAGA </td><td>Zm00001eb296550</td><td>NAD-MDH2         </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BES1       </td><td>NA</td><td>scACR_51667;bundle_sheath</td><td>chr8:181465456-181465957</td><td>175</td><td>184</td><td>-</td><td>15.9024</td><td>2.10e-06</td><td>0.35600</td><td>TGCACGTGTG     </td><td>Zm00001eb371140</td><td>NADP_ME          </td><td>bs</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_8033;mesophyll     </td><td>chr1:289348515-289349016</td><td>237</td><td>250</td><td>-</td><td>15.3293</td><td>3.26e-06</td><td>0.00790</td><td>GGGAGAGAGGCAGA </td><td>Zm00001eb059070</td><td>NAD-MDH12        </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_24813;mesophyll    </td><td>chr3:220994969-220995470</td><td>364</td><td>377</td><td>-</td><td>15.0732</td><td>3.74e-06</td><td>0.00877</td><td>GCGAGAGAGCGCGA </td><td>Zm00001eb158800</td><td>CA2              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_24813;mesophyll    </td><td>chr3:220994969-220995470</td><td>364</td><td>377</td><td>-</td><td>15.0732</td><td>3.74e-06</td><td>0.00877</td><td>GCGAGAGAGCGCGA </td><td>Zm00001eb158810</td><td>CA1              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>MADS_MIKC  </td><td>NA</td><td>scACR_1560;mesophyll     </td><td>chr1:35695232-35695733  </td><td> 20</td><td> 32</td><td>+</td><td>13.6747</td><td>4.88e-06</td><td>0.67100</td><td>CCAAAAATGGTTA  </td><td>Zm00001eb011080</td><td>PEPCK            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>MADS_MIKC  </td><td>NA</td><td>scACR_50193;mesophyll    </td><td>chr8:150379201-150379702</td><td> 37</td><td> 49</td><td>+</td><td>13.1928</td><td>7.96e-06</td><td>0.69900</td><td>CCAAAAAAAGATA  </td><td>Zm00001eb359190</td><td>CA6              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>NA</td><td>scACR_1560;mesophyll     </td><td>chr1:35695232-35695733  </td><td>287</td><td>294</td><td>+</td><td>14.6633</td><td>8.13e-06</td><td>0.13300</td><td>GGGCCCAC       </td><td>Zm00001eb011080</td><td>PEPCK            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>NA</td><td>scACR_24813;mesophyll    </td><td>chr3:220994969-220995470</td><td> 90</td><td>104</td><td>-</td><td>13.8444</td><td>8.48e-06</td><td>0.21700</td><td>CCGCATCCACCGGCG</td><td>Zm00001eb158800</td><td>CA2              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>NA</td><td>scACR_24813;mesophyll    </td><td>chr3:220994969-220995470</td><td> 90</td><td>104</td><td>-</td><td>13.8444</td><td>8.48e-06</td><td>0.21700</td><td>CCGCATCCACCGGCG</td><td>Zm00001eb158810</td><td>CA1              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>HSF        </td><td>NA</td><td>scACR_5926;mesophyll     </td><td>chr1:226422323-226422824</td><td>454</td><td>467</td><td>+</td><td>13.4679</td><td>9.70e-06</td><td>0.65200</td><td>AGAACCTTCTCGAG </td><td>Zm00001eb042970</td><td>PPT              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>NA</td><td>scACR_8034;bundle_sheath </td><td>chr1:289354685-289355186</td><td> 24</td><td> 38</td><td>+</td><td>12.9889</td><td>1.66e-05</td><td>0.23000</td><td>CCGCGGAAGCCGGAG</td><td>Zm00001eb059070</td><td>NAD-MDH12        </td><td>bs</td><td>zm</td></tr>\n",
       "\t<tr><td>HSF        </td><td>NA</td><td>scACR_25494;bundle_sheath</td><td>chr4:1220891-1221392    </td><td>317</td><td>330</td><td>-</td><td>12.4312</td><td>1.77e-05</td><td>0.77600</td><td>TGAATATTCTAGAG </td><td>Zm00001eb164390</td><td>RUBISCO_activase1</td><td>bs</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_16177;mesophyll    </td><td>chr2:147004111-147004612</td><td>226</td><td>239</td><td>+</td><td>12.7073</td><td>1.96e-05</td><td>0.03110</td><td>TCGAGAGAGACGGC </td><td>Zm00001eb092540</td><td>RBCA2            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_16177;mesophyll    </td><td>chr2:147004111-147004612</td><td>228</td><td>241</td><td>+</td><td>12.6341</td><td>2.07e-05</td><td>0.03240</td><td>GAGAGAGACGGCGG </td><td>Zm00001eb092540</td><td>RBCA2            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_29627;mesophyll    </td><td>chr4:196205013-196205514</td><td>330</td><td>343</td><td>-</td><td>12.6341</td><td>2.07e-05</td><td>0.03240</td><td>GAGAGAGACGGCGG </td><td>Zm00001eb197410</td><td>RBCA1            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II</td><td>NA</td><td>scACR_40882;mesophyll    </td><td>chr6:157026520-157027021</td><td> 44</td><td> 51</td><td>-</td><td>12.5149</td><td>2.18e-05</td><td>1.00000</td><td>CAATCATT       </td><td>Zm00001eb287770</td><td>PPDK             </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II</td><td>NA</td><td>scACR_29625;bundle_sheath</td><td>chr4:196176098-196176599</td><td>486</td><td>493</td><td>+</td><td>12.5149</td><td>2.18e-05</td><td>1.00000</td><td>CAATCATT       </td><td>Zm00001eb197410</td><td>RBCA1            </td><td>bs</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_42073;mesophyll    </td><td>chr6:177097950-177098451</td><td>351</td><td>364</td><td>-</td><td>12.5366</td><td>2.21e-05</td><td>0.03390</td><td>GAAAGAGAGAGAAG </td><td>Zm00001eb296550</td><td>NAD-MDH2         </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>NA</td><td>scACR_50193;mesophyll    </td><td>chr8:150379201-150379702</td><td>436</td><td>450</td><td>+</td><td>12.5667</td><td>2.28e-05</td><td>0.24800</td><td>CCGTCATCTTCGCCA</td><td>Zm00001eb359190</td><td>CA6              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_29627;mesophyll    </td><td>chr4:196205013-196205514</td><td>332</td><td>345</td><td>-</td><td>12.4878</td><td>2.29e-05</td><td>0.03490</td><td>TTGAGAGAGACGGC </td><td>Zm00001eb197410</td><td>RBCA1            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>NA</td><td>scACR_5926;mesophyll     </td><td>chr1:226422323-226422824</td><td>187</td><td>194</td><td>+</td><td>12.3571</td><td>2.62e-05</td><td>0.19900</td><td>GGCCCCAC       </td><td>Zm00001eb042970</td><td>PPT              </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>NA</td><td>scACR_25495;bundle_sheath</td><td>chr4:1225196-1225697    </td><td>185</td><td>192</td><td>-</td><td>12.3571</td><td>2.62e-05</td><td>0.19900</td><td>GGCCCCAC       </td><td>Zm00001eb164390</td><td>RUBISCO_activase1</td><td>bs</td><td>zm</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>NA</td><td>scACR_25494;bundle_sheath</td><td>chr4:1220891-1221392    </td><td>342</td><td>349</td><td>-</td><td>12.3571</td><td>2.62e-05</td><td>0.19900</td><td>GGCCCCAC       </td><td>Zm00001eb164390</td><td>RUBISCO_activase1</td><td>bs</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>NA</td><td>scACR_8033;mesophyll     </td><td>chr1:289348515-289349016</td><td>235</td><td>248</td><td>-</td><td>12.1585</td><td>2.87e-05</td><td>0.04140</td><td>GAGAGAGGCAGAGG </td><td>Zm00001eb059070</td><td>NAD-MDH12        </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>NA</td><td>scACR_29627;mesophyll    </td><td>chr4:196205013-196205514</td><td>321</td><td>335</td><td>+</td><td>12.2000</td><td>2.97e-05</td><td>0.25900</td><td>ACGCCGCATCCGCCG</td><td>Zm00001eb197410</td><td>RBCA1            </td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>GARP_G2-like</td><td>NA</td><td>scACR_9874;mesophyll     </td><td>Chr03:17302413-17302914</td><td>464</td><td>470</td><td>+</td><td>10.01550</td><td>7.93e-05</td><td>1.0000</td><td>AAGATTC              </td><td>Urofu.3G289200.v1.1</td><td>PPDK                                                      </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>E2FDP       </td><td>NA</td><td>scACR_24584;bundle_sheath</td><td>Chr09:3736455-3736956  </td><td>375</td><td>387</td><td>-</td><td>10.75610</td><td>8.08e-05</td><td>0.7030</td><td>ATTTGGCGCGGGG        </td><td>Urofu.9G071000.v1.1</td><td>NAD-MDH12                                                 </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>NAC         </td><td>NA</td><td>scACR_25711;bundle_sheath</td><td>Chr09:11759335-11759836</td><td> 15</td><td> 35</td><td>+</td><td>10.08160</td><td>8.14e-05</td><td>1.0000</td><td>TGACTTATATTAGACGTTAAG</td><td>Urofu.9G196500.v1.1</td><td>BASS                                                      </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>NA</td><td>scACR_19222;bundle_sheath</td><td>Chr06:21745880-21746381</td><td>468</td><td>481</td><td>-</td><td>10.28050</td><td>8.23e-05</td><td>0.0870</td><td>GAGAAAGGGAAAGG       </td><td>Urofu.6G175300.v1.1</td><td>DIT2                                                      </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>NA</td><td>scACR_10421;bundle_sheath</td><td>Chr03:27572579-27573080</td><td>274</td><td>281</td><td>+</td><td>11.03060</td><td>8.23e-05</td><td>0.2040</td><td>GGGCCCAA             </td><td>Urofu.3G358900.v1.1</td><td>RBCA2_ortho_count_2__RBCA1_ortho_count_2                  </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>NA</td><td>scACR_5528;mesophyll     </td><td>Chr02:25948078-25948579</td><td>304</td><td>317</td><td>-</td><td>10.25610</td><td>8.34e-05</td><td>0.0875</td><td>GAGAAAGAACGCGA       </td><td>Urofu.2G245100.v1.1</td><td>DIT1                                                      </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>NAC         </td><td>NA</td><td>scACR_18603;bundle_sheath</td><td>Chr06:14687042-14687543</td><td>350</td><td>370</td><td>-</td><td> 9.98980</td><td>8.51e-05</td><td>1.0000</td><td>TCCCGTATTTCACAAGGAAAA</td><td>Urofu.6G101400.v1.1</td><td>PPT                                                       </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>NA</td><td>scACR_19222;bundle_sheath</td><td>Chr06:21745880-21746381</td><td>230</td><td>243</td><td>+</td><td>10.20730</td><td>8.58e-05</td><td>0.0885</td><td>GAGAGAGGGCTAGG       </td><td>Urofu.6G175300.v1.1</td><td>DIT2                                                      </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>NA</td><td>scACR_25711;bundle_sheath</td><td>Chr09:11759335-11759836</td><td>353</td><td>366</td><td>-</td><td>10.18290</td><td>8.69e-05</td><td>0.0891</td><td>AAAAAAGAGAGAGA       </td><td>Urofu.9G196500.v1.1</td><td>BASS                                                      </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>NA</td><td>scACR_6986;mesophyll     </td><td>Chr02:35964365-35964866</td><td>295</td><td>308</td><td>+</td><td>10.17070</td><td>8.75e-05</td><td>0.0895</td><td>GCGAGAGACGGACG       </td><td>Urofu.2G409400.v1.1</td><td>AlaAT2                                                    </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>E2FDP       </td><td>NA</td><td>scACR_9869;mesophyll     </td><td>Chr03:17293724-17294225</td><td>485</td><td>497</td><td>-</td><td>10.57320</td><td>9.00e-05</td><td>0.7030</td><td>TTGGGGCGGGCGA        </td><td>Urofu.3G289200.v1.1</td><td>PPDK                                                      </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>NA</td><td>scACR_24041;bundle_sheath</td><td>Chr08:29179180-29179681</td><td>240</td><td>252</td><td>-</td><td>10.43980</td><td>9.02e-05</td><td>0.7600</td><td>TCAGAAAAGGAAA        </td><td>Urofu.8G225900.v1.1</td><td>RUBISCO_activase1                                         </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>NAC         </td><td>NA</td><td>scACR_18603;bundle_sheath</td><td>Chr06:14687042-14687543</td><td>352</td><td>372</td><td>+</td><td> 9.83673</td><td>9.18e-05</td><td>1.0000</td><td>TTCCTTGTGAAATACGGGATA</td><td>Urofu.6G101400.v1.1</td><td>PPT                                                       </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>NA</td><td>scACR_5523;mesophyll     </td><td>Chr02:25923685-25924186</td><td>348</td><td>362</td><td>-</td><td>10.46670</td><td>9.30e-05</td><td>0.2210</td><td>CCGACGGCGACGGCG      </td><td>Urofu.2G245100.v1.1</td><td>DIT1                                                      </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>NA</td><td>scACR_4503;mesophyll     </td><td>Chr02:11507082-11507583</td><td>330</td><td>342</td><td>+</td><td>10.37950</td><td>9.43e-05</td><td>0.7700</td><td>CCAATATTAGAAC        </td><td>Urofu.2G135100.v1.1</td><td>NHD1__NHD2                                                </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>NA</td><td>scACR_17712;bundle_sheath</td><td>Chr05:36445199-36445700</td><td>359</td><td>371</td><td>-</td><td>10.37950</td><td>9.43e-05</td><td>0.7700</td><td>ACGAAAATGGAAT        </td><td>Urofu.5G464900.v1.1</td><td>NADP_ME1__NADP-ME_1_ortho_count_1__NADP-ME_2_ortho_count_1</td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>BES1        </td><td>NA</td><td>scACR_19222;bundle_sheath</td><td>Chr06:21745880-21746381</td><td> 86</td><td> 95</td><td>+</td><td>10.12200</td><td>9.49e-05</td><td>0.5000</td><td>CCCACGTGTA           </td><td>Urofu.6G175300.v1.1</td><td>DIT2                                                      </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>E2FDP       </td><td>NA</td><td>scACR_9870;mesophyll     </td><td>Chr03:17294852-17295353</td><td>190</td><td>202</td><td>-</td><td>10.47560</td><td>9.52e-05</td><td>0.7030</td><td>TTGTGGCGGCTCC        </td><td>Urofu.3G289200.v1.1</td><td>PPDK                                                      </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>HD          </td><td>NA</td><td>scACR_9874;mesophyll     </td><td>Chr03:17302413-17302914</td><td> 57</td><td> 63</td><td>-</td><td>12.09520</td><td>9.66e-05</td><td>1.0000</td><td>CAATAAT              </td><td>Urofu.3G289200.v1.1</td><td>PPDK                                                      </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>HD          </td><td>NA</td><td>scACR_7066;mesophyll     </td><td>Chr02:36483775-36484276</td><td>166</td><td>172</td><td>-</td><td>12.09520</td><td>9.66e-05</td><td>1.0000</td><td>CAATAAT              </td><td>Urofu.2G419400.v1.1</td><td>NAD_ME10                                                  </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>HD          </td><td>NA</td><td>scACR_2712;mesophyll     </td><td>Chr01:30173454-30173955</td><td>290</td><td>296</td><td>-</td><td>12.09520</td><td>9.66e-05</td><td>1.0000</td><td>CAATAAT              </td><td>Urofu.1G324800.v1.1</td><td>ASP_AT1                                                   </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>HD          </td><td>NA</td><td>scACR_10428;bundle_sheath</td><td>Chr03:27708920-27709421</td><td>294</td><td>300</td><td>-</td><td>12.09520</td><td>9.66e-05</td><td>1.0000</td><td>CAATAAT              </td><td>Urofu.3G359400.v1.1</td><td>RBCA2_ortho_count_3__RBCA1_ortho_count_3                  </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>HD          </td><td>NA</td><td>scACR_10428;bundle_sheath</td><td>Chr03:27708920-27709421</td><td>338</td><td>344</td><td>+</td><td>12.09520</td><td>9.66e-05</td><td>1.0000</td><td>CAATAAT              </td><td>Urofu.3G359400.v1.1</td><td>RBCA2_ortho_count_3__RBCA1_ortho_count_3                  </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>HD          </td><td>NA</td><td>scACR_10427;bundle_sheath</td><td>Chr03:27708345-27708846</td><td>411</td><td>417</td><td>+</td><td>12.09520</td><td>9.66e-05</td><td>1.0000</td><td>CAATAAT              </td><td>Urofu.3G359400.v1.1</td><td>RBCA2_ortho_count_3__RBCA1_ortho_count_3                  </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>HD          </td><td>NA</td><td>scACR_28334;bundle_sheath</td><td>Chr09:39461270-39461771</td><td>462</td><td>468</td><td>+</td><td>12.09520</td><td>9.66e-05</td><td>1.0000</td><td>CAATAAT              </td><td>Urofu.9G489700.v1.1</td><td>PEPCK                                                     </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>NA</td><td>scACR_6986;mesophyll     </td><td>Chr02:35964365-35964866</td><td>213</td><td>227</td><td>+</td><td>10.38890</td><td>9.76e-05</td><td>0.2230</td><td>CCGTCAAATCTGCAG      </td><td>Urofu.2G409400.v1.1</td><td>AlaAT2                                                    </td><td>ms</td><td>uf</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>NA</td><td>scACR_28334;bundle_sheath</td><td>Chr09:39461270-39461771</td><td> 34</td><td> 48</td><td>+</td><td>10.37780</td><td>9.83e-05</td><td>0.2230</td><td>TCGTCGACGCCGCCG      </td><td>Urofu.9G489700.v1.1</td><td>PEPCK                                                     </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>NA</td><td>scACR_28333;bundle_sheath</td><td>Chr09:39458150-39458651</td><td>451</td><td>465</td><td>+</td><td>10.37780</td><td>9.83e-05</td><td>0.2230</td><td>CCGCGATGGCCGCCA      </td><td>Urofu.9G489700.v1.1</td><td>PEPCK                                                     </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>NAC         </td><td>NA</td><td>scACR_28337;bundle_sheath</td><td>Chr09:39468323-39468824</td><td> 78</td><td> 98</td><td>-</td><td> 9.67347</td><td>9.94e-05</td><td>1.0000</td><td>TGCCTTACTGGGGACGAAAAA</td><td>Urofu.9G489700.v1.1</td><td>PEPCK                                                     </td><td>bs</td><td>uf</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>NA</td><td>scACR_9870;mesophyll     </td><td>Chr03:17294852-17295353</td><td>194</td><td>208</td><td>+</td><td>10.35560</td><td>9.96e-05</td><td>0.2230</td><td>CCGCCACAACAGCCG      </td><td>Urofu.3G289200.v1.1</td><td>PPDK                                                      </td><td>ms</td><td>uf</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 707 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " motif\\_id & motif\\_alt\\_id & acrID & region & start & stop & strand & score & p-value & q-value & matched\\_sequence & gene\\_id & gene\\_name & type & species\\\\\n",
       " <chr> & <lgl> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t HSF         & NA & scACR\\_5926;mesophyll      & chr1:226422323-226422824 & 450 & 463 & - & 21.5688 & 1.46e-08 & 0.06570 & AGAAGGTTCTAGAA  & Zm00001eb042970 & PPT               & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_42073;mesophyll     & chr6:177097950-177098451 & 349 & 362 & - & 16.5854 & 9.56e-07 & 0.00282 & AAGAGAGAGAAGGA  & Zm00001eb296550 & NAD-MDH2          & ms & zm\\\\\n",
       "\t MADS\\_MIKC   & NA & scACR\\_1560;mesophyll      & chr1:35695232-35695733   & 428 & 440 & + & 14.9458 & 1.22e-06 & 0.56700 & CCCAAAATGGTAA   & Zm00001eb011080 & PEPCK             & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_42073;mesophyll     & chr6:177097950-177098451 & 353 & 366 & - & 16.1829 & 1.56e-06 & 0.00431 & AAGAAAGAGAGAGA  & Zm00001eb296550 & NAD-MDH2          & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_42073;mesophyll     & chr6:177097950-177098451 & 347 & 360 & - & 16.1098 & 1.70e-06 & 0.00465 & GAGAGAGAAGGAGA  & Zm00001eb296550 & NAD-MDH2          & ms & zm\\\\\n",
       "\t BES1        & NA & scACR\\_51667;bundle\\_sheath & chr8:181465456-181465957 & 175 & 184 & - & 15.9024 & 2.10e-06 & 0.35600 & TGCACGTGTG      & Zm00001eb371140 & NADP\\_ME           & bs & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_8033;mesophyll      & chr1:289348515-289349016 & 237 & 250 & - & 15.3293 & 3.26e-06 & 0.00790 & GGGAGAGAGGCAGA  & Zm00001eb059070 & NAD-MDH12         & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_24813;mesophyll     & chr3:220994969-220995470 & 364 & 377 & - & 15.0732 & 3.74e-06 & 0.00877 & GCGAGAGAGCGCGA  & Zm00001eb158800 & CA2               & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_24813;mesophyll     & chr3:220994969-220995470 & 364 & 377 & - & 15.0732 & 3.74e-06 & 0.00877 & GCGAGAGAGCGCGA  & Zm00001eb158810 & CA1               & ms & zm\\\\\n",
       "\t MADS\\_MIKC   & NA & scACR\\_1560;mesophyll      & chr1:35695232-35695733   &  20 &  32 & + & 13.6747 & 4.88e-06 & 0.67100 & CCAAAAATGGTTA   & Zm00001eb011080 & PEPCK             & ms & zm\\\\\n",
       "\t MADS\\_MIKC   & NA & scACR\\_50193;mesophyll     & chr8:150379201-150379702 &  37 &  49 & + & 13.1928 & 7.96e-06 & 0.69900 & CCAAAAAAAGATA   & Zm00001eb359190 & CA6               & ms & zm\\\\\n",
       "\t bHLH\\_TCP    & NA & scACR\\_1560;mesophyll      & chr1:35695232-35695733   & 287 & 294 & + & 14.6633 & 8.13e-06 & 0.13300 & GGGCCCAC        & Zm00001eb011080 & PEPCK             & ms & zm\\\\\n",
       "\t AS2LOB      & NA & scACR\\_24813;mesophyll     & chr3:220994969-220995470 &  90 & 104 & - & 13.8444 & 8.48e-06 & 0.21700 & CCGCATCCACCGGCG & Zm00001eb158800 & CA2               & ms & zm\\\\\n",
       "\t AS2LOB      & NA & scACR\\_24813;mesophyll     & chr3:220994969-220995470 &  90 & 104 & - & 13.8444 & 8.48e-06 & 0.21700 & CCGCATCCACCGGCG & Zm00001eb158810 & CA1               & ms & zm\\\\\n",
       "\t HSF         & NA & scACR\\_5926;mesophyll      & chr1:226422323-226422824 & 454 & 467 & + & 13.4679 & 9.70e-06 & 0.65200 & AGAACCTTCTCGAG  & Zm00001eb042970 & PPT               & ms & zm\\\\\n",
       "\t AS2LOB      & NA & scACR\\_8034;bundle\\_sheath  & chr1:289354685-289355186 &  24 &  38 & + & 12.9889 & 1.66e-05 & 0.23000 & CCGCGGAAGCCGGAG & Zm00001eb059070 & NAD-MDH12         & bs & zm\\\\\n",
       "\t HSF         & NA & scACR\\_25494;bundle\\_sheath & chr4:1220891-1221392     & 317 & 330 & - & 12.4312 & 1.77e-05 & 0.77600 & TGAATATTCTAGAG  & Zm00001eb164390 & RUBISCO\\_activase1 & bs & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_16177;mesophyll     & chr2:147004111-147004612 & 226 & 239 & + & 12.7073 & 1.96e-05 & 0.03110 & TCGAGAGAGACGGC  & Zm00001eb092540 & RBCA2             & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_16177;mesophyll     & chr2:147004111-147004612 & 228 & 241 & + & 12.6341 & 2.07e-05 & 0.03240 & GAGAGAGACGGCGG  & Zm00001eb092540 & RBCA2             & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_29627;mesophyll     & chr4:196205013-196205514 & 330 & 343 & - & 12.6341 & 2.07e-05 & 0.03240 & GAGAGAGACGGCGG  & Zm00001eb197410 & RBCA1             & ms & zm\\\\\n",
       "\t HD-Zip\\_I\\_II & NA & scACR\\_40882;mesophyll     & chr6:157026520-157027021 &  44 &  51 & - & 12.5149 & 2.18e-05 & 1.00000 & CAATCATT        & Zm00001eb287770 & PPDK              & ms & zm\\\\\n",
       "\t HD-Zip\\_I\\_II & NA & scACR\\_29625;bundle\\_sheath & chr4:196176098-196176599 & 486 & 493 & + & 12.5149 & 2.18e-05 & 1.00000 & CAATCATT        & Zm00001eb197410 & RBCA1             & bs & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_42073;mesophyll     & chr6:177097950-177098451 & 351 & 364 & - & 12.5366 & 2.21e-05 & 0.03390 & GAAAGAGAGAGAAG  & Zm00001eb296550 & NAD-MDH2          & ms & zm\\\\\n",
       "\t AS2LOB      & NA & scACR\\_50193;mesophyll     & chr8:150379201-150379702 & 436 & 450 & + & 12.5667 & 2.28e-05 & 0.24800 & CCGTCATCTTCGCCA & Zm00001eb359190 & CA6               & ms & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_29627;mesophyll     & chr4:196205013-196205514 & 332 & 345 & - & 12.4878 & 2.29e-05 & 0.03490 & TTGAGAGAGACGGC  & Zm00001eb197410 & RBCA1             & ms & zm\\\\\n",
       "\t bHLH\\_TCP    & NA & scACR\\_5926;mesophyll      & chr1:226422323-226422824 & 187 & 194 & + & 12.3571 & 2.62e-05 & 0.19900 & GGCCCCAC        & Zm00001eb042970 & PPT               & ms & zm\\\\\n",
       "\t bHLH\\_TCP    & NA & scACR\\_25495;bundle\\_sheath & chr4:1225196-1225697     & 185 & 192 & - & 12.3571 & 2.62e-05 & 0.19900 & GGCCCCAC        & Zm00001eb164390 & RUBISCO\\_activase1 & bs & zm\\\\\n",
       "\t bHLH\\_TCP    & NA & scACR\\_25494;bundle\\_sheath & chr4:1220891-1221392     & 342 & 349 & - & 12.3571 & 2.62e-05 & 0.19900 & GGCCCCAC        & Zm00001eb164390 & RUBISCO\\_activase1 & bs & zm\\\\\n",
       "\t BBRBPC      & NA & scACR\\_8033;mesophyll      & chr1:289348515-289349016 & 235 & 248 & - & 12.1585 & 2.87e-05 & 0.04140 & GAGAGAGGCAGAGG  & Zm00001eb059070 & NAD-MDH12         & ms & zm\\\\\n",
       "\t AS2LOB      & NA & scACR\\_29627;mesophyll     & chr4:196205013-196205514 & 321 & 335 & + & 12.2000 & 2.97e-05 & 0.25900 & ACGCCGCATCCGCCG & Zm00001eb197410 & RBCA1             & ms & zm\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t GARP\\_G2-like & NA & scACR\\_9874;mesophyll      & Chr03:17302413-17302914 & 464 & 470 & + & 10.01550 & 7.93e-05 & 1.0000 & AAGATTC               & Urofu.3G289200.v1.1 & PPDK                                                       & ms & uf\\\\\n",
       "\t E2FDP        & NA & scACR\\_24584;bundle\\_sheath & Chr09:3736455-3736956   & 375 & 387 & - & 10.75610 & 8.08e-05 & 0.7030 & ATTTGGCGCGGGG         & Urofu.9G071000.v1.1 & NAD-MDH12                                                  & bs & uf\\\\\n",
       "\t NAC          & NA & scACR\\_25711;bundle\\_sheath & Chr09:11759335-11759836 &  15 &  35 & + & 10.08160 & 8.14e-05 & 1.0000 & TGACTTATATTAGACGTTAAG & Urofu.9G196500.v1.1 & BASS                                                       & bs & uf\\\\\n",
       "\t BBRBPC       & NA & scACR\\_19222;bundle\\_sheath & Chr06:21745880-21746381 & 468 & 481 & - & 10.28050 & 8.23e-05 & 0.0870 & GAGAAAGGGAAAGG        & Urofu.6G175300.v1.1 & DIT2                                                       & bs & uf\\\\\n",
       "\t bHLH\\_TCP     & NA & scACR\\_10421;bundle\\_sheath & Chr03:27572579-27573080 & 274 & 281 & + & 11.03060 & 8.23e-05 & 0.2040 & GGGCCCAA              & Urofu.3G358900.v1.1 & RBCA2\\_ortho\\_count\\_2\\_\\_RBCA1\\_ortho\\_count\\_2                   & bs & uf\\\\\n",
       "\t BBRBPC       & NA & scACR\\_5528;mesophyll      & Chr02:25948078-25948579 & 304 & 317 & - & 10.25610 & 8.34e-05 & 0.0875 & GAGAAAGAACGCGA        & Urofu.2G245100.v1.1 & DIT1                                                       & ms & uf\\\\\n",
       "\t NAC          & NA & scACR\\_18603;bundle\\_sheath & Chr06:14687042-14687543 & 350 & 370 & - &  9.98980 & 8.51e-05 & 1.0000 & TCCCGTATTTCACAAGGAAAA & Urofu.6G101400.v1.1 & PPT                                                        & bs & uf\\\\\n",
       "\t BBRBPC       & NA & scACR\\_19222;bundle\\_sheath & Chr06:21745880-21746381 & 230 & 243 & + & 10.20730 & 8.58e-05 & 0.0885 & GAGAGAGGGCTAGG        & Urofu.6G175300.v1.1 & DIT2                                                       & bs & uf\\\\\n",
       "\t BBRBPC       & NA & scACR\\_25711;bundle\\_sheath & Chr09:11759335-11759836 & 353 & 366 & - & 10.18290 & 8.69e-05 & 0.0891 & AAAAAAGAGAGAGA        & Urofu.9G196500.v1.1 & BASS                                                       & bs & uf\\\\\n",
       "\t BBRBPC       & NA & scACR\\_6986;mesophyll      & Chr02:35964365-35964866 & 295 & 308 & + & 10.17070 & 8.75e-05 & 0.0895 & GCGAGAGACGGACG        & Urofu.2G409400.v1.1 & AlaAT2                                                     & ms & uf\\\\\n",
       "\t E2FDP        & NA & scACR\\_9869;mesophyll      & Chr03:17293724-17294225 & 485 & 497 & - & 10.57320 & 9.00e-05 & 0.7030 & TTGGGGCGGGCGA         & Urofu.3G289200.v1.1 & PPDK                                                       & ms & uf\\\\\n",
       "\t MADS\\_MIKC    & NA & scACR\\_24041;bundle\\_sheath & Chr08:29179180-29179681 & 240 & 252 & - & 10.43980 & 9.02e-05 & 0.7600 & TCAGAAAAGGAAA         & Urofu.8G225900.v1.1 & RUBISCO\\_activase1                                          & bs & uf\\\\\n",
       "\t NAC          & NA & scACR\\_18603;bundle\\_sheath & Chr06:14687042-14687543 & 352 & 372 & + &  9.83673 & 9.18e-05 & 1.0000 & TTCCTTGTGAAATACGGGATA & Urofu.6G101400.v1.1 & PPT                                                        & bs & uf\\\\\n",
       "\t AS2LOB       & NA & scACR\\_5523;mesophyll      & Chr02:25923685-25924186 & 348 & 362 & - & 10.46670 & 9.30e-05 & 0.2210 & CCGACGGCGACGGCG       & Urofu.2G245100.v1.1 & DIT1                                                       & ms & uf\\\\\n",
       "\t MADS\\_MIKC    & NA & scACR\\_4503;mesophyll      & Chr02:11507082-11507583 & 330 & 342 & + & 10.37950 & 9.43e-05 & 0.7700 & CCAATATTAGAAC         & Urofu.2G135100.v1.1 & NHD1\\_\\_NHD2                                                 & ms & uf\\\\\n",
       "\t MADS\\_MIKC    & NA & scACR\\_17712;bundle\\_sheath & Chr05:36445199-36445700 & 359 & 371 & - & 10.37950 & 9.43e-05 & 0.7700 & ACGAAAATGGAAT         & Urofu.5G464900.v1.1 & NADP\\_ME1\\_\\_NADP-ME\\_1\\_ortho\\_count\\_1\\_\\_NADP-ME\\_2\\_ortho\\_count\\_1 & bs & uf\\\\\n",
       "\t BES1         & NA & scACR\\_19222;bundle\\_sheath & Chr06:21745880-21746381 &  86 &  95 & + & 10.12200 & 9.49e-05 & 0.5000 & CCCACGTGTA            & Urofu.6G175300.v1.1 & DIT2                                                       & bs & uf\\\\\n",
       "\t E2FDP        & NA & scACR\\_9870;mesophyll      & Chr03:17294852-17295353 & 190 & 202 & - & 10.47560 & 9.52e-05 & 0.7030 & TTGTGGCGGCTCC         & Urofu.3G289200.v1.1 & PPDK                                                       & ms & uf\\\\\n",
       "\t HD           & NA & scACR\\_9874;mesophyll      & Chr03:17302413-17302914 &  57 &  63 & - & 12.09520 & 9.66e-05 & 1.0000 & CAATAAT               & Urofu.3G289200.v1.1 & PPDK                                                       & ms & uf\\\\\n",
       "\t HD           & NA & scACR\\_7066;mesophyll      & Chr02:36483775-36484276 & 166 & 172 & - & 12.09520 & 9.66e-05 & 1.0000 & CAATAAT               & Urofu.2G419400.v1.1 & NAD\\_ME10                                                   & ms & uf\\\\\n",
       "\t HD           & NA & scACR\\_2712;mesophyll      & Chr01:30173454-30173955 & 290 & 296 & - & 12.09520 & 9.66e-05 & 1.0000 & CAATAAT               & Urofu.1G324800.v1.1 & ASP\\_AT1                                                    & ms & uf\\\\\n",
       "\t HD           & NA & scACR\\_10428;bundle\\_sheath & Chr03:27708920-27709421 & 294 & 300 & - & 12.09520 & 9.66e-05 & 1.0000 & CAATAAT               & Urofu.3G359400.v1.1 & RBCA2\\_ortho\\_count\\_3\\_\\_RBCA1\\_ortho\\_count\\_3                   & bs & uf\\\\\n",
       "\t HD           & NA & scACR\\_10428;bundle\\_sheath & Chr03:27708920-27709421 & 338 & 344 & + & 12.09520 & 9.66e-05 & 1.0000 & CAATAAT               & Urofu.3G359400.v1.1 & RBCA2\\_ortho\\_count\\_3\\_\\_RBCA1\\_ortho\\_count\\_3                   & bs & uf\\\\\n",
       "\t HD           & NA & scACR\\_10427;bundle\\_sheath & Chr03:27708345-27708846 & 411 & 417 & + & 12.09520 & 9.66e-05 & 1.0000 & CAATAAT               & Urofu.3G359400.v1.1 & RBCA2\\_ortho\\_count\\_3\\_\\_RBCA1\\_ortho\\_count\\_3                   & bs & uf\\\\\n",
       "\t HD           & NA & scACR\\_28334;bundle\\_sheath & Chr09:39461270-39461771 & 462 & 468 & + & 12.09520 & 9.66e-05 & 1.0000 & CAATAAT               & Urofu.9G489700.v1.1 & PEPCK                                                      & bs & uf\\\\\n",
       "\t AS2LOB       & NA & scACR\\_6986;mesophyll      & Chr02:35964365-35964866 & 213 & 227 & + & 10.38890 & 9.76e-05 & 0.2230 & CCGTCAAATCTGCAG       & Urofu.2G409400.v1.1 & AlaAT2                                                     & ms & uf\\\\\n",
       "\t AS2LOB       & NA & scACR\\_28334;bundle\\_sheath & Chr09:39461270-39461771 &  34 &  48 & + & 10.37780 & 9.83e-05 & 0.2230 & TCGTCGACGCCGCCG       & Urofu.9G489700.v1.1 & PEPCK                                                      & bs & uf\\\\\n",
       "\t AS2LOB       & NA & scACR\\_28333;bundle\\_sheath & Chr09:39458150-39458651 & 451 & 465 & + & 10.37780 & 9.83e-05 & 0.2230 & CCGCGATGGCCGCCA       & Urofu.9G489700.v1.1 & PEPCK                                                      & bs & uf\\\\\n",
       "\t NAC          & NA & scACR\\_28337;bundle\\_sheath & Chr09:39468323-39468824 &  78 &  98 & - &  9.67347 & 9.94e-05 & 1.0000 & TGCCTTACTGGGGACGAAAAA & Urofu.9G489700.v1.1 & PEPCK                                                      & bs & uf\\\\\n",
       "\t AS2LOB       & NA & scACR\\_9870;mesophyll      & Chr03:17294852-17295353 & 194 & 208 & + & 10.35560 & 9.96e-05 & 0.2230 & CCGCCACAACAGCCG       & Urofu.3G289200.v1.1 & PPDK                                                       & ms & uf\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 707 × 15\n",
       "\n",
       "| motif_id &lt;chr&gt; | motif_alt_id &lt;lgl&gt; | acrID &lt;chr&gt; | region &lt;chr&gt; | start &lt;dbl&gt; | stop &lt;dbl&gt; | strand &lt;chr&gt; | score &lt;dbl&gt; | p-value &lt;dbl&gt; | q-value &lt;dbl&gt; | matched_sequence &lt;chr&gt; | gene_id &lt;chr&gt; | gene_name &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HSF         | NA | scACR_5926;mesophyll      | chr1:226422323-226422824 | 450 | 463 | - | 21.5688 | 1.46e-08 | 0.06570 | AGAAGGTTCTAGAA  | Zm00001eb042970 | PPT               | ms | zm |\n",
       "| BBRBPC      | NA | scACR_42073;mesophyll     | chr6:177097950-177098451 | 349 | 362 | - | 16.5854 | 9.56e-07 | 0.00282 | AAGAGAGAGAAGGA  | Zm00001eb296550 | NAD-MDH2          | ms | zm |\n",
       "| MADS_MIKC   | NA | scACR_1560;mesophyll      | chr1:35695232-35695733   | 428 | 440 | + | 14.9458 | 1.22e-06 | 0.56700 | CCCAAAATGGTAA   | Zm00001eb011080 | PEPCK             | ms | zm |\n",
       "| BBRBPC      | NA | scACR_42073;mesophyll     | chr6:177097950-177098451 | 353 | 366 | - | 16.1829 | 1.56e-06 | 0.00431 | AAGAAAGAGAGAGA  | Zm00001eb296550 | NAD-MDH2          | ms | zm |\n",
       "| BBRBPC      | NA | scACR_42073;mesophyll     | chr6:177097950-177098451 | 347 | 360 | - | 16.1098 | 1.70e-06 | 0.00465 | GAGAGAGAAGGAGA  | Zm00001eb296550 | NAD-MDH2          | ms | zm |\n",
       "| BES1        | NA | scACR_51667;bundle_sheath | chr8:181465456-181465957 | 175 | 184 | - | 15.9024 | 2.10e-06 | 0.35600 | TGCACGTGTG      | Zm00001eb371140 | NADP_ME           | bs | zm |\n",
       "| BBRBPC      | NA | scACR_8033;mesophyll      | chr1:289348515-289349016 | 237 | 250 | - | 15.3293 | 3.26e-06 | 0.00790 | GGGAGAGAGGCAGA  | Zm00001eb059070 | NAD-MDH12         | ms | zm |\n",
       "| BBRBPC      | NA | scACR_24813;mesophyll     | chr3:220994969-220995470 | 364 | 377 | - | 15.0732 | 3.74e-06 | 0.00877 | GCGAGAGAGCGCGA  | Zm00001eb158800 | CA2               | ms | zm |\n",
       "| BBRBPC      | NA | scACR_24813;mesophyll     | chr3:220994969-220995470 | 364 | 377 | - | 15.0732 | 3.74e-06 | 0.00877 | GCGAGAGAGCGCGA  | Zm00001eb158810 | CA1               | ms | zm |\n",
       "| MADS_MIKC   | NA | scACR_1560;mesophyll      | chr1:35695232-35695733   |  20 |  32 | + | 13.6747 | 4.88e-06 | 0.67100 | CCAAAAATGGTTA   | Zm00001eb011080 | PEPCK             | ms | zm |\n",
       "| MADS_MIKC   | NA | scACR_50193;mesophyll     | chr8:150379201-150379702 |  37 |  49 | + | 13.1928 | 7.96e-06 | 0.69900 | CCAAAAAAAGATA   | Zm00001eb359190 | CA6               | ms | zm |\n",
       "| bHLH_TCP    | NA | scACR_1560;mesophyll      | chr1:35695232-35695733   | 287 | 294 | + | 14.6633 | 8.13e-06 | 0.13300 | GGGCCCAC        | Zm00001eb011080 | PEPCK             | ms | zm |\n",
       "| AS2LOB      | NA | scACR_24813;mesophyll     | chr3:220994969-220995470 |  90 | 104 | - | 13.8444 | 8.48e-06 | 0.21700 | CCGCATCCACCGGCG | Zm00001eb158800 | CA2               | ms | zm |\n",
       "| AS2LOB      | NA | scACR_24813;mesophyll     | chr3:220994969-220995470 |  90 | 104 | - | 13.8444 | 8.48e-06 | 0.21700 | CCGCATCCACCGGCG | Zm00001eb158810 | CA1               | ms | zm |\n",
       "| HSF         | NA | scACR_5926;mesophyll      | chr1:226422323-226422824 | 454 | 467 | + | 13.4679 | 9.70e-06 | 0.65200 | AGAACCTTCTCGAG  | Zm00001eb042970 | PPT               | ms | zm |\n",
       "| AS2LOB      | NA | scACR_8034;bundle_sheath  | chr1:289354685-289355186 |  24 |  38 | + | 12.9889 | 1.66e-05 | 0.23000 | CCGCGGAAGCCGGAG | Zm00001eb059070 | NAD-MDH12         | bs | zm |\n",
       "| HSF         | NA | scACR_25494;bundle_sheath | chr4:1220891-1221392     | 317 | 330 | - | 12.4312 | 1.77e-05 | 0.77600 | TGAATATTCTAGAG  | Zm00001eb164390 | RUBISCO_activase1 | bs | zm |\n",
       "| BBRBPC      | NA | scACR_16177;mesophyll     | chr2:147004111-147004612 | 226 | 239 | + | 12.7073 | 1.96e-05 | 0.03110 | TCGAGAGAGACGGC  | Zm00001eb092540 | RBCA2             | ms | zm |\n",
       "| BBRBPC      | NA | scACR_16177;mesophyll     | chr2:147004111-147004612 | 228 | 241 | + | 12.6341 | 2.07e-05 | 0.03240 | GAGAGAGACGGCGG  | Zm00001eb092540 | RBCA2             | ms | zm |\n",
       "| BBRBPC      | NA | scACR_29627;mesophyll     | chr4:196205013-196205514 | 330 | 343 | - | 12.6341 | 2.07e-05 | 0.03240 | GAGAGAGACGGCGG  | Zm00001eb197410 | RBCA1             | ms | zm |\n",
       "| HD-Zip_I_II | NA | scACR_40882;mesophyll     | chr6:157026520-157027021 |  44 |  51 | - | 12.5149 | 2.18e-05 | 1.00000 | CAATCATT        | Zm00001eb287770 | PPDK              | ms | zm |\n",
       "| HD-Zip_I_II | NA | scACR_29625;bundle_sheath | chr4:196176098-196176599 | 486 | 493 | + | 12.5149 | 2.18e-05 | 1.00000 | CAATCATT        | Zm00001eb197410 | RBCA1             | bs | zm |\n",
       "| BBRBPC      | NA | scACR_42073;mesophyll     | chr6:177097950-177098451 | 351 | 364 | - | 12.5366 | 2.21e-05 | 0.03390 | GAAAGAGAGAGAAG  | Zm00001eb296550 | NAD-MDH2          | ms | zm |\n",
       "| AS2LOB      | NA | scACR_50193;mesophyll     | chr8:150379201-150379702 | 436 | 450 | + | 12.5667 | 2.28e-05 | 0.24800 | CCGTCATCTTCGCCA | Zm00001eb359190 | CA6               | ms | zm |\n",
       "| BBRBPC      | NA | scACR_29627;mesophyll     | chr4:196205013-196205514 | 332 | 345 | - | 12.4878 | 2.29e-05 | 0.03490 | TTGAGAGAGACGGC  | Zm00001eb197410 | RBCA1             | ms | zm |\n",
       "| bHLH_TCP    | NA | scACR_5926;mesophyll      | chr1:226422323-226422824 | 187 | 194 | + | 12.3571 | 2.62e-05 | 0.19900 | GGCCCCAC        | Zm00001eb042970 | PPT               | ms | zm |\n",
       "| bHLH_TCP    | NA | scACR_25495;bundle_sheath | chr4:1225196-1225697     | 185 | 192 | - | 12.3571 | 2.62e-05 | 0.19900 | GGCCCCAC        | Zm00001eb164390 | RUBISCO_activase1 | bs | zm |\n",
       "| bHLH_TCP    | NA | scACR_25494;bundle_sheath | chr4:1220891-1221392     | 342 | 349 | - | 12.3571 | 2.62e-05 | 0.19900 | GGCCCCAC        | Zm00001eb164390 | RUBISCO_activase1 | bs | zm |\n",
       "| BBRBPC      | NA | scACR_8033;mesophyll      | chr1:289348515-289349016 | 235 | 248 | - | 12.1585 | 2.87e-05 | 0.04140 | GAGAGAGGCAGAGG  | Zm00001eb059070 | NAD-MDH12         | ms | zm |\n",
       "| AS2LOB      | NA | scACR_29627;mesophyll     | chr4:196205013-196205514 | 321 | 335 | + | 12.2000 | 2.97e-05 | 0.25900 | ACGCCGCATCCGCCG | Zm00001eb197410 | RBCA1             | ms | zm |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| GARP_G2-like | NA | scACR_9874;mesophyll      | Chr03:17302413-17302914 | 464 | 470 | + | 10.01550 | 7.93e-05 | 1.0000 | AAGATTC               | Urofu.3G289200.v1.1 | PPDK                                                       | ms | uf |\n",
       "| E2FDP        | NA | scACR_24584;bundle_sheath | Chr09:3736455-3736956   | 375 | 387 | - | 10.75610 | 8.08e-05 | 0.7030 | ATTTGGCGCGGGG         | Urofu.9G071000.v1.1 | NAD-MDH12                                                  | bs | uf |\n",
       "| NAC          | NA | scACR_25711;bundle_sheath | Chr09:11759335-11759836 |  15 |  35 | + | 10.08160 | 8.14e-05 | 1.0000 | TGACTTATATTAGACGTTAAG | Urofu.9G196500.v1.1 | BASS                                                       | bs | uf |\n",
       "| BBRBPC       | NA | scACR_19222;bundle_sheath | Chr06:21745880-21746381 | 468 | 481 | - | 10.28050 | 8.23e-05 | 0.0870 | GAGAAAGGGAAAGG        | Urofu.6G175300.v1.1 | DIT2                                                       | bs | uf |\n",
       "| bHLH_TCP     | NA | scACR_10421;bundle_sheath | Chr03:27572579-27573080 | 274 | 281 | + | 11.03060 | 8.23e-05 | 0.2040 | GGGCCCAA              | Urofu.3G358900.v1.1 | RBCA2_ortho_count_2__RBCA1_ortho_count_2                   | bs | uf |\n",
       "| BBRBPC       | NA | scACR_5528;mesophyll      | Chr02:25948078-25948579 | 304 | 317 | - | 10.25610 | 8.34e-05 | 0.0875 | GAGAAAGAACGCGA        | Urofu.2G245100.v1.1 | DIT1                                                       | ms | uf |\n",
       "| NAC          | NA | scACR_18603;bundle_sheath | Chr06:14687042-14687543 | 350 | 370 | - |  9.98980 | 8.51e-05 | 1.0000 | TCCCGTATTTCACAAGGAAAA | Urofu.6G101400.v1.1 | PPT                                                        | bs | uf |\n",
       "| BBRBPC       | NA | scACR_19222;bundle_sheath | Chr06:21745880-21746381 | 230 | 243 | + | 10.20730 | 8.58e-05 | 0.0885 | GAGAGAGGGCTAGG        | Urofu.6G175300.v1.1 | DIT2                                                       | bs | uf |\n",
       "| BBRBPC       | NA | scACR_25711;bundle_sheath | Chr09:11759335-11759836 | 353 | 366 | - | 10.18290 | 8.69e-05 | 0.0891 | AAAAAAGAGAGAGA        | Urofu.9G196500.v1.1 | BASS                                                       | bs | uf |\n",
       "| BBRBPC       | NA | scACR_6986;mesophyll      | Chr02:35964365-35964866 | 295 | 308 | + | 10.17070 | 8.75e-05 | 0.0895 | GCGAGAGACGGACG        | Urofu.2G409400.v1.1 | AlaAT2                                                     | ms | uf |\n",
       "| E2FDP        | NA | scACR_9869;mesophyll      | Chr03:17293724-17294225 | 485 | 497 | - | 10.57320 | 9.00e-05 | 0.7030 | TTGGGGCGGGCGA         | Urofu.3G289200.v1.1 | PPDK                                                       | ms | uf |\n",
       "| MADS_MIKC    | NA | scACR_24041;bundle_sheath | Chr08:29179180-29179681 | 240 | 252 | - | 10.43980 | 9.02e-05 | 0.7600 | TCAGAAAAGGAAA         | Urofu.8G225900.v1.1 | RUBISCO_activase1                                          | bs | uf |\n",
       "| NAC          | NA | scACR_18603;bundle_sheath | Chr06:14687042-14687543 | 352 | 372 | + |  9.83673 | 9.18e-05 | 1.0000 | TTCCTTGTGAAATACGGGATA | Urofu.6G101400.v1.1 | PPT                                                        | bs | uf |\n",
       "| AS2LOB       | NA | scACR_5523;mesophyll      | Chr02:25923685-25924186 | 348 | 362 | - | 10.46670 | 9.30e-05 | 0.2210 | CCGACGGCGACGGCG       | Urofu.2G245100.v1.1 | DIT1                                                       | ms | uf |\n",
       "| MADS_MIKC    | NA | scACR_4503;mesophyll      | Chr02:11507082-11507583 | 330 | 342 | + | 10.37950 | 9.43e-05 | 0.7700 | CCAATATTAGAAC         | Urofu.2G135100.v1.1 | NHD1__NHD2                                                 | ms | uf |\n",
       "| MADS_MIKC    | NA | scACR_17712;bundle_sheath | Chr05:36445199-36445700 | 359 | 371 | - | 10.37950 | 9.43e-05 | 0.7700 | ACGAAAATGGAAT         | Urofu.5G464900.v1.1 | NADP_ME1__NADP-ME_1_ortho_count_1__NADP-ME_2_ortho_count_1 | bs | uf |\n",
       "| BES1         | NA | scACR_19222;bundle_sheath | Chr06:21745880-21746381 |  86 |  95 | + | 10.12200 | 9.49e-05 | 0.5000 | CCCACGTGTA            | Urofu.6G175300.v1.1 | DIT2                                                       | bs | uf |\n",
       "| E2FDP        | NA | scACR_9870;mesophyll      | Chr03:17294852-17295353 | 190 | 202 | - | 10.47560 | 9.52e-05 | 0.7030 | TTGTGGCGGCTCC         | Urofu.3G289200.v1.1 | PPDK                                                       | ms | uf |\n",
       "| HD           | NA | scACR_9874;mesophyll      | Chr03:17302413-17302914 |  57 |  63 | - | 12.09520 | 9.66e-05 | 1.0000 | CAATAAT               | Urofu.3G289200.v1.1 | PPDK                                                       | ms | uf |\n",
       "| HD           | NA | scACR_7066;mesophyll      | Chr02:36483775-36484276 | 166 | 172 | - | 12.09520 | 9.66e-05 | 1.0000 | CAATAAT               | Urofu.2G419400.v1.1 | NAD_ME10                                                   | ms | uf |\n",
       "| HD           | NA | scACR_2712;mesophyll      | Chr01:30173454-30173955 | 290 | 296 | - | 12.09520 | 9.66e-05 | 1.0000 | CAATAAT               | Urofu.1G324800.v1.1 | ASP_AT1                                                    | ms | uf |\n",
       "| HD           | NA | scACR_10428;bundle_sheath | Chr03:27708920-27709421 | 294 | 300 | - | 12.09520 | 9.66e-05 | 1.0000 | CAATAAT               | Urofu.3G359400.v1.1 | RBCA2_ortho_count_3__RBCA1_ortho_count_3                   | bs | uf |\n",
       "| HD           | NA | scACR_10428;bundle_sheath | Chr03:27708920-27709421 | 338 | 344 | + | 12.09520 | 9.66e-05 | 1.0000 | CAATAAT               | Urofu.3G359400.v1.1 | RBCA2_ortho_count_3__RBCA1_ortho_count_3                   | bs | uf |\n",
       "| HD           | NA | scACR_10427;bundle_sheath | Chr03:27708345-27708846 | 411 | 417 | + | 12.09520 | 9.66e-05 | 1.0000 | CAATAAT               | Urofu.3G359400.v1.1 | RBCA2_ortho_count_3__RBCA1_ortho_count_3                   | bs | uf |\n",
       "| HD           | NA | scACR_28334;bundle_sheath | Chr09:39461270-39461771 | 462 | 468 | + | 12.09520 | 9.66e-05 | 1.0000 | CAATAAT               | Urofu.9G489700.v1.1 | PEPCK                                                      | bs | uf |\n",
       "| AS2LOB       | NA | scACR_6986;mesophyll      | Chr02:35964365-35964866 | 213 | 227 | + | 10.38890 | 9.76e-05 | 0.2230 | CCGTCAAATCTGCAG       | Urofu.2G409400.v1.1 | AlaAT2                                                     | ms | uf |\n",
       "| AS2LOB       | NA | scACR_28334;bundle_sheath | Chr09:39461270-39461771 |  34 |  48 | + | 10.37780 | 9.83e-05 | 0.2230 | TCGTCGACGCCGCCG       | Urofu.9G489700.v1.1 | PEPCK                                                      | bs | uf |\n",
       "| AS2LOB       | NA | scACR_28333;bundle_sheath | Chr09:39458150-39458651 | 451 | 465 | + | 10.37780 | 9.83e-05 | 0.2230 | CCGCGATGGCCGCCA       | Urofu.9G489700.v1.1 | PEPCK                                                      | bs | uf |\n",
       "| NAC          | NA | scACR_28337;bundle_sheath | Chr09:39468323-39468824 |  78 |  98 | - |  9.67347 | 9.94e-05 | 1.0000 | TGCCTTACTGGGGACGAAAAA | Urofu.9G489700.v1.1 | PEPCK                                                      | bs | uf |\n",
       "| AS2LOB       | NA | scACR_9870;mesophyll      | Chr03:17294852-17295353 | 194 | 208 | + | 10.35560 | 9.96e-05 | 0.2230 | CCGCCACAACAGCCG       | Urofu.3G289200.v1.1 | PPDK                                                       | ms | uf |\n",
       "\n"
      ],
      "text/plain": [
       "    motif_id     motif_alt_id acrID                    \n",
       "1   HSF          NA           scACR_5926;mesophyll     \n",
       "2   BBRBPC       NA           scACR_42073;mesophyll    \n",
       "3   MADS_MIKC    NA           scACR_1560;mesophyll     \n",
       "4   BBRBPC       NA           scACR_42073;mesophyll    \n",
       "5   BBRBPC       NA           scACR_42073;mesophyll    \n",
       "6   BES1         NA           scACR_51667;bundle_sheath\n",
       "7   BBRBPC       NA           scACR_8033;mesophyll     \n",
       "8   BBRBPC       NA           scACR_24813;mesophyll    \n",
       "9   BBRBPC       NA           scACR_24813;mesophyll    \n",
       "10  MADS_MIKC    NA           scACR_1560;mesophyll     \n",
       "11  MADS_MIKC    NA           scACR_50193;mesophyll    \n",
       "12  bHLH_TCP     NA           scACR_1560;mesophyll     \n",
       "13  AS2LOB       NA           scACR_24813;mesophyll    \n",
       "14  AS2LOB       NA           scACR_24813;mesophyll    \n",
       "15  HSF          NA           scACR_5926;mesophyll     \n",
       "16  AS2LOB       NA           scACR_8034;bundle_sheath \n",
       "17  HSF          NA           scACR_25494;bundle_sheath\n",
       "18  BBRBPC       NA           scACR_16177;mesophyll    \n",
       "19  BBRBPC       NA           scACR_16177;mesophyll    \n",
       "20  BBRBPC       NA           scACR_29627;mesophyll    \n",
       "21  HD-Zip_I_II  NA           scACR_40882;mesophyll    \n",
       "22  HD-Zip_I_II  NA           scACR_29625;bundle_sheath\n",
       "23  BBRBPC       NA           scACR_42073;mesophyll    \n",
       "24  AS2LOB       NA           scACR_50193;mesophyll    \n",
       "25  BBRBPC       NA           scACR_29627;mesophyll    \n",
       "26  bHLH_TCP     NA           scACR_5926;mesophyll     \n",
       "27  bHLH_TCP     NA           scACR_25495;bundle_sheath\n",
       "28  bHLH_TCP     NA           scACR_25494;bundle_sheath\n",
       "29  BBRBPC       NA           scACR_8033;mesophyll     \n",
       "30  AS2LOB       NA           scACR_29627;mesophyll    \n",
       "⋮   ⋮            ⋮            ⋮                        \n",
       "678 GARP_G2-like NA           scACR_9874;mesophyll     \n",
       "679 E2FDP        NA           scACR_24584;bundle_sheath\n",
       "680 NAC          NA           scACR_25711;bundle_sheath\n",
       "681 BBRBPC       NA           scACR_19222;bundle_sheath\n",
       "682 bHLH_TCP     NA           scACR_10421;bundle_sheath\n",
       "683 BBRBPC       NA           scACR_5528;mesophyll     \n",
       "684 NAC          NA           scACR_18603;bundle_sheath\n",
       "685 BBRBPC       NA           scACR_19222;bundle_sheath\n",
       "686 BBRBPC       NA           scACR_25711;bundle_sheath\n",
       "687 BBRBPC       NA           scACR_6986;mesophyll     \n",
       "688 E2FDP        NA           scACR_9869;mesophyll     \n",
       "689 MADS_MIKC    NA           scACR_24041;bundle_sheath\n",
       "690 NAC          NA           scACR_18603;bundle_sheath\n",
       "691 AS2LOB       NA           scACR_5523;mesophyll     \n",
       "692 MADS_MIKC    NA           scACR_4503;mesophyll     \n",
       "693 MADS_MIKC    NA           scACR_17712;bundle_sheath\n",
       "694 BES1         NA           scACR_19222;bundle_sheath\n",
       "695 E2FDP        NA           scACR_9870;mesophyll     \n",
       "696 HD           NA           scACR_9874;mesophyll     \n",
       "697 HD           NA           scACR_7066;mesophyll     \n",
       "698 HD           NA           scACR_2712;mesophyll     \n",
       "699 HD           NA           scACR_10428;bundle_sheath\n",
       "700 HD           NA           scACR_10428;bundle_sheath\n",
       "701 HD           NA           scACR_10427;bundle_sheath\n",
       "702 HD           NA           scACR_28334;bundle_sheath\n",
       "703 AS2LOB       NA           scACR_6986;mesophyll     \n",
       "704 AS2LOB       NA           scACR_28334;bundle_sheath\n",
       "705 AS2LOB       NA           scACR_28333;bundle_sheath\n",
       "706 NAC          NA           scACR_28337;bundle_sheath\n",
       "707 AS2LOB       NA           scACR_9870;mesophyll     \n",
       "    region                   start stop strand score    p-value  q-value\n",
       "1   chr1:226422323-226422824 450   463  -      21.5688  1.46e-08 0.06570\n",
       "2   chr6:177097950-177098451 349   362  -      16.5854  9.56e-07 0.00282\n",
       "3   chr1:35695232-35695733   428   440  +      14.9458  1.22e-06 0.56700\n",
       "4   chr6:177097950-177098451 353   366  -      16.1829  1.56e-06 0.00431\n",
       "5   chr6:177097950-177098451 347   360  -      16.1098  1.70e-06 0.00465\n",
       "6   chr8:181465456-181465957 175   184  -      15.9024  2.10e-06 0.35600\n",
       "7   chr1:289348515-289349016 237   250  -      15.3293  3.26e-06 0.00790\n",
       "8   chr3:220994969-220995470 364   377  -      15.0732  3.74e-06 0.00877\n",
       "9   chr3:220994969-220995470 364   377  -      15.0732  3.74e-06 0.00877\n",
       "10  chr1:35695232-35695733    20    32  +      13.6747  4.88e-06 0.67100\n",
       "11  chr8:150379201-150379702  37    49  +      13.1928  7.96e-06 0.69900\n",
       "12  chr1:35695232-35695733   287   294  +      14.6633  8.13e-06 0.13300\n",
       "13  chr3:220994969-220995470  90   104  -      13.8444  8.48e-06 0.21700\n",
       "14  chr3:220994969-220995470  90   104  -      13.8444  8.48e-06 0.21700\n",
       "15  chr1:226422323-226422824 454   467  +      13.4679  9.70e-06 0.65200\n",
       "16  chr1:289354685-289355186  24    38  +      12.9889  1.66e-05 0.23000\n",
       "17  chr4:1220891-1221392     317   330  -      12.4312  1.77e-05 0.77600\n",
       "18  chr2:147004111-147004612 226   239  +      12.7073  1.96e-05 0.03110\n",
       "19  chr2:147004111-147004612 228   241  +      12.6341  2.07e-05 0.03240\n",
       "20  chr4:196205013-196205514 330   343  -      12.6341  2.07e-05 0.03240\n",
       "21  chr6:157026520-157027021  44    51  -      12.5149  2.18e-05 1.00000\n",
       "22  chr4:196176098-196176599 486   493  +      12.5149  2.18e-05 1.00000\n",
       "23  chr6:177097950-177098451 351   364  -      12.5366  2.21e-05 0.03390\n",
       "24  chr8:150379201-150379702 436   450  +      12.5667  2.28e-05 0.24800\n",
       "25  chr4:196205013-196205514 332   345  -      12.4878  2.29e-05 0.03490\n",
       "26  chr1:226422323-226422824 187   194  +      12.3571  2.62e-05 0.19900\n",
       "27  chr4:1225196-1225697     185   192  -      12.3571  2.62e-05 0.19900\n",
       "28  chr4:1220891-1221392     342   349  -      12.3571  2.62e-05 0.19900\n",
       "29  chr1:289348515-289349016 235   248  -      12.1585  2.87e-05 0.04140\n",
       "30  chr4:196205013-196205514 321   335  +      12.2000  2.97e-05 0.25900\n",
       "⋮   ⋮                        ⋮     ⋮    ⋮      ⋮        ⋮        ⋮      \n",
       "678 Chr03:17302413-17302914  464   470  +      10.01550 7.93e-05 1.0000 \n",
       "679 Chr09:3736455-3736956    375   387  -      10.75610 8.08e-05 0.7030 \n",
       "680 Chr09:11759335-11759836   15    35  +      10.08160 8.14e-05 1.0000 \n",
       "681 Chr06:21745880-21746381  468   481  -      10.28050 8.23e-05 0.0870 \n",
       "682 Chr03:27572579-27573080  274   281  +      11.03060 8.23e-05 0.2040 \n",
       "683 Chr02:25948078-25948579  304   317  -      10.25610 8.34e-05 0.0875 \n",
       "684 Chr06:14687042-14687543  350   370  -       9.98980 8.51e-05 1.0000 \n",
       "685 Chr06:21745880-21746381  230   243  +      10.20730 8.58e-05 0.0885 \n",
       "686 Chr09:11759335-11759836  353   366  -      10.18290 8.69e-05 0.0891 \n",
       "687 Chr02:35964365-35964866  295   308  +      10.17070 8.75e-05 0.0895 \n",
       "688 Chr03:17293724-17294225  485   497  -      10.57320 9.00e-05 0.7030 \n",
       "689 Chr08:29179180-29179681  240   252  -      10.43980 9.02e-05 0.7600 \n",
       "690 Chr06:14687042-14687543  352   372  +       9.83673 9.18e-05 1.0000 \n",
       "691 Chr02:25923685-25924186  348   362  -      10.46670 9.30e-05 0.2210 \n",
       "692 Chr02:11507082-11507583  330   342  +      10.37950 9.43e-05 0.7700 \n",
       "693 Chr05:36445199-36445700  359   371  -      10.37950 9.43e-05 0.7700 \n",
       "694 Chr06:21745880-21746381   86    95  +      10.12200 9.49e-05 0.5000 \n",
       "695 Chr03:17294852-17295353  190   202  -      10.47560 9.52e-05 0.7030 \n",
       "696 Chr03:17302413-17302914   57    63  -      12.09520 9.66e-05 1.0000 \n",
       "697 Chr02:36483775-36484276  166   172  -      12.09520 9.66e-05 1.0000 \n",
       "698 Chr01:30173454-30173955  290   296  -      12.09520 9.66e-05 1.0000 \n",
       "699 Chr03:27708920-27709421  294   300  -      12.09520 9.66e-05 1.0000 \n",
       "700 Chr03:27708920-27709421  338   344  +      12.09520 9.66e-05 1.0000 \n",
       "701 Chr03:27708345-27708846  411   417  +      12.09520 9.66e-05 1.0000 \n",
       "702 Chr09:39461270-39461771  462   468  +      12.09520 9.66e-05 1.0000 \n",
       "703 Chr02:35964365-35964866  213   227  +      10.38890 9.76e-05 0.2230 \n",
       "704 Chr09:39461270-39461771   34    48  +      10.37780 9.83e-05 0.2230 \n",
       "705 Chr09:39458150-39458651  451   465  +      10.37780 9.83e-05 0.2230 \n",
       "706 Chr09:39468323-39468824   78    98  -       9.67347 9.94e-05 1.0000 \n",
       "707 Chr03:17294852-17295353  194   208  +      10.35560 9.96e-05 0.2230 \n",
       "    matched_sequence      gene_id            \n",
       "1   AGAAGGTTCTAGAA        Zm00001eb042970    \n",
       "2   AAGAGAGAGAAGGA        Zm00001eb296550    \n",
       "3   CCCAAAATGGTAA         Zm00001eb011080    \n",
       "4   AAGAAAGAGAGAGA        Zm00001eb296550    \n",
       "5   GAGAGAGAAGGAGA        Zm00001eb296550    \n",
       "6   TGCACGTGTG            Zm00001eb371140    \n",
       "7   GGGAGAGAGGCAGA        Zm00001eb059070    \n",
       "8   GCGAGAGAGCGCGA        Zm00001eb158800    \n",
       "9   GCGAGAGAGCGCGA        Zm00001eb158810    \n",
       "10  CCAAAAATGGTTA         Zm00001eb011080    \n",
       "11  CCAAAAAAAGATA         Zm00001eb359190    \n",
       "12  GGGCCCAC              Zm00001eb011080    \n",
       "13  CCGCATCCACCGGCG       Zm00001eb158800    \n",
       "14  CCGCATCCACCGGCG       Zm00001eb158810    \n",
       "15  AGAACCTTCTCGAG        Zm00001eb042970    \n",
       "16  CCGCGGAAGCCGGAG       Zm00001eb059070    \n",
       "17  TGAATATTCTAGAG        Zm00001eb164390    \n",
       "18  TCGAGAGAGACGGC        Zm00001eb092540    \n",
       "19  GAGAGAGACGGCGG        Zm00001eb092540    \n",
       "20  GAGAGAGACGGCGG        Zm00001eb197410    \n",
       "21  CAATCATT              Zm00001eb287770    \n",
       "22  CAATCATT              Zm00001eb197410    \n",
       "23  GAAAGAGAGAGAAG        Zm00001eb296550    \n",
       "24  CCGTCATCTTCGCCA       Zm00001eb359190    \n",
       "25  TTGAGAGAGACGGC        Zm00001eb197410    \n",
       "26  GGCCCCAC              Zm00001eb042970    \n",
       "27  GGCCCCAC              Zm00001eb164390    \n",
       "28  GGCCCCAC              Zm00001eb164390    \n",
       "29  GAGAGAGGCAGAGG        Zm00001eb059070    \n",
       "30  ACGCCGCATCCGCCG       Zm00001eb197410    \n",
       "⋮   ⋮                     ⋮                  \n",
       "678 AAGATTC               Urofu.3G289200.v1.1\n",
       "679 ATTTGGCGCGGGG         Urofu.9G071000.v1.1\n",
       "680 TGACTTATATTAGACGTTAAG Urofu.9G196500.v1.1\n",
       "681 GAGAAAGGGAAAGG        Urofu.6G175300.v1.1\n",
       "682 GGGCCCAA              Urofu.3G358900.v1.1\n",
       "683 GAGAAAGAACGCGA        Urofu.2G245100.v1.1\n",
       "684 TCCCGTATTTCACAAGGAAAA Urofu.6G101400.v1.1\n",
       "685 GAGAGAGGGCTAGG        Urofu.6G175300.v1.1\n",
       "686 AAAAAAGAGAGAGA        Urofu.9G196500.v1.1\n",
       "687 GCGAGAGACGGACG        Urofu.2G409400.v1.1\n",
       "688 TTGGGGCGGGCGA         Urofu.3G289200.v1.1\n",
       "689 TCAGAAAAGGAAA         Urofu.8G225900.v1.1\n",
       "690 TTCCTTGTGAAATACGGGATA Urofu.6G101400.v1.1\n",
       "691 CCGACGGCGACGGCG       Urofu.2G245100.v1.1\n",
       "692 CCAATATTAGAAC         Urofu.2G135100.v1.1\n",
       "693 ACGAAAATGGAAT         Urofu.5G464900.v1.1\n",
       "694 CCCACGTGTA            Urofu.6G175300.v1.1\n",
       "695 TTGTGGCGGCTCC         Urofu.3G289200.v1.1\n",
       "696 CAATAAT               Urofu.3G289200.v1.1\n",
       "697 CAATAAT               Urofu.2G419400.v1.1\n",
       "698 CAATAAT               Urofu.1G324800.v1.1\n",
       "699 CAATAAT               Urofu.3G359400.v1.1\n",
       "700 CAATAAT               Urofu.3G359400.v1.1\n",
       "701 CAATAAT               Urofu.3G359400.v1.1\n",
       "702 CAATAAT               Urofu.9G489700.v1.1\n",
       "703 CCGTCAAATCTGCAG       Urofu.2G409400.v1.1\n",
       "704 TCGTCGACGCCGCCG       Urofu.9G489700.v1.1\n",
       "705 CCGCGATGGCCGCCA       Urofu.9G489700.v1.1\n",
       "706 TGCCTTACTGGGGACGAAAAA Urofu.9G489700.v1.1\n",
       "707 CCGCCACAACAGCCG       Urofu.3G289200.v1.1\n",
       "    gene_name                                                  type species\n",
       "1   PPT                                                        ms   zm     \n",
       "2   NAD-MDH2                                                   ms   zm     \n",
       "3   PEPCK                                                      ms   zm     \n",
       "4   NAD-MDH2                                                   ms   zm     \n",
       "5   NAD-MDH2                                                   ms   zm     \n",
       "6   NADP_ME                                                    bs   zm     \n",
       "7   NAD-MDH12                                                  ms   zm     \n",
       "8   CA2                                                        ms   zm     \n",
       "9   CA1                                                        ms   zm     \n",
       "10  PEPCK                                                      ms   zm     \n",
       "11  CA6                                                        ms   zm     \n",
       "12  PEPCK                                                      ms   zm     \n",
       "13  CA2                                                        ms   zm     \n",
       "14  CA1                                                        ms   zm     \n",
       "15  PPT                                                        ms   zm     \n",
       "16  NAD-MDH12                                                  bs   zm     \n",
       "17  RUBISCO_activase1                                          bs   zm     \n",
       "18  RBCA2                                                      ms   zm     \n",
       "19  RBCA2                                                      ms   zm     \n",
       "20  RBCA1                                                      ms   zm     \n",
       "21  PPDK                                                       ms   zm     \n",
       "22  RBCA1                                                      bs   zm     \n",
       "23  NAD-MDH2                                                   ms   zm     \n",
       "24  CA6                                                        ms   zm     \n",
       "25  RBCA1                                                      ms   zm     \n",
       "26  PPT                                                        ms   zm     \n",
       "27  RUBISCO_activase1                                          bs   zm     \n",
       "28  RUBISCO_activase1                                          bs   zm     \n",
       "29  NAD-MDH12                                                  ms   zm     \n",
       "30  RBCA1                                                      ms   zm     \n",
       "⋮   ⋮                                                          ⋮    ⋮      \n",
       "678 PPDK                                                       ms   uf     \n",
       "679 NAD-MDH12                                                  bs   uf     \n",
       "680 BASS                                                       bs   uf     \n",
       "681 DIT2                                                       bs   uf     \n",
       "682 RBCA2_ortho_count_2__RBCA1_ortho_count_2                   bs   uf     \n",
       "683 DIT1                                                       ms   uf     \n",
       "684 PPT                                                        bs   uf     \n",
       "685 DIT2                                                       bs   uf     \n",
       "686 BASS                                                       bs   uf     \n",
       "687 AlaAT2                                                     ms   uf     \n",
       "688 PPDK                                                       ms   uf     \n",
       "689 RUBISCO_activase1                                          bs   uf     \n",
       "690 PPT                                                        bs   uf     \n",
       "691 DIT1                                                       ms   uf     \n",
       "692 NHD1__NHD2                                                 ms   uf     \n",
       "693 NADP_ME1__NADP-ME_1_ortho_count_1__NADP-ME_2_ortho_count_1 bs   uf     \n",
       "694 DIT2                                                       bs   uf     \n",
       "695 PPDK                                                       ms   uf     \n",
       "696 PPDK                                                       ms   uf     \n",
       "697 NAD_ME10                                                   ms   uf     \n",
       "698 ASP_AT1                                                    ms   uf     \n",
       "699 RBCA2_ortho_count_3__RBCA1_ortho_count_3                   bs   uf     \n",
       "700 RBCA2_ortho_count_3__RBCA1_ortho_count_3                   bs   uf     \n",
       "701 RBCA2_ortho_count_3__RBCA1_ortho_count_3                   bs   uf     \n",
       "702 PEPCK                                                      bs   uf     \n",
       "703 AlaAT2                                                     ms   uf     \n",
       "704 PEPCK                                                      bs   uf     \n",
       "705 PEPCK                                                      bs   uf     \n",
       "706 PEPCK                                                      bs   uf     \n",
       "707 PPDK                                                       ms   uf     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_species_fimo_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4e53b7f5-a7d0-475c-b22b-1193cb1b48a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path <- \"/home/jpm73279/genome_downloads/C4_markers/generate_null_ACR_files/\"  # Adjust this to your directory\n",
    "file_lists_null <- list(\n",
    "  Os = list(\n",
    "    bs = paste0(base_path, \"Os_bs_acrs_intersect.null.bed..broad_acr_null_list.bed\"),\n",
    "    ms = paste0(base_path, \"Os_ms_acrs_intersect.null.bed..broad_acr_null_list.bed\")\n",
    "  ),\n",
    "  Pm = list(\n",
    "    bs = paste0(base_path, \"Pm_bs_acrs_intersect.null.bed..broad_acr_null_list.bed\"),\n",
    "    ms = paste0(base_path, \"Pm_ms_acrs_intersect.null.bed..broad_acr_null_list.bed\")\n",
    "  ),\n",
    "  Sb = list(\n",
    "    bs = paste0(base_path, \"Sb_bs_acrs_intersect.null.bed..broad_acr_null_list.bed\"),\n",
    "    ms = paste0(base_path, \"Sb_ms_acrs_intersect.null.bed..broad_acr_null_list.bed\")\n",
    "  ),\n",
    "  Uf = list(\n",
    "    bs = paste0(base_path, \"Uf_bs_acrs_intersect.null.bed..broad_acr_null_list.bed\"),\n",
    "    ms = paste0(base_path, \"Uf_ms_acrs_intersect.null.bed..broad_acr_null_list.bed\")\n",
    "  ),\n",
    "  Zm = list(\n",
    "    bs = paste0(base_path, \"Zm_bs_acrs_intersect.null.bed..broad_acr_null_list.bed\"),\n",
    "    ms = paste0(base_path, \"Zm_ms_acrs_intersect.null.bed..broad_acr_null_list.bed\")\n",
    "  )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dc79af27-b9f6-4e5c-9f33-15f3b07788ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mOne or more parsing issues, call `problems()` on your data frame for details,\n",
      "e.g.:\n",
      "  dat <- vroom(...)\n",
      "  problems(dat)”\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m149440\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): motif_id, sequence_name, strand, matched_sequence\n",
      "\u001b[32mdbl\u001b[39m (5): start, stop, score, p-value, q-value\n",
      "\u001b[33mlgl\u001b[39m (1): motif_alt_id\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m15\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m27\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mOne or more parsing issues, call `problems()` on your data frame for details,\n",
      "e.g.:\n",
      "  dat <- vroom(...)\n",
      "  problems(dat)”\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m194707\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): motif_id, sequence_name, strand, matched_sequence\n",
      "\u001b[32mdbl\u001b[39m (5): start, stop, score, p-value, q-value\n",
      "\u001b[33mlgl\u001b[39m (1): motif_alt_id\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m54\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m38\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mOne or more parsing issues, call `problems()` on your data frame for details,\n",
      "e.g.:\n",
      "  dat <- vroom(...)\n",
      "  problems(dat)”\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m221103\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): motif_id, sequence_name, strand, matched_sequence\n",
      "\u001b[32mdbl\u001b[39m (5): start, stop, score, p-value, q-value\n",
      "\u001b[33mlgl\u001b[39m (1): motif_alt_id\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m14\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m24\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mOne or more parsing issues, call `problems()` on your data frame for details,\n",
      "e.g.:\n",
      "  dat <- vroom(...)\n",
      "  problems(dat)”\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m82705\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): motif_id, sequence_name, strand, matched_sequence\n",
      "\u001b[32mdbl\u001b[39m (5): start, stop, score, p-value, q-value\n",
      "\u001b[33mlgl\u001b[39m (1): motif_alt_id\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m35\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m32\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mOne or more parsing issues, call `problems()` on your data frame for details,\n",
      "e.g.:\n",
      "  dat <- vroom(...)\n",
      "  problems(dat)”\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m137863\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): motif_id, sequence_name, strand, matched_sequence\n",
      "\u001b[32mdbl\u001b[39m (5): start, stop, score, p-value, q-value\n",
      "\u001b[33mlgl\u001b[39m (1): motif_alt_id\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m14\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m28\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (2): chromosome, acrs_id\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\u001b[33mlgl\u001b[39m (1): p_value\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "zm_fimo_counts_null <- process_data_null_files(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/06.motif_enrichment/zm_dir/all_acrs/zm_fimo_output/fimo.tsv\",\n",
    "                 c(file_lists_null$Zm$bs, file_lists_null$Zm$ms), \"zm\")       \n",
    "pm_fimo_counts_null <- process_data_null_files(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/06.motif_enrichment/pm_dir/all_acrs/pm_fimo_output/fimo.tsv\",\n",
    "                 c(file_lists_null$Pm$bs, file_lists_null$Pm$ms), \"pm\")       \n",
    "os_fimo_counts_null <- process_data_null_files(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/06.motif_enrichment/os_dir/all_acrs/os_fimo_output/fimo.tsv\",\n",
    "                 c(file_lists_null$Os$bs, file_lists_null$Os$ms), \"os\")       \n",
    "uf_fimo_counts_null <- process_data_null_files(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/06.motif_enrichment/uf_dir/all_acrs/uf_fimo_output/fimo.tsv\",\n",
    "                 c(file_lists_null$Uf$bs, file_lists_null$Uf$ms), \"uf\")       \n",
    "sb_fimo_counts_null <- process_data_null_files(\"/scratch/jpm73279/comparative_single_cell/07.call.ACRs/06.motif_enrichment/sb_dir/all_acrs/sb_fimo_output/fimo.tsv\",\n",
    "                 c(file_lists_null$Sb$bs, file_lists_null$Sb$ms), \"sb\")       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "73a0940d-0ca5-4af8-8dd8-aeed16af0cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fimo_counts_combined_null <- bind_rows(\n",
    "  zm_fimo_counts_null,\n",
    "  pm_fimo_counts_null,\n",
    "  os_fimo_counts_null,\n",
    "  uf_fimo_counts_null,\n",
    "  sb_fimo_counts_null\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8df2a5c1-09d7-4937-ac8b-de97acb07feb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>motif_alt_id</th><th scope=col>acrID</th><th scope=col>region</th><th scope=col>start</th><th scope=col>stop</th><th scope=col>strand</th><th scope=col>score</th><th scope=col>p-value</th><th scope=col>q-value</th><th scope=col>matched_sequence</th><th scope=col>gene_name</th><th scope=col>type</th><th scope=col>species</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BBRBPC</td><td>NA</td><td>scACR_7004;broadly_accessible</td><td>chr1:261782646-261783147</td><td>266</td><td>279</td><td>-</td><td>21.9268</td><td>3.48e-09</td><td>1.93e-05</td><td>GAGAGAGAGAGAGA</td><td>null</td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC</td><td>NA</td><td>scACR_7004;broadly_accessible</td><td>chr1:261782646-261783147</td><td>268</td><td>281</td><td>-</td><td>21.9268</td><td>3.48e-09</td><td>1.93e-05</td><td>GAGAGAGAGAGAGA</td><td>null</td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC</td><td>NA</td><td>scACR_7004;broadly_accessible</td><td>chr1:261782646-261783147</td><td>264</td><td>277</td><td>-</td><td>19.8171</td><td>4.38e-08</td><td>1.93e-04</td><td>GAGAGAGAGAGAGT</td><td>null</td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC</td><td>NA</td><td>scACR_5293;broadly_accessible</td><td>chr1:207077821-207078322</td><td>209</td><td>222</td><td>+</td><td>19.7683</td><td>5.50e-08</td><td>2.33e-04</td><td>GAGAGAGAGAGACA</td><td>null</td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC</td><td>NA</td><td>scACR_5293;broadly_accessible</td><td>chr1:207077821-207078322</td><td>207</td><td>220</td><td>+</td><td>17.9756</td><td>2.96e-07</td><td>1.04e-03</td><td>AGGAGAGAGAGAGA</td><td>null</td><td>ms</td><td>zm</td></tr>\n",
       "\t<tr><td>BBRBPC</td><td>NA</td><td>scACR_7004;broadly_accessible</td><td>chr1:261782646-261783147</td><td>270</td><td>283</td><td>-</td><td>17.8171</td><td>3.94e-07</td><td>1.33e-03</td><td>ACGAGAGAGAGAGA</td><td>null</td><td>ms</td><td>zm</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " motif\\_id & motif\\_alt\\_id & acrID & region & start & stop & strand & score & p-value & q-value & matched\\_sequence & gene\\_name & type & species\\\\\n",
       " <chr> & <lgl> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t BBRBPC & NA & scACR\\_7004;broadly\\_accessible & chr1:261782646-261783147 & 266 & 279 & - & 21.9268 & 3.48e-09 & 1.93e-05 & GAGAGAGAGAGAGA & null & ms & zm\\\\\n",
       "\t BBRBPC & NA & scACR\\_7004;broadly\\_accessible & chr1:261782646-261783147 & 268 & 281 & - & 21.9268 & 3.48e-09 & 1.93e-05 & GAGAGAGAGAGAGA & null & ms & zm\\\\\n",
       "\t BBRBPC & NA & scACR\\_7004;broadly\\_accessible & chr1:261782646-261783147 & 264 & 277 & - & 19.8171 & 4.38e-08 & 1.93e-04 & GAGAGAGAGAGAGT & null & ms & zm\\\\\n",
       "\t BBRBPC & NA & scACR\\_5293;broadly\\_accessible & chr1:207077821-207078322 & 209 & 222 & + & 19.7683 & 5.50e-08 & 2.33e-04 & GAGAGAGAGAGACA & null & ms & zm\\\\\n",
       "\t BBRBPC & NA & scACR\\_5293;broadly\\_accessible & chr1:207077821-207078322 & 207 & 220 & + & 17.9756 & 2.96e-07 & 1.04e-03 & AGGAGAGAGAGAGA & null & ms & zm\\\\\n",
       "\t BBRBPC & NA & scACR\\_7004;broadly\\_accessible & chr1:261782646-261783147 & 270 & 283 & - & 17.8171 & 3.94e-07 & 1.33e-03 & ACGAGAGAGAGAGA & null & ms & zm\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 14\n",
       "\n",
       "| motif_id &lt;chr&gt; | motif_alt_id &lt;lgl&gt; | acrID &lt;chr&gt; | region &lt;chr&gt; | start &lt;dbl&gt; | stop &lt;dbl&gt; | strand &lt;chr&gt; | score &lt;dbl&gt; | p-value &lt;dbl&gt; | q-value &lt;dbl&gt; | matched_sequence &lt;chr&gt; | gene_name &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| BBRBPC | NA | scACR_7004;broadly_accessible | chr1:261782646-261783147 | 266 | 279 | - | 21.9268 | 3.48e-09 | 1.93e-05 | GAGAGAGAGAGAGA | null | ms | zm |\n",
       "| BBRBPC | NA | scACR_7004;broadly_accessible | chr1:261782646-261783147 | 268 | 281 | - | 21.9268 | 3.48e-09 | 1.93e-05 | GAGAGAGAGAGAGA | null | ms | zm |\n",
       "| BBRBPC | NA | scACR_7004;broadly_accessible | chr1:261782646-261783147 | 264 | 277 | - | 19.8171 | 4.38e-08 | 1.93e-04 | GAGAGAGAGAGAGT | null | ms | zm |\n",
       "| BBRBPC | NA | scACR_5293;broadly_accessible | chr1:207077821-207078322 | 209 | 222 | + | 19.7683 | 5.50e-08 | 2.33e-04 | GAGAGAGAGAGACA | null | ms | zm |\n",
       "| BBRBPC | NA | scACR_5293;broadly_accessible | chr1:207077821-207078322 | 207 | 220 | + | 17.9756 | 2.96e-07 | 1.04e-03 | AGGAGAGAGAGAGA | null | ms | zm |\n",
       "| BBRBPC | NA | scACR_7004;broadly_accessible | chr1:261782646-261783147 | 270 | 283 | - | 17.8171 | 3.94e-07 | 1.33e-03 | ACGAGAGAGAGAGA | null | ms | zm |\n",
       "\n"
      ],
      "text/plain": [
       "  motif_id motif_alt_id acrID                         region                  \n",
       "1 BBRBPC   NA           scACR_7004;broadly_accessible chr1:261782646-261783147\n",
       "2 BBRBPC   NA           scACR_7004;broadly_accessible chr1:261782646-261783147\n",
       "3 BBRBPC   NA           scACR_7004;broadly_accessible chr1:261782646-261783147\n",
       "4 BBRBPC   NA           scACR_5293;broadly_accessible chr1:207077821-207078322\n",
       "5 BBRBPC   NA           scACR_5293;broadly_accessible chr1:207077821-207078322\n",
       "6 BBRBPC   NA           scACR_7004;broadly_accessible chr1:261782646-261783147\n",
       "  start stop strand score   p-value  q-value  matched_sequence gene_name type\n",
       "1 266   279  -      21.9268 3.48e-09 1.93e-05 GAGAGAGAGAGAGA   null      ms  \n",
       "2 268   281  -      21.9268 3.48e-09 1.93e-05 GAGAGAGAGAGAGA   null      ms  \n",
       "3 264   277  -      19.8171 4.38e-08 1.93e-04 GAGAGAGAGAGAGT   null      ms  \n",
       "4 209   222  +      19.7683 5.50e-08 2.33e-04 GAGAGAGAGAGACA   null      ms  \n",
       "5 207   220  +      17.9756 2.96e-07 1.04e-03 AGGAGAGAGAGAGA   null      ms  \n",
       "6 270   283  -      17.8171 3.94e-07 1.33e-03 ACGAGAGAGAGAGA   null      ms  \n",
       "  species\n",
       "1 zm     \n",
       "2 zm     \n",
       "3 zm     \n",
       "4 zm     \n",
       "5 zm     \n",
       "6 zm     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fimo_counts_combined_null)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eb97418-fcad-4d60-8fdd-aa0846618a52",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Analysis by Species and Cell Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c03f3d52-3f17-4a52-b3dd-6bea44dbbc12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'motif_id', 'type'. You can override using\n",
      "the `.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 20 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>type</th><th scope=col>species</th><th scope=col>count_motif_occurence</th><th scope=col>counts</th><th scope=col>non_motif_occurence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB      </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>ms</td><td>sb</td><td>11</td><td>28</td><td>17</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>bs</td><td>sb</td><td>12</td><td>14</td><td> 2</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>ms</td><td>sb</td><td>30</td><td>28</td><td>-2</td></tr>\n",
       "\t<tr><td>CPP         </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>E2FDP       </td><td>bs</td><td>sb</td><td> 2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>E2FDP       </td><td>ms</td><td>sb</td><td> 4</td><td>28</td><td>24</td></tr>\n",
       "\t<tr><td>GARP_ARR-B  </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>GARP_G2-like</td><td>bs</td><td>sb</td><td> 2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>HD          </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>HD          </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>HD-Zip_IV   </td><td>bs</td><td>sb</td><td> 3</td><td>14</td><td>11</td></tr>\n",
       "\t<tr><td>HD-Zip_IV   </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>HSF         </td><td>bs</td><td>sb</td><td> 2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>HSF         </td><td>ms</td><td>sb</td><td> 4</td><td>28</td><td>24</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>NAC         </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>bs</td><td>sb</td><td> 2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 20 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " motif\\_id & type & species & count\\_motif\\_occurence & counts & non\\_motif\\_occurence\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t AS2LOB       & bs & sb &  1 & 14 & 13\\\\\n",
       "\t AS2LOB       & ms & sb & 11 & 28 & 17\\\\\n",
       "\t BBRBPC       & bs & sb & 12 & 14 &  2\\\\\n",
       "\t BBRBPC       & ms & sb & 30 & 28 & -2\\\\\n",
       "\t CPP          & bs & sb &  1 & 14 & 13\\\\\n",
       "\t E2FDP        & bs & sb &  2 & 14 & 12\\\\\n",
       "\t E2FDP        & ms & sb &  4 & 28 & 24\\\\\n",
       "\t GARP\\_ARR-B   & ms & sb &  2 & 28 & 26\\\\\n",
       "\t GARP\\_G2-like & bs & sb &  2 & 14 & 12\\\\\n",
       "\t HD           & bs & sb &  1 & 14 & 13\\\\\n",
       "\t HD           & ms & sb &  2 & 28 & 26\\\\\n",
       "\t HD-Zip\\_IV    & bs & sb &  3 & 14 & 11\\\\\n",
       "\t HD-Zip\\_IV    & ms & sb &  2 & 28 & 26\\\\\n",
       "\t HD-Zip\\_I\\_II  & bs & sb &  1 & 14 & 13\\\\\n",
       "\t HSF          & bs & sb &  2 & 14 & 12\\\\\n",
       "\t HSF          & ms & sb &  4 & 28 & 24\\\\\n",
       "\t MADS\\_MIKC    & bs & sb &  1 & 14 & 13\\\\\n",
       "\t NAC          & ms & sb &  2 & 28 & 26\\\\\n",
       "\t bHLH\\_TCP     & bs & sb &  2 & 14 & 12\\\\\n",
       "\t bHLH\\_TCP     & ms & sb &  2 & 28 & 26\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 20 × 6\n",
       "\n",
       "| motif_id &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; | count_motif_occurence &lt;int&gt; | counts &lt;int&gt; | non_motif_occurence &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| AS2LOB       | bs | sb |  1 | 14 | 13 |\n",
       "| AS2LOB       | ms | sb | 11 | 28 | 17 |\n",
       "| BBRBPC       | bs | sb | 12 | 14 |  2 |\n",
       "| BBRBPC       | ms | sb | 30 | 28 | -2 |\n",
       "| CPP          | bs | sb |  1 | 14 | 13 |\n",
       "| E2FDP        | bs | sb |  2 | 14 | 12 |\n",
       "| E2FDP        | ms | sb |  4 | 28 | 24 |\n",
       "| GARP_ARR-B   | ms | sb |  2 | 28 | 26 |\n",
       "| GARP_G2-like | bs | sb |  2 | 14 | 12 |\n",
       "| HD           | bs | sb |  1 | 14 | 13 |\n",
       "| HD           | ms | sb |  2 | 28 | 26 |\n",
       "| HD-Zip_IV    | bs | sb |  3 | 14 | 11 |\n",
       "| HD-Zip_IV    | ms | sb |  2 | 28 | 26 |\n",
       "| HD-Zip_I_II  | bs | sb |  1 | 14 | 13 |\n",
       "| HSF          | bs | sb |  2 | 14 | 12 |\n",
       "| HSF          | ms | sb |  4 | 28 | 24 |\n",
       "| MADS_MIKC    | bs | sb |  1 | 14 | 13 |\n",
       "| NAC          | ms | sb |  2 | 28 | 26 |\n",
       "| bHLH_TCP     | bs | sb |  2 | 14 | 12 |\n",
       "| bHLH_TCP     | ms | sb |  2 | 28 | 26 |\n",
       "\n"
      ],
      "text/plain": [
       "   motif_id     type species count_motif_occurence counts non_motif_occurence\n",
       "1  AS2LOB       bs   sb       1                    14     13                 \n",
       "2  AS2LOB       ms   sb      11                    28     17                 \n",
       "3  BBRBPC       bs   sb      12                    14      2                 \n",
       "4  BBRBPC       ms   sb      30                    28     -2                 \n",
       "5  CPP          bs   sb       1                    14     13                 \n",
       "6  E2FDP        bs   sb       2                    14     12                 \n",
       "7  E2FDP        ms   sb       4                    28     24                 \n",
       "8  GARP_ARR-B   ms   sb       2                    28     26                 \n",
       "9  GARP_G2-like bs   sb       2                    14     12                 \n",
       "10 HD           bs   sb       1                    14     13                 \n",
       "11 HD           ms   sb       2                    28     26                 \n",
       "12 HD-Zip_IV    bs   sb       3                    14     11                 \n",
       "13 HD-Zip_IV    ms   sb       2                    28     26                 \n",
       "14 HD-Zip_I_II  bs   sb       1                    14     13                 \n",
       "15 HSF          bs   sb       2                    14     12                 \n",
       "16 HSF          ms   sb       4                    28     24                 \n",
       "17 MADS_MIKC    bs   sb       1                    14     13                 \n",
       "18 NAC          ms   sb       2                    28     26                 \n",
       "19 bHLH_TCP     bs   sb       2                    14     12                 \n",
       "20 bHLH_TCP     ms   sb       2                    28     26                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_species_fimo_counts %>% \n",
    "    dplyr::select(motif_id, acrID, type,species) %>% \n",
    "    dplyr::group_by(motif_id, type, species) %>% \n",
    "    dplyr::summarise(count_motif_occurence = n()) %>% \n",
    "    left_join(., all_species_total_counts, by = c(\"type\", \"species\")) %>% \n",
    "    dplyr::rowwise() %>% \n",
    "    mutate(non_motif_occurence = counts - count_motif_occurence) %>% \n",
    "    dplyr::filter(species == \"sb\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "40da1ee7-9061-414f-8c34-59db70c6294d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'motif_id', 'type'. You can override using\n",
      "the `.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 27 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>type</th><th scope=col>species</th><th scope=col>count_motif_occurence</th><th scope=col>counts</th><th scope=col>non_motif_occurence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB      </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>ms</td><td>sb</td><td>16</td><td>28</td><td>12</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>bs</td><td>sb</td><td> 6</td><td>14</td><td> 8</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>ms</td><td>sb</td><td>17</td><td>28</td><td>11</td></tr>\n",
       "\t<tr><td>BES1        </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>BES1        </td><td>ms</td><td>sb</td><td> 4</td><td>28</td><td>24</td></tr>\n",
       "\t<tr><td>CPP         </td><td>bs</td><td>sb</td><td> 3</td><td>14</td><td>11</td></tr>\n",
       "\t<tr><td>CPP         </td><td>ms</td><td>sb</td><td> 5</td><td>28</td><td>23</td></tr>\n",
       "\t<tr><td>E2FDP       </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>GARP_ARR-B  </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>GARP_ARR-B  </td><td>ms</td><td>sb</td><td> 1</td><td>28</td><td>27</td></tr>\n",
       "\t<tr><td>GARP_G2-like</td><td>bs</td><td>sb</td><td> 3</td><td>14</td><td>11</td></tr>\n",
       "\t<tr><td>GARP_G2-like</td><td>ms</td><td>sb</td><td> 4</td><td>28</td><td>24</td></tr>\n",
       "\t<tr><td>HD          </td><td>bs</td><td>sb</td><td> 2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>HD          </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>HD-Zip_IV   </td><td>ms</td><td>sb</td><td> 3</td><td>28</td><td>25</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II </td><td>bs</td><td>sb</td><td> 2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II </td><td>ms</td><td>sb</td><td> 1</td><td>28</td><td>27</td></tr>\n",
       "\t<tr><td>HSF         </td><td>bs</td><td>sb</td><td> 5</td><td>14</td><td> 9</td></tr>\n",
       "\t<tr><td>HSF         </td><td>ms</td><td>sb</td><td> 3</td><td>28</td><td>25</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>bs</td><td>sb</td><td> 3</td><td>14</td><td>11</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>ms</td><td>sb</td><td> 5</td><td>28</td><td>23</td></tr>\n",
       "\t<tr><td>NAC         </td><td>bs</td><td>sb</td><td> 3</td><td>14</td><td>11</td></tr>\n",
       "\t<tr><td>NAC         </td><td>ms</td><td>sb</td><td> 1</td><td>28</td><td>27</td></tr>\n",
       "\t<tr><td>WRKY        </td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>WRKY        </td><td>ms</td><td>sb</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>ms</td><td>sb</td><td> 6</td><td>28</td><td>22</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 27 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " motif\\_id & type & species & count\\_motif\\_occurence & counts & non\\_motif\\_occurence\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t AS2LOB       & bs & sb &  1 & 14 & 13\\\\\n",
       "\t AS2LOB       & ms & sb & 16 & 28 & 12\\\\\n",
       "\t BBRBPC       & bs & sb &  6 & 14 &  8\\\\\n",
       "\t BBRBPC       & ms & sb & 17 & 28 & 11\\\\\n",
       "\t BES1         & bs & sb &  1 & 14 & 13\\\\\n",
       "\t BES1         & ms & sb &  4 & 28 & 24\\\\\n",
       "\t CPP          & bs & sb &  3 & 14 & 11\\\\\n",
       "\t CPP          & ms & sb &  5 & 28 & 23\\\\\n",
       "\t E2FDP        & ms & sb &  2 & 28 & 26\\\\\n",
       "\t GARP\\_ARR-B   & bs & sb &  1 & 14 & 13\\\\\n",
       "\t GARP\\_ARR-B   & ms & sb &  1 & 28 & 27\\\\\n",
       "\t GARP\\_G2-like & bs & sb &  3 & 14 & 11\\\\\n",
       "\t GARP\\_G2-like & ms & sb &  4 & 28 & 24\\\\\n",
       "\t HD           & bs & sb &  2 & 14 & 12\\\\\n",
       "\t HD           & ms & sb &  2 & 28 & 26\\\\\n",
       "\t HD-Zip\\_IV    & ms & sb &  3 & 28 & 25\\\\\n",
       "\t HD-Zip\\_I\\_II  & bs & sb &  2 & 14 & 12\\\\\n",
       "\t HD-Zip\\_I\\_II  & ms & sb &  1 & 28 & 27\\\\\n",
       "\t HSF          & bs & sb &  5 & 14 &  9\\\\\n",
       "\t HSF          & ms & sb &  3 & 28 & 25\\\\\n",
       "\t MADS\\_MIKC    & bs & sb &  3 & 14 & 11\\\\\n",
       "\t MADS\\_MIKC    & ms & sb &  5 & 28 & 23\\\\\n",
       "\t NAC          & bs & sb &  3 & 14 & 11\\\\\n",
       "\t NAC          & ms & sb &  1 & 28 & 27\\\\\n",
       "\t WRKY         & bs & sb &  1 & 14 & 13\\\\\n",
       "\t WRKY         & ms & sb &  2 & 28 & 26\\\\\n",
       "\t bHLH\\_TCP     & ms & sb &  6 & 28 & 22\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 27 × 6\n",
       "\n",
       "| motif_id &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; | count_motif_occurence &lt;int&gt; | counts &lt;int&gt; | non_motif_occurence &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| AS2LOB       | bs | sb |  1 | 14 | 13 |\n",
       "| AS2LOB       | ms | sb | 16 | 28 | 12 |\n",
       "| BBRBPC       | bs | sb |  6 | 14 |  8 |\n",
       "| BBRBPC       | ms | sb | 17 | 28 | 11 |\n",
       "| BES1         | bs | sb |  1 | 14 | 13 |\n",
       "| BES1         | ms | sb |  4 | 28 | 24 |\n",
       "| CPP          | bs | sb |  3 | 14 | 11 |\n",
       "| CPP          | ms | sb |  5 | 28 | 23 |\n",
       "| E2FDP        | ms | sb |  2 | 28 | 26 |\n",
       "| GARP_ARR-B   | bs | sb |  1 | 14 | 13 |\n",
       "| GARP_ARR-B   | ms | sb |  1 | 28 | 27 |\n",
       "| GARP_G2-like | bs | sb |  3 | 14 | 11 |\n",
       "| GARP_G2-like | ms | sb |  4 | 28 | 24 |\n",
       "| HD           | bs | sb |  2 | 14 | 12 |\n",
       "| HD           | ms | sb |  2 | 28 | 26 |\n",
       "| HD-Zip_IV    | ms | sb |  3 | 28 | 25 |\n",
       "| HD-Zip_I_II  | bs | sb |  2 | 14 | 12 |\n",
       "| HD-Zip_I_II  | ms | sb |  1 | 28 | 27 |\n",
       "| HSF          | bs | sb |  5 | 14 |  9 |\n",
       "| HSF          | ms | sb |  3 | 28 | 25 |\n",
       "| MADS_MIKC    | bs | sb |  3 | 14 | 11 |\n",
       "| MADS_MIKC    | ms | sb |  5 | 28 | 23 |\n",
       "| NAC          | bs | sb |  3 | 14 | 11 |\n",
       "| NAC          | ms | sb |  1 | 28 | 27 |\n",
       "| WRKY         | bs | sb |  1 | 14 | 13 |\n",
       "| WRKY         | ms | sb |  2 | 28 | 26 |\n",
       "| bHLH_TCP     | ms | sb |  6 | 28 | 22 |\n",
       "\n"
      ],
      "text/plain": [
       "   motif_id     type species count_motif_occurence counts non_motif_occurence\n",
       "1  AS2LOB       bs   sb       1                    14     13                 \n",
       "2  AS2LOB       ms   sb      16                    28     12                 \n",
       "3  BBRBPC       bs   sb       6                    14      8                 \n",
       "4  BBRBPC       ms   sb      17                    28     11                 \n",
       "5  BES1         bs   sb       1                    14     13                 \n",
       "6  BES1         ms   sb       4                    28     24                 \n",
       "7  CPP          bs   sb       3                    14     11                 \n",
       "8  CPP          ms   sb       5                    28     23                 \n",
       "9  E2FDP        ms   sb       2                    28     26                 \n",
       "10 GARP_ARR-B   bs   sb       1                    14     13                 \n",
       "11 GARP_ARR-B   ms   sb       1                    28     27                 \n",
       "12 GARP_G2-like bs   sb       3                    14     11                 \n",
       "13 GARP_G2-like ms   sb       4                    28     24                 \n",
       "14 HD           bs   sb       2                    14     12                 \n",
       "15 HD           ms   sb       2                    28     26                 \n",
       "16 HD-Zip_IV    ms   sb       3                    28     25                 \n",
       "17 HD-Zip_I_II  bs   sb       2                    14     12                 \n",
       "18 HD-Zip_I_II  ms   sb       1                    28     27                 \n",
       "19 HSF          bs   sb       5                    14      9                 \n",
       "20 HSF          ms   sb       3                    28     25                 \n",
       "21 MADS_MIKC    bs   sb       3                    14     11                 \n",
       "22 MADS_MIKC    ms   sb       5                    28     23                 \n",
       "23 NAC          bs   sb       3                    14     11                 \n",
       "24 NAC          ms   sb       1                    28     27                 \n",
       "25 WRKY         bs   sb       1                    14     13                 \n",
       "26 WRKY         ms   sb       2                    28     26                 \n",
       "27 bHLH_TCP     ms   sb       6                    28     22                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fimo_counts_combined_null %>% \n",
    "    dplyr::select(motif_id, acrID, type,species) %>% \n",
    "    dplyr::group_by(motif_id, type, species) %>% \n",
    "    dplyr::summarise(count_motif_occurence = n()) %>% \n",
    "    left_join(., all_species_total_counts, by = c(\"type\", \"species\")) %>% \n",
    "    dplyr::rowwise() %>% \n",
    "    mutate(non_motif_occurence = counts - count_motif_occurence) %>% \n",
    "    dplyr::filter(species == \"sb\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "40505401-8458-4536-b263-525e68d3d7e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'motif_id', 'type', 'species'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 32 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>type</th><th scope=col>species</th><th scope=col>gene_name</th><th scope=col>count_motif_occurence</th><th scope=col>counts</th><th scope=col>non_motif_occurence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB     </td><td>ms</td><td>os</td><td>CA_1__CA_2       </td><td> 1</td><td>25</td><td>24</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>ms</td><td>sb</td><td>CA1              </td><td> 3</td><td>28</td><td>25</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>ms</td><td>sb</td><td>CA6_ortho_count_2</td><td> 2</td><td>28</td><td>26</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>ms</td><td>zm</td><td>CA1              </td><td> 2</td><td>29</td><td>27</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>ms</td><td>zm</td><td>CA2              </td><td> 2</td><td>29</td><td>27</td></tr>\n",
       "\t<tr><td>AS2LOB     </td><td>ms</td><td>zm</td><td>CA6              </td><td> 3</td><td>29</td><td>26</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>ms</td><td>sb</td><td>CA1              </td><td>10</td><td>28</td><td>18</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>ms</td><td>sb</td><td>CA6_ortho_count_2</td><td>12</td><td>28</td><td>16</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>ms</td><td>zm</td><td>CA1              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>BBRBPC     </td><td>ms</td><td>zm</td><td>CA2              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>BES1       </td><td>bs</td><td>pm</td><td>CA6_ortho_count_2</td><td> 2</td><td>54</td><td>52</td></tr>\n",
       "\t<tr><td>BES1       </td><td>bs</td><td>pm</td><td>CA_2             </td><td> 2</td><td>54</td><td>52</td></tr>\n",
       "\t<tr><td>BES1       </td><td>ms</td><td>zm</td><td>CA6              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>E2FDP      </td><td>ms</td><td>os</td><td>CA_1__CA_2       </td><td> 1</td><td>25</td><td>24</td></tr>\n",
       "\t<tr><td>E2FDP      </td><td>ms</td><td>sb</td><td>CA6_ortho_count_2</td><td> 1</td><td>28</td><td>27</td></tr>\n",
       "\t<tr><td>GARP_ARR-B </td><td>ms</td><td>sb</td><td>CA1              </td><td> 1</td><td>28</td><td>27</td></tr>\n",
       "\t<tr><td>GARP_ARR-B </td><td>ms</td><td>sb</td><td>CA6_ortho_count_2</td><td> 1</td><td>28</td><td>27</td></tr>\n",
       "\t<tr><td>GARP_ARR-B </td><td>ms</td><td>zm</td><td>CA1              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>GARP_ARR-B </td><td>ms</td><td>zm</td><td>CA2              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>GARP_ARR-B </td><td>ms</td><td>zm</td><td>CA6              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>HD         </td><td>ms</td><td>zm</td><td>CA6              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>HD-Zip_IV  </td><td>ms</td><td>os</td><td>CA_1__CA_2       </td><td> 1</td><td>25</td><td>24</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II</td><td>ms</td><td>zm</td><td>CA6              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>MADS_MIKC  </td><td>ms</td><td>uf</td><td>CA6              </td><td> 1</td><td>33</td><td>32</td></tr>\n",
       "\t<tr><td>MADS_MIKC  </td><td>ms</td><td>zm</td><td>CA6              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>NAC        </td><td>ms</td><td>zm</td><td>CA1              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>NAC        </td><td>ms</td><td>zm</td><td>CA2              </td><td> 1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>bs</td><td>pm</td><td>CA6_ortho_count_2</td><td> 1</td><td>54</td><td>53</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>bs</td><td>pm</td><td>CA_2             </td><td> 1</td><td>54</td><td>53</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>ms</td><td>pm</td><td>CA6_ortho_count_2</td><td> 2</td><td>38</td><td>36</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>ms</td><td>pm</td><td>CA_2             </td><td> 2</td><td>38</td><td>36</td></tr>\n",
       "\t<tr><td>bHLH_TCP   </td><td>ms</td><td>uf</td><td>CA6              </td><td> 1</td><td>33</td><td>32</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 32 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " motif\\_id & type & species & gene\\_name & count\\_motif\\_occurence & counts & non\\_motif\\_occurence\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t AS2LOB      & ms & os & CA\\_1\\_\\_CA\\_2        &  1 & 25 & 24\\\\\n",
       "\t AS2LOB      & ms & sb & CA1               &  3 & 28 & 25\\\\\n",
       "\t AS2LOB      & ms & sb & CA6\\_ortho\\_count\\_2 &  2 & 28 & 26\\\\\n",
       "\t AS2LOB      & ms & zm & CA1               &  2 & 29 & 27\\\\\n",
       "\t AS2LOB      & ms & zm & CA2               &  2 & 29 & 27\\\\\n",
       "\t AS2LOB      & ms & zm & CA6               &  3 & 29 & 26\\\\\n",
       "\t BBRBPC      & ms & sb & CA1               & 10 & 28 & 18\\\\\n",
       "\t BBRBPC      & ms & sb & CA6\\_ortho\\_count\\_2 & 12 & 28 & 16\\\\\n",
       "\t BBRBPC      & ms & zm & CA1               &  1 & 29 & 28\\\\\n",
       "\t BBRBPC      & ms & zm & CA2               &  1 & 29 & 28\\\\\n",
       "\t BES1        & bs & pm & CA6\\_ortho\\_count\\_2 &  2 & 54 & 52\\\\\n",
       "\t BES1        & bs & pm & CA\\_2              &  2 & 54 & 52\\\\\n",
       "\t BES1        & ms & zm & CA6               &  1 & 29 & 28\\\\\n",
       "\t E2FDP       & ms & os & CA\\_1\\_\\_CA\\_2        &  1 & 25 & 24\\\\\n",
       "\t E2FDP       & ms & sb & CA6\\_ortho\\_count\\_2 &  1 & 28 & 27\\\\\n",
       "\t GARP\\_ARR-B  & ms & sb & CA1               &  1 & 28 & 27\\\\\n",
       "\t GARP\\_ARR-B  & ms & sb & CA6\\_ortho\\_count\\_2 &  1 & 28 & 27\\\\\n",
       "\t GARP\\_ARR-B  & ms & zm & CA1               &  1 & 29 & 28\\\\\n",
       "\t GARP\\_ARR-B  & ms & zm & CA2               &  1 & 29 & 28\\\\\n",
       "\t GARP\\_ARR-B  & ms & zm & CA6               &  1 & 29 & 28\\\\\n",
       "\t HD          & ms & zm & CA6               &  1 & 29 & 28\\\\\n",
       "\t HD-Zip\\_IV   & ms & os & CA\\_1\\_\\_CA\\_2        &  1 & 25 & 24\\\\\n",
       "\t HD-Zip\\_I\\_II & ms & zm & CA6               &  1 & 29 & 28\\\\\n",
       "\t MADS\\_MIKC   & ms & uf & CA6               &  1 & 33 & 32\\\\\n",
       "\t MADS\\_MIKC   & ms & zm & CA6               &  1 & 29 & 28\\\\\n",
       "\t NAC         & ms & zm & CA1               &  1 & 29 & 28\\\\\n",
       "\t NAC         & ms & zm & CA2               &  1 & 29 & 28\\\\\n",
       "\t bHLH\\_TCP    & bs & pm & CA6\\_ortho\\_count\\_2 &  1 & 54 & 53\\\\\n",
       "\t bHLH\\_TCP    & bs & pm & CA\\_2              &  1 & 54 & 53\\\\\n",
       "\t bHLH\\_TCP    & ms & pm & CA6\\_ortho\\_count\\_2 &  2 & 38 & 36\\\\\n",
       "\t bHLH\\_TCP    & ms & pm & CA\\_2              &  2 & 38 & 36\\\\\n",
       "\t bHLH\\_TCP    & ms & uf & CA6               &  1 & 33 & 32\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 32 × 7\n",
       "\n",
       "| motif_id &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; | gene_name &lt;chr&gt; | count_motif_occurence &lt;int&gt; | counts &lt;int&gt; | non_motif_occurence &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| AS2LOB      | ms | os | CA_1__CA_2        |  1 | 25 | 24 |\n",
       "| AS2LOB      | ms | sb | CA1               |  3 | 28 | 25 |\n",
       "| AS2LOB      | ms | sb | CA6_ortho_count_2 |  2 | 28 | 26 |\n",
       "| AS2LOB      | ms | zm | CA1               |  2 | 29 | 27 |\n",
       "| AS2LOB      | ms | zm | CA2               |  2 | 29 | 27 |\n",
       "| AS2LOB      | ms | zm | CA6               |  3 | 29 | 26 |\n",
       "| BBRBPC      | ms | sb | CA1               | 10 | 28 | 18 |\n",
       "| BBRBPC      | ms | sb | CA6_ortho_count_2 | 12 | 28 | 16 |\n",
       "| BBRBPC      | ms | zm | CA1               |  1 | 29 | 28 |\n",
       "| BBRBPC      | ms | zm | CA2               |  1 | 29 | 28 |\n",
       "| BES1        | bs | pm | CA6_ortho_count_2 |  2 | 54 | 52 |\n",
       "| BES1        | bs | pm | CA_2              |  2 | 54 | 52 |\n",
       "| BES1        | ms | zm | CA6               |  1 | 29 | 28 |\n",
       "| E2FDP       | ms | os | CA_1__CA_2        |  1 | 25 | 24 |\n",
       "| E2FDP       | ms | sb | CA6_ortho_count_2 |  1 | 28 | 27 |\n",
       "| GARP_ARR-B  | ms | sb | CA1               |  1 | 28 | 27 |\n",
       "| GARP_ARR-B  | ms | sb | CA6_ortho_count_2 |  1 | 28 | 27 |\n",
       "| GARP_ARR-B  | ms | zm | CA1               |  1 | 29 | 28 |\n",
       "| GARP_ARR-B  | ms | zm | CA2               |  1 | 29 | 28 |\n",
       "| GARP_ARR-B  | ms | zm | CA6               |  1 | 29 | 28 |\n",
       "| HD          | ms | zm | CA6               |  1 | 29 | 28 |\n",
       "| HD-Zip_IV   | ms | os | CA_1__CA_2        |  1 | 25 | 24 |\n",
       "| HD-Zip_I_II | ms | zm | CA6               |  1 | 29 | 28 |\n",
       "| MADS_MIKC   | ms | uf | CA6               |  1 | 33 | 32 |\n",
       "| MADS_MIKC   | ms | zm | CA6               |  1 | 29 | 28 |\n",
       "| NAC         | ms | zm | CA1               |  1 | 29 | 28 |\n",
       "| NAC         | ms | zm | CA2               |  1 | 29 | 28 |\n",
       "| bHLH_TCP    | bs | pm | CA6_ortho_count_2 |  1 | 54 | 53 |\n",
       "| bHLH_TCP    | bs | pm | CA_2              |  1 | 54 | 53 |\n",
       "| bHLH_TCP    | ms | pm | CA6_ortho_count_2 |  2 | 38 | 36 |\n",
       "| bHLH_TCP    | ms | pm | CA_2              |  2 | 38 | 36 |\n",
       "| bHLH_TCP    | ms | uf | CA6               |  1 | 33 | 32 |\n",
       "\n"
      ],
      "text/plain": [
       "   motif_id    type species gene_name         count_motif_occurence counts\n",
       "1  AS2LOB      ms   os      CA_1__CA_2         1                    25    \n",
       "2  AS2LOB      ms   sb      CA1                3                    28    \n",
       "3  AS2LOB      ms   sb      CA6_ortho_count_2  2                    28    \n",
       "4  AS2LOB      ms   zm      CA1                2                    29    \n",
       "5  AS2LOB      ms   zm      CA2                2                    29    \n",
       "6  AS2LOB      ms   zm      CA6                3                    29    \n",
       "7  BBRBPC      ms   sb      CA1               10                    28    \n",
       "8  BBRBPC      ms   sb      CA6_ortho_count_2 12                    28    \n",
       "9  BBRBPC      ms   zm      CA1                1                    29    \n",
       "10 BBRBPC      ms   zm      CA2                1                    29    \n",
       "11 BES1        bs   pm      CA6_ortho_count_2  2                    54    \n",
       "12 BES1        bs   pm      CA_2               2                    54    \n",
       "13 BES1        ms   zm      CA6                1                    29    \n",
       "14 E2FDP       ms   os      CA_1__CA_2         1                    25    \n",
       "15 E2FDP       ms   sb      CA6_ortho_count_2  1                    28    \n",
       "16 GARP_ARR-B  ms   sb      CA1                1                    28    \n",
       "17 GARP_ARR-B  ms   sb      CA6_ortho_count_2  1                    28    \n",
       "18 GARP_ARR-B  ms   zm      CA1                1                    29    \n",
       "19 GARP_ARR-B  ms   zm      CA2                1                    29    \n",
       "20 GARP_ARR-B  ms   zm      CA6                1                    29    \n",
       "21 HD          ms   zm      CA6                1                    29    \n",
       "22 HD-Zip_IV   ms   os      CA_1__CA_2         1                    25    \n",
       "23 HD-Zip_I_II ms   zm      CA6                1                    29    \n",
       "24 MADS_MIKC   ms   uf      CA6                1                    33    \n",
       "25 MADS_MIKC   ms   zm      CA6                1                    29    \n",
       "26 NAC         ms   zm      CA1                1                    29    \n",
       "27 NAC         ms   zm      CA2                1                    29    \n",
       "28 bHLH_TCP    bs   pm      CA6_ortho_count_2  1                    54    \n",
       "29 bHLH_TCP    bs   pm      CA_2               1                    54    \n",
       "30 bHLH_TCP    ms   pm      CA6_ortho_count_2  2                    38    \n",
       "31 bHLH_TCP    ms   pm      CA_2               2                    38    \n",
       "32 bHLH_TCP    ms   uf      CA6                1                    33    \n",
       "   non_motif_occurence\n",
       "1  24                 \n",
       "2  25                 \n",
       "3  26                 \n",
       "4  27                 \n",
       "5  27                 \n",
       "6  26                 \n",
       "7  18                 \n",
       "8  16                 \n",
       "9  28                 \n",
       "10 28                 \n",
       "11 52                 \n",
       "12 52                 \n",
       "13 28                 \n",
       "14 24                 \n",
       "15 27                 \n",
       "16 27                 \n",
       "17 27                 \n",
       "18 28                 \n",
       "19 28                 \n",
       "20 28                 \n",
       "21 28                 \n",
       "22 24                 \n",
       "23 28                 \n",
       "24 32                 \n",
       "25 28                 \n",
       "26 28                 \n",
       "27 28                 \n",
       "28 53                 \n",
       "29 53                 \n",
       "30 36                 \n",
       "31 36                 \n",
       "32 32                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_species_fimo_counts %>% \n",
    "    dplyr::select(motif_id, type,species, gene_name) %>% \n",
    "    dplyr::group_by(motif_id, type, species,gene_name) %>% \n",
    "    dplyr::summarise(count_motif_occurence = n()) %>% \n",
    "    left_join(., all_species_total_counts, by = c(\"type\", \"species\")) %>% \n",
    "    dplyr::rowwise() %>% \n",
    "    mutate(non_motif_occurence = counts - count_motif_occurence) %>% \n",
    "    dplyr::filter(gene_name %in% c(\"CA2\", \"CA1\", \"CA6\", \"CA_1__CA_2\", \"CA_2\", \"CA6_ortho_count_2\")) %>% \n",
    "    arrange(motif_id, type, species)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f93e650e-5735-47bb-b945-e701a1cdc4b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'PPT'</li><li>'NAD-MDH2'</li><li>'PEPCK'</li><li>'NADP_ME'</li><li>'NAD-MDH12'</li><li>'CA2'</li><li>'CA1'</li><li>'CA6'</li><li>'RUBISCO_activase1'</li><li>'RBCA2'</li><li>'RBCA1'</li><li>'PPDK'</li><li>'NADP_ME1'</li><li>'PEPC1'</li><li>'AlaAT3_ortho_count_2__AlaAT4_ortho_count_2'</li><li>'NAD-ME_2'</li><li>'NAD-ME_1'</li><li>'CA_1'</li><li>'PPDK_1'</li><li>'RUBISCO_activase1_ortho_count_1'</li><li>'NADP-ME_1'</li><li>'AlaAT2_ortho_count_1'</li><li>'AlaAT2_ortho_count_2'</li><li>'PPT_4'</li><li>'PEP-CK_3'</li><li>'NADP-MDH_1'</li><li>'PPDK-RP'</li><li>'AlaAT1_ortho_count_1'</li><li>'AlaAT3'</li><li>'NADP-ME_2'</li><li>'PPT_3'</li><li>'NAD-MDH1'</li><li>'CA_2'</li><li>'CA6_ortho_count_2'</li><li>'RBCA2_ortho_count_7__RBCA1_ortho_count_7'</li><li>'BASS2'</li><li>'PEPC_2'</li><li>'RBCA2_ortho_count_4__RBCA1_ortho_count_4'</li><li>'AspAT_1'</li><li>'PPDK_2'</li><li>'RBCA2_ortho_count_1__RBCA1_ortho_count_1'</li><li>'RBCA2_ortho_count_2__RBCA1_ortho_count_2'</li><li>'RBCA2_ortho_count_3__RBCA1_ortho_count_3'</li><li>'PEPC_1'</li><li>'PPT_2'</li><li>'NHD2'</li><li>'NADP-MDH_2'</li><li>'DIC1_'</li><li>'PEP-CK_2'</li><li>'Dit_1'</li><li>'AspAT_2'</li><li>'NAD-ME_1__NAD-ME_2'</li><li>'NAD_ME1'</li><li>'AlaAT1'</li><li>'DIT1'</li><li>'DIT2'</li><li>'AlaAT8'</li><li>'CA_1__CA_2'</li><li>'PPA2_2__PPA3_2'</li><li>'DIC1__ortho_count_3'</li><li>'DIC1__ortho_count_2'</li><li>'NAD_ME12'</li><li>'Dit_1__Dit_2'</li><li>'AlaAT2'</li><li>'NADP-MDH_1__NADP-MDH_2'</li><li>'NHD1__NHD2'</li><li>'NADP_ME1_ortho_count_2'</li><li>'NADP_ME1_ortho_count_1'</li><li>'ASP_AT1'</li><li>'RBCA2__RBCA1'</li><li>'CA6_ortho_count_1'</li><li>'DIT2_ortho_count_1'</li><li>'BASS'</li><li>'NADP_MDH'</li><li>'AlaAT3_ortho_count_1__AlaAT4_ortho_count_1__AlaAT3_ortho_count_2__AlaAT4_ortho_count_2'</li><li>'PEPC1_ortho_count_1'</li><li>'NAD_ME10'</li><li>'NADP_ME1__NADP-ME_1_ortho_count_1__NADP-ME_2_ortho_count_1'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'PPT'\n",
       "\\item 'NAD-MDH2'\n",
       "\\item 'PEPCK'\n",
       "\\item 'NADP\\_ME'\n",
       "\\item 'NAD-MDH12'\n",
       "\\item 'CA2'\n",
       "\\item 'CA1'\n",
       "\\item 'CA6'\n",
       "\\item 'RUBISCO\\_activase1'\n",
       "\\item 'RBCA2'\n",
       "\\item 'RBCA1'\n",
       "\\item 'PPDK'\n",
       "\\item 'NADP\\_ME1'\n",
       "\\item 'PEPC1'\n",
       "\\item 'AlaAT3\\_ortho\\_count\\_2\\_\\_AlaAT4\\_ortho\\_count\\_2'\n",
       "\\item 'NAD-ME\\_2'\n",
       "\\item 'NAD-ME\\_1'\n",
       "\\item 'CA\\_1'\n",
       "\\item 'PPDK\\_1'\n",
       "\\item 'RUBISCO\\_activase1\\_ortho\\_count\\_1'\n",
       "\\item 'NADP-ME\\_1'\n",
       "\\item 'AlaAT2\\_ortho\\_count\\_1'\n",
       "\\item 'AlaAT2\\_ortho\\_count\\_2'\n",
       "\\item 'PPT\\_4'\n",
       "\\item 'PEP-CK\\_3'\n",
       "\\item 'NADP-MDH\\_1'\n",
       "\\item 'PPDK-RP'\n",
       "\\item 'AlaAT1\\_ortho\\_count\\_1'\n",
       "\\item 'AlaAT3'\n",
       "\\item 'NADP-ME\\_2'\n",
       "\\item 'PPT\\_3'\n",
       "\\item 'NAD-MDH1'\n",
       "\\item 'CA\\_2'\n",
       "\\item 'CA6\\_ortho\\_count\\_2'\n",
       "\\item 'RBCA2\\_ortho\\_count\\_7\\_\\_RBCA1\\_ortho\\_count\\_7'\n",
       "\\item 'BASS2'\n",
       "\\item 'PEPC\\_2'\n",
       "\\item 'RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4'\n",
       "\\item 'AspAT\\_1'\n",
       "\\item 'PPDK\\_2'\n",
       "\\item 'RBCA2\\_ortho\\_count\\_1\\_\\_RBCA1\\_ortho\\_count\\_1'\n",
       "\\item 'RBCA2\\_ortho\\_count\\_2\\_\\_RBCA1\\_ortho\\_count\\_2'\n",
       "\\item 'RBCA2\\_ortho\\_count\\_3\\_\\_RBCA1\\_ortho\\_count\\_3'\n",
       "\\item 'PEPC\\_1'\n",
       "\\item 'PPT\\_2'\n",
       "\\item 'NHD2'\n",
       "\\item 'NADP-MDH\\_2'\n",
       "\\item 'DIC1\\_'\n",
       "\\item 'PEP-CK\\_2'\n",
       "\\item 'Dit\\_1'\n",
       "\\item 'AspAT\\_2'\n",
       "\\item 'NAD-ME\\_1\\_\\_NAD-ME\\_2'\n",
       "\\item 'NAD\\_ME1'\n",
       "\\item 'AlaAT1'\n",
       "\\item 'DIT1'\n",
       "\\item 'DIT2'\n",
       "\\item 'AlaAT8'\n",
       "\\item 'CA\\_1\\_\\_CA\\_2'\n",
       "\\item 'PPA2\\_2\\_\\_PPA3\\_2'\n",
       "\\item 'DIC1\\_\\_ortho\\_count\\_3'\n",
       "\\item 'DIC1\\_\\_ortho\\_count\\_2'\n",
       "\\item 'NAD\\_ME12'\n",
       "\\item 'Dit\\_1\\_\\_Dit\\_2'\n",
       "\\item 'AlaAT2'\n",
       "\\item 'NADP-MDH\\_1\\_\\_NADP-MDH\\_2'\n",
       "\\item 'NHD1\\_\\_NHD2'\n",
       "\\item 'NADP\\_ME1\\_ortho\\_count\\_2'\n",
       "\\item 'NADP\\_ME1\\_ortho\\_count\\_1'\n",
       "\\item 'ASP\\_AT1'\n",
       "\\item 'RBCA2\\_\\_RBCA1'\n",
       "\\item 'CA6\\_ortho\\_count\\_1'\n",
       "\\item 'DIT2\\_ortho\\_count\\_1'\n",
       "\\item 'BASS'\n",
       "\\item 'NADP\\_MDH'\n",
       "\\item 'AlaAT3\\_ortho\\_count\\_1\\_\\_AlaAT4\\_ortho\\_count\\_1\\_\\_AlaAT3\\_ortho\\_count\\_2\\_\\_AlaAT4\\_ortho\\_count\\_2'\n",
       "\\item 'PEPC1\\_ortho\\_count\\_1'\n",
       "\\item 'NAD\\_ME10'\n",
       "\\item 'NADP\\_ME1\\_\\_NADP-ME\\_1\\_ortho\\_count\\_1\\_\\_NADP-ME\\_2\\_ortho\\_count\\_1'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'PPT'\n",
       "2. 'NAD-MDH2'\n",
       "3. 'PEPCK'\n",
       "4. 'NADP_ME'\n",
       "5. 'NAD-MDH12'\n",
       "6. 'CA2'\n",
       "7. 'CA1'\n",
       "8. 'CA6'\n",
       "9. 'RUBISCO_activase1'\n",
       "10. 'RBCA2'\n",
       "11. 'RBCA1'\n",
       "12. 'PPDK'\n",
       "13. 'NADP_ME1'\n",
       "14. 'PEPC1'\n",
       "15. 'AlaAT3_ortho_count_2__AlaAT4_ortho_count_2'\n",
       "16. 'NAD-ME_2'\n",
       "17. 'NAD-ME_1'\n",
       "18. 'CA_1'\n",
       "19. 'PPDK_1'\n",
       "20. 'RUBISCO_activase1_ortho_count_1'\n",
       "21. 'NADP-ME_1'\n",
       "22. 'AlaAT2_ortho_count_1'\n",
       "23. 'AlaAT2_ortho_count_2'\n",
       "24. 'PPT_4'\n",
       "25. 'PEP-CK_3'\n",
       "26. 'NADP-MDH_1'\n",
       "27. 'PPDK-RP'\n",
       "28. 'AlaAT1_ortho_count_1'\n",
       "29. 'AlaAT3'\n",
       "30. 'NADP-ME_2'\n",
       "31. 'PPT_3'\n",
       "32. 'NAD-MDH1'\n",
       "33. 'CA_2'\n",
       "34. 'CA6_ortho_count_2'\n",
       "35. 'RBCA2_ortho_count_7__RBCA1_ortho_count_7'\n",
       "36. 'BASS2'\n",
       "37. 'PEPC_2'\n",
       "38. 'RBCA2_ortho_count_4__RBCA1_ortho_count_4'\n",
       "39. 'AspAT_1'\n",
       "40. 'PPDK_2'\n",
       "41. 'RBCA2_ortho_count_1__RBCA1_ortho_count_1'\n",
       "42. 'RBCA2_ortho_count_2__RBCA1_ortho_count_2'\n",
       "43. 'RBCA2_ortho_count_3__RBCA1_ortho_count_3'\n",
       "44. 'PEPC_1'\n",
       "45. 'PPT_2'\n",
       "46. 'NHD2'\n",
       "47. 'NADP-MDH_2'\n",
       "48. 'DIC1_'\n",
       "49. 'PEP-CK_2'\n",
       "50. 'Dit_1'\n",
       "51. 'AspAT_2'\n",
       "52. 'NAD-ME_1__NAD-ME_2'\n",
       "53. 'NAD_ME1'\n",
       "54. 'AlaAT1'\n",
       "55. 'DIT1'\n",
       "56. 'DIT2'\n",
       "57. 'AlaAT8'\n",
       "58. 'CA_1__CA_2'\n",
       "59. 'PPA2_2__PPA3_2'\n",
       "60. 'DIC1__ortho_count_3'\n",
       "61. 'DIC1__ortho_count_2'\n",
       "62. 'NAD_ME12'\n",
       "63. 'Dit_1__Dit_2'\n",
       "64. 'AlaAT2'\n",
       "65. 'NADP-MDH_1__NADP-MDH_2'\n",
       "66. 'NHD1__NHD2'\n",
       "67. 'NADP_ME1_ortho_count_2'\n",
       "68. 'NADP_ME1_ortho_count_1'\n",
       "69. 'ASP_AT1'\n",
       "70. 'RBCA2__RBCA1'\n",
       "71. 'CA6_ortho_count_1'\n",
       "72. 'DIT2_ortho_count_1'\n",
       "73. 'BASS'\n",
       "74. 'NADP_MDH'\n",
       "75. 'AlaAT3_ortho_count_1__AlaAT4_ortho_count_1__AlaAT3_ortho_count_2__AlaAT4_ortho_count_2'\n",
       "76. 'PEPC1_ortho_count_1'\n",
       "77. 'NAD_ME10'\n",
       "78. 'NADP_ME1__NADP-ME_1_ortho_count_1__NADP-ME_2_ortho_count_1'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"PPT\"                                                                                   \n",
       " [2] \"NAD-MDH2\"                                                                              \n",
       " [3] \"PEPCK\"                                                                                 \n",
       " [4] \"NADP_ME\"                                                                               \n",
       " [5] \"NAD-MDH12\"                                                                             \n",
       " [6] \"CA2\"                                                                                   \n",
       " [7] \"CA1\"                                                                                   \n",
       " [8] \"CA6\"                                                                                   \n",
       " [9] \"RUBISCO_activase1\"                                                                     \n",
       "[10] \"RBCA2\"                                                                                 \n",
       "[11] \"RBCA1\"                                                                                 \n",
       "[12] \"PPDK\"                                                                                  \n",
       "[13] \"NADP_ME1\"                                                                              \n",
       "[14] \"PEPC1\"                                                                                 \n",
       "[15] \"AlaAT3_ortho_count_2__AlaAT4_ortho_count_2\"                                            \n",
       "[16] \"NAD-ME_2\"                                                                              \n",
       "[17] \"NAD-ME_1\"                                                                              \n",
       "[18] \"CA_1\"                                                                                  \n",
       "[19] \"PPDK_1\"                                                                                \n",
       "[20] \"RUBISCO_activase1_ortho_count_1\"                                                       \n",
       "[21] \"NADP-ME_1\"                                                                             \n",
       "[22] \"AlaAT2_ortho_count_1\"                                                                  \n",
       "[23] \"AlaAT2_ortho_count_2\"                                                                  \n",
       "[24] \"PPT_4\"                                                                                 \n",
       "[25] \"PEP-CK_3\"                                                                              \n",
       "[26] \"NADP-MDH_1\"                                                                            \n",
       "[27] \"PPDK-RP\"                                                                               \n",
       "[28] \"AlaAT1_ortho_count_1\"                                                                  \n",
       "[29] \"AlaAT3\"                                                                                \n",
       "[30] \"NADP-ME_2\"                                                                             \n",
       "[31] \"PPT_3\"                                                                                 \n",
       "[32] \"NAD-MDH1\"                                                                              \n",
       "[33] \"CA_2\"                                                                                  \n",
       "[34] \"CA6_ortho_count_2\"                                                                     \n",
       "[35] \"RBCA2_ortho_count_7__RBCA1_ortho_count_7\"                                              \n",
       "[36] \"BASS2\"                                                                                 \n",
       "[37] \"PEPC_2\"                                                                                \n",
       "[38] \"RBCA2_ortho_count_4__RBCA1_ortho_count_4\"                                              \n",
       "[39] \"AspAT_1\"                                                                               \n",
       "[40] \"PPDK_2\"                                                                                \n",
       "[41] \"RBCA2_ortho_count_1__RBCA1_ortho_count_1\"                                              \n",
       "[42] \"RBCA2_ortho_count_2__RBCA1_ortho_count_2\"                                              \n",
       "[43] \"RBCA2_ortho_count_3__RBCA1_ortho_count_3\"                                              \n",
       "[44] \"PEPC_1\"                                                                                \n",
       "[45] \"PPT_2\"                                                                                 \n",
       "[46] \"NHD2\"                                                                                  \n",
       "[47] \"NADP-MDH_2\"                                                                            \n",
       "[48] \"DIC1_\"                                                                                 \n",
       "[49] \"PEP-CK_2\"                                                                              \n",
       "[50] \"Dit_1\"                                                                                 \n",
       "[51] \"AspAT_2\"                                                                               \n",
       "[52] \"NAD-ME_1__NAD-ME_2\"                                                                    \n",
       "[53] \"NAD_ME1\"                                                                               \n",
       "[54] \"AlaAT1\"                                                                                \n",
       "[55] \"DIT1\"                                                                                  \n",
       "[56] \"DIT2\"                                                                                  \n",
       "[57] \"AlaAT8\"                                                                                \n",
       "[58] \"CA_1__CA_2\"                                                                            \n",
       "[59] \"PPA2_2__PPA3_2\"                                                                        \n",
       "[60] \"DIC1__ortho_count_3\"                                                                   \n",
       "[61] \"DIC1__ortho_count_2\"                                                                   \n",
       "[62] \"NAD_ME12\"                                                                              \n",
       "[63] \"Dit_1__Dit_2\"                                                                          \n",
       "[64] \"AlaAT2\"                                                                                \n",
       "[65] \"NADP-MDH_1__NADP-MDH_2\"                                                                \n",
       "[66] \"NHD1__NHD2\"                                                                            \n",
       "[67] \"NADP_ME1_ortho_count_2\"                                                                \n",
       "[68] \"NADP_ME1_ortho_count_1\"                                                                \n",
       "[69] \"ASP_AT1\"                                                                               \n",
       "[70] \"RBCA2__RBCA1\"                                                                          \n",
       "[71] \"CA6_ortho_count_1\"                                                                     \n",
       "[72] \"DIT2_ortho_count_1\"                                                                    \n",
       "[73] \"BASS\"                                                                                  \n",
       "[74] \"NADP_MDH\"                                                                              \n",
       "[75] \"AlaAT3_ortho_count_1__AlaAT4_ortho_count_1__AlaAT3_ortho_count_2__AlaAT4_ortho_count_2\"\n",
       "[76] \"PEPC1_ortho_count_1\"                                                                   \n",
       "[77] \"NAD_ME10\"                                                                              \n",
       "[78] \"NADP_ME1__NADP-ME_1_ortho_count_1__NADP-ME_2_ortho_count_1\"                            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(all_species_fimo_counts$gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f3a4f169-a8f6-4108-bb57-9c218534a600",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'motif_id', 'type', 'species'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 23 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>type</th><th scope=col>species</th><th scope=col>gene_name</th><th scope=col>count_motif_occurence</th><th scope=col>counts</th><th scope=col>non_motif_occurence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB      </td><td>bs</td><td>pm</td><td>RBCA2_ortho_count_4__RBCA1_ortho_count_4</td><td>2</td><td>54</td><td>52</td></tr>\n",
       "\t<tr><td>AS2LOB      </td><td>ms</td><td>zm</td><td>RBCA1                                   </td><td>1</td><td>29</td><td>28</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>bs</td><td>uf</td><td>RUBISCO_activase1                       </td><td>1</td><td>35</td><td>34</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>ms</td><td>zm</td><td>RBCA1                                   </td><td>2</td><td>29</td><td>27</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>ms</td><td>zm</td><td>RBCA2                                   </td><td>4</td><td>29</td><td>25</td></tr>\n",
       "\t<tr><td>CPP         </td><td>bs</td><td>pm</td><td>RBCA2_ortho_count_4__RBCA1_ortho_count_4</td><td>1</td><td>54</td><td>53</td></tr>\n",
       "\t<tr><td>CPP         </td><td>bs</td><td>uf</td><td>RUBISCO_activase1                       </td><td>1</td><td>35</td><td>34</td></tr>\n",
       "\t<tr><td>GARP_G2-like</td><td>bs</td><td>pm</td><td>RBCA2_ortho_count_4__RBCA1_ortho_count_4</td><td>1</td><td>54</td><td>53</td></tr>\n",
       "\t<tr><td>HD-Zip_IV   </td><td>bs</td><td>sb</td><td>RUBISCO_activase1                       </td><td>1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>HD-Zip_IV   </td><td>bs</td><td>uf</td><td>RUBISCO_activase1                       </td><td>2</td><td>35</td><td>33</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II </td><td>bs</td><td>pm</td><td>RBCA2_ortho_count_4__RBCA1_ortho_count_4</td><td>1</td><td>54</td><td>53</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II </td><td>bs</td><td>zm</td><td>RBCA1                                   </td><td>1</td><td>13</td><td>12</td></tr>\n",
       "\t<tr><td>HD-Zip_I_II </td><td>bs</td><td>zm</td><td>RUBISCO_activase1                       </td><td>1</td><td>13</td><td>12</td></tr>\n",
       "\t<tr><td>HSF         </td><td>bs</td><td>sb</td><td>RUBISCO_activase1                       </td><td>2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>HSF         </td><td>bs</td><td>zm</td><td>RUBISCO_activase1                       </td><td>1</td><td>13</td><td>12</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>bs</td><td>sb</td><td>RUBISCO_activase1                       </td><td>1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>bs</td><td>uf</td><td>RUBISCO_activase1                       </td><td>1</td><td>35</td><td>34</td></tr>\n",
       "\t<tr><td>MADS_MIKC   </td><td>bs</td><td>zm</td><td>RUBISCO_activase1                       </td><td>1</td><td>13</td><td>12</td></tr>\n",
       "\t<tr><td>NAC         </td><td>bs</td><td>pm</td><td>RBCA2_ortho_count_4__RBCA1_ortho_count_4</td><td>2</td><td>54</td><td>52</td></tr>\n",
       "\t<tr><td>WRKY        </td><td>bs</td><td>pm</td><td>RBCA2_ortho_count_4__RBCA1_ortho_count_4</td><td>1</td><td>54</td><td>53</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>bs</td><td>pm</td><td>RBCA2_ortho_count_4__RBCA1_ortho_count_4</td><td>1</td><td>54</td><td>53</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>bs</td><td>sb</td><td>RUBISCO_activase1                       </td><td>1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>bHLH_TCP    </td><td>bs</td><td>zm</td><td>RUBISCO_activase1                       </td><td>2</td><td>13</td><td>11</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 23 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " motif\\_id & type & species & gene\\_name & count\\_motif\\_occurence & counts & non\\_motif\\_occurence\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t AS2LOB       & bs & pm & RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4 & 2 & 54 & 52\\\\\n",
       "\t AS2LOB       & ms & zm & RBCA1                                    & 1 & 29 & 28\\\\\n",
       "\t BBRBPC       & bs & uf & RUBISCO\\_activase1                        & 1 & 35 & 34\\\\\n",
       "\t BBRBPC       & ms & zm & RBCA1                                    & 2 & 29 & 27\\\\\n",
       "\t BBRBPC       & ms & zm & RBCA2                                    & 4 & 29 & 25\\\\\n",
       "\t CPP          & bs & pm & RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4 & 1 & 54 & 53\\\\\n",
       "\t CPP          & bs & uf & RUBISCO\\_activase1                        & 1 & 35 & 34\\\\\n",
       "\t GARP\\_G2-like & bs & pm & RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4 & 1 & 54 & 53\\\\\n",
       "\t HD-Zip\\_IV    & bs & sb & RUBISCO\\_activase1                        & 1 & 14 & 13\\\\\n",
       "\t HD-Zip\\_IV    & bs & uf & RUBISCO\\_activase1                        & 2 & 35 & 33\\\\\n",
       "\t HD-Zip\\_I\\_II  & bs & pm & RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4 & 1 & 54 & 53\\\\\n",
       "\t HD-Zip\\_I\\_II  & bs & zm & RBCA1                                    & 1 & 13 & 12\\\\\n",
       "\t HD-Zip\\_I\\_II  & bs & zm & RUBISCO\\_activase1                        & 1 & 13 & 12\\\\\n",
       "\t HSF          & bs & sb & RUBISCO\\_activase1                        & 2 & 14 & 12\\\\\n",
       "\t HSF          & bs & zm & RUBISCO\\_activase1                        & 1 & 13 & 12\\\\\n",
       "\t MADS\\_MIKC    & bs & sb & RUBISCO\\_activase1                        & 1 & 14 & 13\\\\\n",
       "\t MADS\\_MIKC    & bs & uf & RUBISCO\\_activase1                        & 1 & 35 & 34\\\\\n",
       "\t MADS\\_MIKC    & bs & zm & RUBISCO\\_activase1                        & 1 & 13 & 12\\\\\n",
       "\t NAC          & bs & pm & RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4 & 2 & 54 & 52\\\\\n",
       "\t WRKY         & bs & pm & RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4 & 1 & 54 & 53\\\\\n",
       "\t bHLH\\_TCP     & bs & pm & RBCA2\\_ortho\\_count\\_4\\_\\_RBCA1\\_ortho\\_count\\_4 & 1 & 54 & 53\\\\\n",
       "\t bHLH\\_TCP     & bs & sb & RUBISCO\\_activase1                        & 1 & 14 & 13\\\\\n",
       "\t bHLH\\_TCP     & bs & zm & RUBISCO\\_activase1                        & 2 & 13 & 11\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 23 × 7\n",
       "\n",
       "| motif_id &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; | gene_name &lt;chr&gt; | count_motif_occurence &lt;int&gt; | counts &lt;int&gt; | non_motif_occurence &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| AS2LOB       | bs | pm | RBCA2_ortho_count_4__RBCA1_ortho_count_4 | 2 | 54 | 52 |\n",
       "| AS2LOB       | ms | zm | RBCA1                                    | 1 | 29 | 28 |\n",
       "| BBRBPC       | bs | uf | RUBISCO_activase1                        | 1 | 35 | 34 |\n",
       "| BBRBPC       | ms | zm | RBCA1                                    | 2 | 29 | 27 |\n",
       "| BBRBPC       | ms | zm | RBCA2                                    | 4 | 29 | 25 |\n",
       "| CPP          | bs | pm | RBCA2_ortho_count_4__RBCA1_ortho_count_4 | 1 | 54 | 53 |\n",
       "| CPP          | bs | uf | RUBISCO_activase1                        | 1 | 35 | 34 |\n",
       "| GARP_G2-like | bs | pm | RBCA2_ortho_count_4__RBCA1_ortho_count_4 | 1 | 54 | 53 |\n",
       "| HD-Zip_IV    | bs | sb | RUBISCO_activase1                        | 1 | 14 | 13 |\n",
       "| HD-Zip_IV    | bs | uf | RUBISCO_activase1                        | 2 | 35 | 33 |\n",
       "| HD-Zip_I_II  | bs | pm | RBCA2_ortho_count_4__RBCA1_ortho_count_4 | 1 | 54 | 53 |\n",
       "| HD-Zip_I_II  | bs | zm | RBCA1                                    | 1 | 13 | 12 |\n",
       "| HD-Zip_I_II  | bs | zm | RUBISCO_activase1                        | 1 | 13 | 12 |\n",
       "| HSF          | bs | sb | RUBISCO_activase1                        | 2 | 14 | 12 |\n",
       "| HSF          | bs | zm | RUBISCO_activase1                        | 1 | 13 | 12 |\n",
       "| MADS_MIKC    | bs | sb | RUBISCO_activase1                        | 1 | 14 | 13 |\n",
       "| MADS_MIKC    | bs | uf | RUBISCO_activase1                        | 1 | 35 | 34 |\n",
       "| MADS_MIKC    | bs | zm | RUBISCO_activase1                        | 1 | 13 | 12 |\n",
       "| NAC          | bs | pm | RBCA2_ortho_count_4__RBCA1_ortho_count_4 | 2 | 54 | 52 |\n",
       "| WRKY         | bs | pm | RBCA2_ortho_count_4__RBCA1_ortho_count_4 | 1 | 54 | 53 |\n",
       "| bHLH_TCP     | bs | pm | RBCA2_ortho_count_4__RBCA1_ortho_count_4 | 1 | 54 | 53 |\n",
       "| bHLH_TCP     | bs | sb | RUBISCO_activase1                        | 1 | 14 | 13 |\n",
       "| bHLH_TCP     | bs | zm | RUBISCO_activase1                        | 2 | 13 | 11 |\n",
       "\n"
      ],
      "text/plain": [
       "   motif_id     type species gene_name                               \n",
       "1  AS2LOB       bs   pm      RBCA2_ortho_count_4__RBCA1_ortho_count_4\n",
       "2  AS2LOB       ms   zm      RBCA1                                   \n",
       "3  BBRBPC       bs   uf      RUBISCO_activase1                       \n",
       "4  BBRBPC       ms   zm      RBCA1                                   \n",
       "5  BBRBPC       ms   zm      RBCA2                                   \n",
       "6  CPP          bs   pm      RBCA2_ortho_count_4__RBCA1_ortho_count_4\n",
       "7  CPP          bs   uf      RUBISCO_activase1                       \n",
       "8  GARP_G2-like bs   pm      RBCA2_ortho_count_4__RBCA1_ortho_count_4\n",
       "9  HD-Zip_IV    bs   sb      RUBISCO_activase1                       \n",
       "10 HD-Zip_IV    bs   uf      RUBISCO_activase1                       \n",
       "11 HD-Zip_I_II  bs   pm      RBCA2_ortho_count_4__RBCA1_ortho_count_4\n",
       "12 HD-Zip_I_II  bs   zm      RBCA1                                   \n",
       "13 HD-Zip_I_II  bs   zm      RUBISCO_activase1                       \n",
       "14 HSF          bs   sb      RUBISCO_activase1                       \n",
       "15 HSF          bs   zm      RUBISCO_activase1                       \n",
       "16 MADS_MIKC    bs   sb      RUBISCO_activase1                       \n",
       "17 MADS_MIKC    bs   uf      RUBISCO_activase1                       \n",
       "18 MADS_MIKC    bs   zm      RUBISCO_activase1                       \n",
       "19 NAC          bs   pm      RBCA2_ortho_count_4__RBCA1_ortho_count_4\n",
       "20 WRKY         bs   pm      RBCA2_ortho_count_4__RBCA1_ortho_count_4\n",
       "21 bHLH_TCP     bs   pm      RBCA2_ortho_count_4__RBCA1_ortho_count_4\n",
       "22 bHLH_TCP     bs   sb      RUBISCO_activase1                       \n",
       "23 bHLH_TCP     bs   zm      RUBISCO_activase1                       \n",
       "   count_motif_occurence counts non_motif_occurence\n",
       "1  2                     54     52                 \n",
       "2  1                     29     28                 \n",
       "3  1                     35     34                 \n",
       "4  2                     29     27                 \n",
       "5  4                     29     25                 \n",
       "6  1                     54     53                 \n",
       "7  1                     35     34                 \n",
       "8  1                     54     53                 \n",
       "9  1                     14     13                 \n",
       "10 2                     35     33                 \n",
       "11 1                     54     53                 \n",
       "12 1                     13     12                 \n",
       "13 1                     13     12                 \n",
       "14 2                     14     12                 \n",
       "15 1                     13     12                 \n",
       "16 1                     14     13                 \n",
       "17 1                     35     34                 \n",
       "18 1                     13     12                 \n",
       "19 2                     54     52                 \n",
       "20 1                     54     53                 \n",
       "21 1                     54     53                 \n",
       "22 1                     14     13                 \n",
       "23 2                     13     11                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_species_fimo_counts %>% \n",
    "    dplyr::select(motif_id, type,species, gene_name) %>% \n",
    "    dplyr::group_by(motif_id, type, species,gene_name) %>% \n",
    "    dplyr::summarise(count_motif_occurence = n()) %>% \n",
    "    left_join(., all_species_total_counts, by = c(\"type\", \"species\")) %>% \n",
    "    dplyr::rowwise() %>% \n",
    "    mutate(non_motif_occurence = counts - count_motif_occurence) %>% \n",
    "    dplyr::filter(gene_name %in% c(\"RBCA2_ortho_count_4__RBCA1_ortho_count_4\", \"RBCA2_ortho_count_2__RBCA1_ortho_count_2-CK_2\",\n",
    "                                  \"RUBISCO_activase1\", \"RBCA2\", \"RBCA1\")) %>% \n",
    "    arrange(motif_id, type, species)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "529bbccd-eee0-4b44-a6e4-3751869bda0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'motif_id', 'type'. You can override using\n",
      "the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "collpsed_by_species_cell_type <- all_species_fimo_counts %>% \n",
    "    dplyr::select(motif_id, acrID, type,species) %>% \n",
    "    unique() %>% \n",
    "    dplyr::group_by(motif_id, type, species) %>% \n",
    "    dplyr::summarise(count_motif_occurence = n()) %>% \n",
    "    left_join(., all_species_total_counts, by = c(\"type\", \"species\")) %>% \n",
    "    dplyr::rowwise() %>% \n",
    "    mutate(non_motif_occurence = counts - count_motif_occurence)\n",
    "\n",
    "\n",
    "complete_combination <- expand.grid(\n",
    "  motif_id = unique(fimo_counts_combined_null$motif_id),\n",
    "  type = unique(fimo_counts_combined_null$type),\n",
    "  species = unique(fimo_counts_combined_null$species)\n",
    ")\n",
    "\n",
    "null_collapsed_species_cell_type <- fimo_counts_combined_null %>%\n",
    "  dplyr::select(motif_id, acrID, type, species) %>% \n",
    "  unique() %>% \n",
    "  dplyr::group_by(motif_id, type, species) %>% \n",
    "  dplyr::summarise(null_count_motif_occurence = n(), .groups = 'drop') %>% \n",
    "  # left join with complete combination to ensure every possible instance exists\n",
    "  right_join(complete_combination, by = c(\"motif_id\", \"type\", \"species\")) %>% \n",
    "  replace_na(list(null_count_motif_occurence = 0)) %>% # replace NA with 0\n",
    "  left_join(., all_species_total_counts, by = c(\"type\", \"species\")) %>% \n",
    "  dplyr::rowwise() %>% \n",
    "  mutate(null_count_non_motif_occurence = counts - null_count_motif_occurence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "74199f06-9e6c-4f21-bc9e-07f747bd4cab",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_null_real <- left_join(collpsed_by_species_cell_type, null_collapsed_species_cell_type,by = join_by(motif_id, type, species, counts)) %>% \n",
    "    dplyr::select(-counts) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b38cf419-694e-43be-8b7e-c286b954c08a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 115 × 8\u001b[39m\n",
      "\u001b[90m# Rowwise: \u001b[39m\n",
      "   motif_id type  species count_motif_occurence non_motif_occurence\n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m               \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m 1\u001b[39m AS2LOB   bs    os                          3                  11\n",
      "\u001b[90m 2\u001b[39m AS2LOB   bs    pm                          7                  47\n",
      "\u001b[90m 3\u001b[39m AS2LOB   bs    sb                          1                  13\n",
      "\u001b[90m 4\u001b[39m AS2LOB   bs    uf                          9                  26\n",
      "\u001b[90m 5\u001b[39m AS2LOB   bs    zm                          1                  12\n",
      "\u001b[90m 6\u001b[39m AS2LOB   ms    os                          9                  16\n",
      "\u001b[90m 7\u001b[39m AS2LOB   ms    pm                          6                  32\n",
      "\u001b[90m 8\u001b[39m AS2LOB   ms    sb                          6                  22\n",
      "\u001b[90m 9\u001b[39m AS2LOB   ms    uf                          8                  25\n",
      "\u001b[90m10\u001b[39m AS2LOB   ms    zm                          5                  24\n",
      "\u001b[90m# ℹ 105 more rows\u001b[39m\n",
      "\u001b[90m# ℹ 3 more variables: null_count_motif_occurence <int>,\u001b[39m\n",
      "\u001b[90m#   null_count_non_motif_occurence <int>, p_value <dbl>\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "fisher_test_from_row <- function(count_motif_occurence, non_motif_occurence, null_count_motif_occurence, null_count_non_motif_occurence) {\n",
    "  matrix_data <- c(count_motif_occurence, null_count_motif_occurence, non_motif_occurence, null_count_non_motif_occurence)\n",
    "  matrix_for_test <- matrix(matrix_data, nrow = 2)\n",
    "  fisher_result <- fisher.test(matrix_for_test)\n",
    "  return(fisher_result$p.value)\n",
    "}\n",
    "\n",
    "# Apply the function to each row\n",
    "combined_null_real.pvals <- combined_null_real %>%\n",
    "  rowwise() %>%\n",
    "  mutate(p_value = purrr::pmap_dbl(list(count_motif_occurence, non_motif_occurence, null_count_motif_occurence, null_count_non_motif_occurence), fisher_test_from_row))\n",
    "\n",
    "print(combined_null_real.pvals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "abf67e9c-9368-4ddc-a455-89eaea1328bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 3 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>type</th><th scope=col>species</th><th scope=col>count_motif_occurence</th><th scope=col>non_motif_occurence</th><th scope=col>null_count_motif_occurence</th><th scope=col>null_count_non_motif_occurence</th><th scope=col>p_value</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB      </td><td>ms</td><td>pm</td><td>6</td><td>32</td><td>16</td><td>22</td><td>0.02164525</td></tr>\n",
       "\t<tr><td>BBRBPC      </td><td>bs</td><td>pm</td><td>8</td><td>46</td><td>19</td><td>35</td><td>0.02518857</td></tr>\n",
       "\t<tr><td>GARP_G2-like</td><td>bs</td><td>pm</td><td>8</td><td>46</td><td> 1</td><td>53</td><td>0.03144815</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 3 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " motif\\_id & type & species & count\\_motif\\_occurence & non\\_motif\\_occurence & null\\_count\\_motif\\_occurence & null\\_count\\_non\\_motif\\_occurence & p\\_value\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <int> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AS2LOB       & ms & pm & 6 & 32 & 16 & 22 & 0.02164525\\\\\n",
       "\t BBRBPC       & bs & pm & 8 & 46 & 19 & 35 & 0.02518857\\\\\n",
       "\t GARP\\_G2-like & bs & pm & 8 & 46 &  1 & 53 & 0.03144815\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 3 × 8\n",
       "\n",
       "| motif_id &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; | count_motif_occurence &lt;int&gt; | non_motif_occurence &lt;int&gt; | null_count_motif_occurence &lt;int&gt; | null_count_non_motif_occurence &lt;int&gt; | p_value &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| AS2LOB       | ms | pm | 6 | 32 | 16 | 22 | 0.02164525 |\n",
       "| BBRBPC       | bs | pm | 8 | 46 | 19 | 35 | 0.02518857 |\n",
       "| GARP_G2-like | bs | pm | 8 | 46 |  1 | 53 | 0.03144815 |\n",
       "\n"
      ],
      "text/plain": [
       "  motif_id     type species count_motif_occurence non_motif_occurence\n",
       "1 AS2LOB       ms   pm      6                     32                 \n",
       "2 BBRBPC       bs   pm      8                     46                 \n",
       "3 GARP_G2-like bs   pm      8                     46                 \n",
       "  null_count_motif_occurence null_count_non_motif_occurence p_value   \n",
       "1 16                         22                             0.02164525\n",
       "2 19                         35                             0.02518857\n",
       "3  1                         53                             0.03144815"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "combined_null_real.pvals %>% \n",
    "    dplyr::filter(p_value < .05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8f3f9419-ac2f-4ff5-bc5f-89149b5c6f4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>type</th><th scope=col>species</th><th scope=col>count_motif_occurence</th><th scope=col>counts</th><th scope=col>non_motif_occurence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>os</td><td>3</td><td>14</td><td>11</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>pm</td><td>7</td><td>54</td><td>47</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>sb</td><td>1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>uf</td><td>9</td><td>35</td><td>26</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>zm</td><td>1</td><td>13</td><td>12</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>ms</td><td>os</td><td>9</td><td>25</td><td>16</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 6 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " motif\\_id & type & species & count\\_motif\\_occurence & counts & non\\_motif\\_occurence\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t AS2LOB & bs & os & 3 & 14 & 11\\\\\n",
       "\t AS2LOB & bs & pm & 7 & 54 & 47\\\\\n",
       "\t AS2LOB & bs & sb & 1 & 14 & 13\\\\\n",
       "\t AS2LOB & bs & uf & 9 & 35 & 26\\\\\n",
       "\t AS2LOB & bs & zm & 1 & 13 & 12\\\\\n",
       "\t AS2LOB & ms & os & 9 & 25 & 16\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 6 × 6\n",
       "\n",
       "| motif_id &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; | count_motif_occurence &lt;int&gt; | counts &lt;int&gt; | non_motif_occurence &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| AS2LOB | bs | os | 3 | 14 | 11 |\n",
       "| AS2LOB | bs | pm | 7 | 54 | 47 |\n",
       "| AS2LOB | bs | sb | 1 | 14 | 13 |\n",
       "| AS2LOB | bs | uf | 9 | 35 | 26 |\n",
       "| AS2LOB | bs | zm | 1 | 13 | 12 |\n",
       "| AS2LOB | ms | os | 9 | 25 | 16 |\n",
       "\n"
      ],
      "text/plain": [
       "  motif_id type species count_motif_occurence counts non_motif_occurence\n",
       "1 AS2LOB   bs   os      3                     14     11                 \n",
       "2 AS2LOB   bs   pm      7                     54     47                 \n",
       "3 AS2LOB   bs   sb      1                     14     13                 \n",
       "4 AS2LOB   bs   uf      9                     35     26                 \n",
       "5 AS2LOB   bs   zm      1                     13     12                 \n",
       "6 AS2LOB   ms   os      9                     25     16                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif_id</th><th scope=col>type</th><th scope=col>species</th><th scope=col>null_count_motif_occurence</th><th scope=col>counts</th><th scope=col>null_count_non_motif_occurence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>os</td><td> 2</td><td>14</td><td>12</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>pm</td><td>11</td><td>54</td><td>43</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>sb</td><td> 1</td><td>14</td><td>13</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>uf</td><td> 7</td><td>35</td><td>28</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>bs</td><td>zm</td><td> 3</td><td>13</td><td>10</td></tr>\n",
       "\t<tr><td>AS2LOB</td><td>ms</td><td>os</td><td> 9</td><td>25</td><td>16</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 6 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " motif\\_id & type & species & null\\_count\\_motif\\_occurence & counts & null\\_count\\_non\\_motif\\_occurence\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t AS2LOB & bs & os &  2 & 14 & 12\\\\\n",
       "\t AS2LOB & bs & pm & 11 & 54 & 43\\\\\n",
       "\t AS2LOB & bs & sb &  1 & 14 & 13\\\\\n",
       "\t AS2LOB & bs & uf &  7 & 35 & 28\\\\\n",
       "\t AS2LOB & bs & zm &  3 & 13 & 10\\\\\n",
       "\t AS2LOB & ms & os &  9 & 25 & 16\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 6 × 6\n",
       "\n",
       "| motif_id &lt;chr&gt; | type &lt;chr&gt; | species &lt;chr&gt; | null_count_motif_occurence &lt;int&gt; | counts &lt;int&gt; | null_count_non_motif_occurence &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| AS2LOB | bs | os |  2 | 14 | 12 |\n",
       "| AS2LOB | bs | pm | 11 | 54 | 43 |\n",
       "| AS2LOB | bs | sb |  1 | 14 | 13 |\n",
       "| AS2LOB | bs | uf |  7 | 35 | 28 |\n",
       "| AS2LOB | bs | zm |  3 | 13 | 10 |\n",
       "| AS2LOB | ms | os |  9 | 25 | 16 |\n",
       "\n"
      ],
      "text/plain": [
       "  motif_id type species null_count_motif_occurence counts\n",
       "1 AS2LOB   bs   os       2                         14    \n",
       "2 AS2LOB   bs   pm      11                         54    \n",
       "3 AS2LOB   bs   sb       1                         14    \n",
       "4 AS2LOB   bs   uf       7                         35    \n",
       "5 AS2LOB   bs   zm       3                         13    \n",
       "6 AS2LOB   ms   os       9                         25    \n",
       "  null_count_non_motif_occurence\n",
       "1 12                            \n",
       "2 43                            \n",
       "3 13                            \n",
       "4 28                            \n",
       "5 10                            \n",
       "6 16                            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(collpsed_by_species_cell_type)\n",
    "head(null_collapsed_species_cell_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d65ac54-ebfc-4a87-9538-8d2911340aa3",
   "metadata": {},
   "source": [
    "## Look into specific genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "eba20557-0fc6-4058-af0e-ed8d04c8b1c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'species', 'type', 'gene_name'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "just_counts <- all_species_fimo_counts %>% \n",
    "    group_by(species, type, gene_name,motif_id) %>% \n",
    "    dplyr::summarise(motif_count = n()) %>% \n",
    "    dplyr::arrange(desc(motif_count))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "24f66084-5d70-4725-94c9-3a144b3af008",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 16 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>species</th><th scope=col>type</th><th scope=col>gene_name</th><th scope=col>motif_id</th><th scope=col>motif_count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>E2FDP       </td><td>4</td></tr>\n",
       "\t<tr><td>sb</td><td>ms</td><td>PPDK</td><td>BBRBPC      </td><td>3</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>BBRBPC      </td><td>2</td></tr>\n",
       "\t<tr><td>sb</td><td>ms</td><td>PPDK</td><td>E2FDP       </td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>bs</td><td>PPDK</td><td>AS2LOB      </td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>AS2LOB      </td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>CPP         </td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>GARP_ARR-B  </td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>GARP_G2-like</td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>HD          </td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>HSF         </td><td>1</td></tr>\n",
       "\t<tr><td>uf</td><td>ms</td><td>PPDK</td><td>MADS_MIKC   </td><td>1</td></tr>\n",
       "\t<tr><td>zm</td><td>ms</td><td>PPDK</td><td>BBRBPC      </td><td>1</td></tr>\n",
       "\t<tr><td>zm</td><td>ms</td><td>PPDK</td><td>CPP         </td><td>1</td></tr>\n",
       "\t<tr><td>zm</td><td>ms</td><td>PPDK</td><td>HD-Zip_I_II </td><td>1</td></tr>\n",
       "\t<tr><td>zm</td><td>ms</td><td>PPDK</td><td>HSF         </td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 16 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " species & type & gene\\_name & motif\\_id & motif\\_count\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t uf & ms & PPDK & E2FDP        & 4\\\\\n",
       "\t sb & ms & PPDK & BBRBPC       & 3\\\\\n",
       "\t uf & ms & PPDK & BBRBPC       & 2\\\\\n",
       "\t sb & ms & PPDK & E2FDP        & 1\\\\\n",
       "\t uf & bs & PPDK & AS2LOB       & 1\\\\\n",
       "\t uf & ms & PPDK & AS2LOB       & 1\\\\\n",
       "\t uf & ms & PPDK & CPP          & 1\\\\\n",
       "\t uf & ms & PPDK & GARP\\_ARR-B   & 1\\\\\n",
       "\t uf & ms & PPDK & GARP\\_G2-like & 1\\\\\n",
       "\t uf & ms & PPDK & HD           & 1\\\\\n",
       "\t uf & ms & PPDK & HSF          & 1\\\\\n",
       "\t uf & ms & PPDK & MADS\\_MIKC    & 1\\\\\n",
       "\t zm & ms & PPDK & BBRBPC       & 1\\\\\n",
       "\t zm & ms & PPDK & CPP          & 1\\\\\n",
       "\t zm & ms & PPDK & HD-Zip\\_I\\_II  & 1\\\\\n",
       "\t zm & ms & PPDK & HSF          & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 16 × 5\n",
       "\n",
       "| species &lt;chr&gt; | type &lt;chr&gt; | gene_name &lt;chr&gt; | motif_id &lt;chr&gt; | motif_count &lt;int&gt; |\n",
       "|---|---|---|---|---|\n",
       "| uf | ms | PPDK | E2FDP        | 4 |\n",
       "| sb | ms | PPDK | BBRBPC       | 3 |\n",
       "| uf | ms | PPDK | BBRBPC       | 2 |\n",
       "| sb | ms | PPDK | E2FDP        | 1 |\n",
       "| uf | bs | PPDK | AS2LOB       | 1 |\n",
       "| uf | ms | PPDK | AS2LOB       | 1 |\n",
       "| uf | ms | PPDK | CPP          | 1 |\n",
       "| uf | ms | PPDK | GARP_ARR-B   | 1 |\n",
       "| uf | ms | PPDK | GARP_G2-like | 1 |\n",
       "| uf | ms | PPDK | HD           | 1 |\n",
       "| uf | ms | PPDK | HSF          | 1 |\n",
       "| uf | ms | PPDK | MADS_MIKC    | 1 |\n",
       "| zm | ms | PPDK | BBRBPC       | 1 |\n",
       "| zm | ms | PPDK | CPP          | 1 |\n",
       "| zm | ms | PPDK | HD-Zip_I_II  | 1 |\n",
       "| zm | ms | PPDK | HSF          | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   species type gene_name motif_id     motif_count\n",
       "1  uf      ms   PPDK      E2FDP        4          \n",
       "2  sb      ms   PPDK      BBRBPC       3          \n",
       "3  uf      ms   PPDK      BBRBPC       2          \n",
       "4  sb      ms   PPDK      E2FDP        1          \n",
       "5  uf      bs   PPDK      AS2LOB       1          \n",
       "6  uf      ms   PPDK      AS2LOB       1          \n",
       "7  uf      ms   PPDK      CPP          1          \n",
       "8  uf      ms   PPDK      GARP_ARR-B   1          \n",
       "9  uf      ms   PPDK      GARP_G2-like 1          \n",
       "10 uf      ms   PPDK      HD           1          \n",
       "11 uf      ms   PPDK      HSF          1          \n",
       "12 uf      ms   PPDK      MADS_MIKC    1          \n",
       "13 zm      ms   PPDK      BBRBPC       1          \n",
       "14 zm      ms   PPDK      CPP          1          \n",
       "15 zm      ms   PPDK      HD-Zip_I_II  1          \n",
       "16 zm      ms   PPDK      HSF          1          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "just_counts %>% \n",
    "    dplyr::filter(gene_name == \"PPDK\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
